---
title: "R Notebook"
output:
    html_document:
      toc: true
      toc_float: true

---


```{r message=FALSE, warning = FALSE}
library(gprofiler2)
library(viridis)
library(stringr)
library(signatureSearch)
library(ggplot2)
library(dplyr)
library(here)
library(viridis)
```

## Custom functions for FEA

Enrichment function for up and down genes
```{r}
enrich <- function(upgenes, downgenes, topreturned) {#, filepath) {
  up_fea <- gost(upgenes, multi_query = FALSE, correction_method = "bonferroni", evcodes = TRUE)
  #remove abitrary pathways -- any pathways with > 1000 genes
    up_fea <- up_fea$result %>% dplyr::filter(., term_size < 1000 & term_size > 5) %>% dplyr::mutate(., .keep = "all", direction = "upregulated")
    up_mat <- up_fea  %>% dplyr::arrange(., desc(intersection_size)) %>% dplyr::slice_head(., n = topreturned)
  
  down_fea <- gost(downgenes, multi_query = FALSE, correction_method = "bonferroni", evcodes = TRUE)
  #remove abitrary pathways -- any pathways with > 1000 genes
  down_fea <- down_fea$result %>% dplyr::filter(., term_size < 1000 & term_size > 5) %>% dplyr::mutate(., .keep = "all", direction = "downregulated") 
  down_mat <- down_fea  %>% dplyr::arrange(., desc(intersection_size)) %>% dplyr::slice_head(., n = topreturned)
  combined_fea <- rbind(up_mat, down_mat)
  
  fea_res <- list(downregulated = down_fea, upregulated = up_fea, topcompared = combined_fea)

return(fea_res)
}
```

MOUSE GENES Enrichment function for up and down genes  

```{r}
#Runs an ordered query
fea <- function(genelist, topreturned){
  fea <- gost(genelist, organism = "mmusculus", ordered_query = FALSE, multi_query = FALSE, evcodes = TRUE, correction_method = "bonferroni", sources = c("GO:BP", "GO:MF", "GO:CC", "KEGG", "REAC", "MIRNA", "CORUM", "HP", "HPA", "WP"))
  #remove abitrary pathways -- any pathways with > 1000 genes
  fea <- fea$result %>% dplyr::filter(., term_size < 1000 & term_size > 10)
return(fea)
}

mouseenrich <- function(upgenes, downgenes, topreturned) {
  upgenes$LFC <- sort(as.numeric(upgenes$LFC), decreasing = TRUE)
  up_fea <- fea(upgenes$Mouse.gene.stable.ID, topreturned$Mouse.gene.stable.ID)
  up_fea <- dplyr::mutate(up_fea, .keep = "all", direction = "upregulated")
  
  downgenes$LFC <- sort(as.numeric(downgenes$LFC))
  down_fea <- fea(downgenes$Mouse.gene.stable.ID, topreturned$Mouse.gene.stable.ID)
  down_fea <- dplyr::mutate(down_fea, .keep = "all", direction = "downregulated")
  
  up_mat <- up_fea  %>% dplyr::arrange(., desc(recall)) %>% dplyr::slice_head(., n = topreturned)
  down_mat <- down_fea  %>% dplyr::arrange(., desc(recall)) %>% dplyr::slice_head(., n = topreturned)
  combined_fea <- rbind(up_mat, down_mat)
  fea_res <- list(downregulated = down_fea, upregulated = up_fea, topcompared = combined_fea)

return(fea_res)
}

```

Custom background for FEA (for enrichment using mouse to human to entrez to lincs conversion -- background gene list is all possible measured genes that have gone through those conversions)
```{r}
#Needs to only be possible genes -- all genes measured in salmon gene counts that map to human entrez
#mesgenes <- data@NAMES
#remove version numbers
#mesgenes <- gsub("\\..*", "", mesgenes)

#genebg <- convertMouseGeneList(mesgenes)
#write.csv2(genebg, file = "./data/background_genelist_genestoLINCSgenes.csv")

genebg <- read.csv(file = here("data", "background_genelist_genestoLINCSgenes.csv"))
```


bubbleplot, up and down for x axis and terms on y axis
```{r}
bubbleplot <- function(combined_fea){
  bubbleplot <- ggplot(combined_fea, aes(x=direction, y=term_name, size = intersection_size, fill = p_value)) +
  geom_point(alpha=0.7, shape = 21) +
  scale_size(range = c(2, 10), name = "# Genes Matched to Term") + 
  scale_fill_distiller(palette = "Purples") + 
  labs(x = "Differential Expression Direction", y = "Functional Enrichment Terms") + 
  theme_minimal(base_size = 8) #+ labs(title = "Top Enriched Terms")
#  ggsave(filepath, bubbleplot, width = 10, height = 7)
return(bubbleplot)
}
```

barplot, up and down merged, colored by p-value, terms on y axis and # genes on x axis
```{r}
barplot <- function(combined_fea){
  barplot <- ggplot(combined_fea, aes(x=intersection_size, y=term_name,  fill = recall)) + 
    geom_bar( stat = "identity") + 
    scale_fill_continuous(type = "viridis") + 
    labs(x = "Number of Genes Matched to Term", y = "Functional Terms") + 
    theme_minimal(base_size = 8)
return(barplot)
}

```

subsetting terms that are mitochondria dysfunction-related
```{r eval=FALSE, include=FALSE}
mito_terms <- function(fea_res){
  up <- fea_res$upregulated
  down <- fea_res$downregulated
  tbl <- rbind(up, down)
  
  metabol <- stringr::str_subset(tbl$term_name, "metabol")
  mitochon <- stringr::str_subset(tbl$term_name, "mitochon")
  oxid <- stringr::str_subset(tbl$term_name, "oxid")
  #signal <- stringr::str_subset(tbl$term_name, "signal")
  NAD <- stringr::str_subset(tbl$term_name, "NAD")
  glyc <- stringr::str_subset(tbl$term_name, "glyc")
  cAMP <- stringr::str_subset(tbl$term_name, "cAMP")
  mTOR <- stringr::str_subset(tbl$term_name, "mTOR")
  AMPK <- stringr::str_subset(tbl$term_name, "AMPK")
  ERK <- stringr::str_subset(tbl$term_name, "ERK")
  calcium <- stringr::str_subset(tbl$term_name, "calcium")
  MAPK <- stringr::str_subset(tbl$term_name, "MAPK")
  ATPase <-  stringr::str_subset(tbl$term_name, "ATPase")
  ATP <- stringr::str_subset(tbl$term_name, "ATP")
  STIM1 <- stringr::str_subset(tbl$term_name, "STIM1")
  #channel <- stringr::str_subset(tbl$term_name, "channel")
  TRP <-  stringr::str_subset(tbl$term_name, "TRP")
  NFAT <-  stringr::str_subset(tbl$term_name, "NFAT")
  
  mitoterms <- c(metabol, mitochon, oxid, NAD, glyc, cAMP, mTOR, AMPK, ERK, calcium, MAPK, ATPase, ATP, STIM1, TRP, NFAT)
  mitoterms <- as.data.frame(mitoterms)
  colnames(mitoterms) <- c("term_name")
  
  mitochondrial_terms <- right_join(tbl, mitoterms, by = "term_name")
return(mitochondrial_terms)
}
  

```

subsetting terms that are  ciliopathy-related terms
```{r eval=FALSE, include=FALSE}
cilia_terms <- function(fea_res){
  up <- fea_res$upregulated
  down <- fea_res$downregulated
  tbl <- rbind(up, down)
  
  gpcr <- stringr::str_subset(tbl$term_name, "GPCR")
  sens <- stringr::str_subset(tbl$term_name, "sens")
  trans <- stringr::str_subset(tbl$term_name, "trans") #transduction and transport roles
  cilia <- c(gpcr, sens, trans)
  cilia <- as.data.frame(cilia)
  colnames(cilia) <- c("term_name")
  #print(mitoterms)
  cilia_terms <- right_join(tbl, cilia, by = "term_name")
return(cilia_terms)
}


```

Subsetting for fibrosis-realted terms
```{r eval=FALSE, include=FALSE}
fibrosis_terms <- function(fea_res){
  up <- fea_res$upregulated
  down <- fea_res$downregulated
  tbl <- rbind(up, down)
  
  mitosis <- stringr::str_subset(tbl$term_name, "mito")
  cycle <- stringr::str_subset(tbl$term_name, "cell cycle")
  collagen <- stringr::str_subset(tbl$term_name, "collagen") 
  ECM <- stringr::str_subset(tbl$term_name, "ECM")
  extracellular <- stringr::str_subset(tbl$term_name, "extracellular")
  TGF <- stringr::str_subset(tbl$term_name, "TGF") 
  growth <- stringr::str_subset(tbl$term_name, "growth")
  fibro <- stringr::str_subset(tbl$term_name, "fibro")
  collagen <- stringr::str_subset(tbl$term_name, "collagen")
  TIMP <- stringr::str_subset(tbl$term_name, "TIMP")
  PAI <- stringr::str_subset(tbl$term_name, "PAI") 
  EGFR <- stringr::str_subset(tbl$term_name, "EGFR")
  prolif <- stringr::str_subset(tbl$term_name, "prolif")
  miR17 <- stringr::str_subset(tbl$term_name, "miR17") 
  
  fibrosis_terms <- c(mitosis, cycle, collagen, ECM, extracellular, TGF, growth, fibro, collagen, TIMP, PAI, EGFR, prolif, miR17)
  fibrosis_terms <- as.data.frame(fibrosis_terms)
  colnames(fibrosis_terms) <- c("term_name")
  #print(mitoterms)
  fibrosis_terms <- right_join(tbl, fibrosis_terms, by = "term_name")
return(fibrosis_terms)
}

```

immune terms
Subsetting for fibrosis-realted terms
```{r eval=FALSE, include=FALSE}
immune_terms <- function(fea_res){
  up <- fea_res$upregulated
  down <- fea_res$downregulated
  tbl <- rbind(up, down)
  
  cytok <- stringr::str_subset(tbl$term_name, "cytok")
  inflam <- stringr::str_subset(tbl$term_name, "inflam")
  immun <- stringr::str_subset(tbl$term_name, "immun")
  leuk <- stringr::str_subset(tbl$term_name, "leuk")
  extracellular <- stringr::str_subset(tbl$term_name, "extracellular")
  ECM <- stringr::str_subset(tbl$term_name, "ECM")
  macroph <- stringr::str_subset(tbl$term_name, "macroph")
  monoc <- stringr::str_subset(tbl$term_name, "monoc")
  MHC <- stringr::str_subset(tbl$term_name, "MHC")
  CD <- stringr::str_subset(tbl$term_name, "CD")
  Tcell <- stringr::str_subset(tbl$term_name, "T cell")
  complement <- stringr::str_subset(tbl$term_name, "complement")
  Bcell <- stringr::str_subset(tbl$term_name, "B cell")
  kcell <- stringr::str_subset(tbl$term_name, "killer cell")
  hcell <- stringr::str_subset(tbl$term_name, "helper cell")
  FCERI <- stringr::str_subset(tbl$term_name, "FCERI")
  
  immune_terms <- c(cytok, inflam, immun, leuk, extracellular, ECM, macroph, monoc, MHC, CD, Tcell, complement, Bcell, kcell, hcell, FCERI)
  immune_terms <- as.data.frame(immune_terms)
  colnames(immune_terms) <- c("term_name")

  immune_terms <- right_join(tbl, immune_terms, by = "term_name")
return(immune_terms)
}

```

subsetting terms that are SKF-96365 related (a store-operated Ca2+ entry (SOCE) inhibitor)
```{r eval=FALSE, include=FALSE}
soce <- function(up, down){
  tbl <- rbind(up, down)
  
  ERK <- stringr::str_subset(tbl$term_name, "ERK")
  calc <- stringr::str_subset(tbl$term_name, "calc")
  MAPK <- stringr::str_subset(tbl$term_name, "MAPK")
  ATPase <-  stringr::str_subset(tbl$term_name, "ATPase")
  ATP <- stringr::str_subset(tbl$term_name, "ATP")
  STIM1 <- stringr::str_subset(tbl$term_name, "STIM1")
  channel <- stringr::str_subset(tbl$term_name, "channel")
  TRP <-  stringr::str_subset(tbl$term_name, "TRP")
  NFAT <-  stringr::str_subset(tbl$term_name, "NFAT")
  soce <- c(ERK, calc, MAPK, ATPase, ATP, STIM1, channel, TRP, NFAT)
  soce <- as.data.frame(soce)
  colnames(soce) <- c("term_name")
  #print(mitoterms)
  soce_terms <- right_join(tbl, soce, by = "term_name")
  
return(soce_terms)
}

```

 ### Load in (top 100 human) data  
Load in top 100 up and down genes used as signatures for signatureSearch

'39 precystic p70
```{r}
hom_degs_UP <- read.csv(here("res", "deseq2_outputs", "GSE149739_degs_UP_220421.csv"))

hom_degs_DOWN <- read.csv(here("res", "deseq2_outputs", "GSE149739_degs_DOWN_220421.csv"))
```

'19 cystic p28
```{r}
hom_19_UP <- read.csv(here("res", "deseq2_outputs", "GSE134719_degs_UP_220421.csv"))
hom_19_DOWN <- read.csv(here("res", "deseq2_outputs", "GSE134719_degs_DOWN_220421.csv"))
```

'56 cystic p21
```{r}
hom_56_UP <- read.csv(here("res", "deseq2_outputs", "GSE69556_degs_UP_220421.csv"))

hom_56_DOWN <- read.csv(here("res", "deseq2_outputs", "GSE69556_degs_DOWN_220421.csv"))
```


Comparing results across datasets 

"Hom" refers to homosapiens -- genes that have been mapped from mouse to human
### Compare number of DEGs - datasets
```{r eval=FALSE, include=FALSE}
#in common, 19 and 56
comp19vs56_intersect_up <- dplyr::intersect(hom_19_UP$entrezgene_id, hom_56_UP$entrezgene_id) 
length(comp19vs56_intersect_up)
comp19vs56_intersect_down <- dplyr::intersect(hom_19_DOWN$entrezgene_id, hom_19_DOWN$entrezgene_id) 
length(comp19vs56_intersect_down)

#in common, 39 and 56
comp39vs56_intersect_up <- dplyr::intersect(hom_degs_UP$entrezgene_id, hom_56_UP$entrezgene_id) 
length(comp39vs56_intersect_up)

comp39vs56_intersect_down <- dplyr::intersect(hom_degs_DOWN$entrezgene_id, hom_56_DOWN$entrezgene_id) 
length(comp39vs56_intersect_down)

#difference, 39 and 56, specific to 56
comp39vs56_56specific_up <- dplyr::setdiff(hom_56_UP$entrezgene_id, hom_degs_UP$entrezgene_id) # specific to 56 compared to 39
length(comp39vs56_56specific_up)
comp39vs56_56specific_down <- dplyr::setdiff(hom_56_DOWN$entrezgene_id, hom_degs_DOWN$entrezgene_id) # specific to 56 compared to 39
length(comp39vs56_56specific_down)

#difference, 39 and 56, specific to 39
comp39vs56_39specific_up <-  dplyr::setdiff(hom_degs_UP$entrezgene_id, hom_56_UP$entrezgene_id) # specific to 39 compared to 56
length(comp39vs56_39specific_up)
comp39vs56_39specific_down <- dplyr::setdiff(hom_degs_DOWN$entrezgene_id, hom_56_DOWN$entrezgene_id) # specific to 39 compared to 56
length(comp39vs56_39specific_down)

#in common, 39 and 19
comp39vs19_intersect_up <- dplyr::intersect(hom_degs_UP$entrezgene_id, hom_19_UP$entrezgene_id) 
length(comp39vs19_intersect_up)

comp39vs19_intersect_down <- dplyr::intersect(hom_degs_DOWN$entrezgene_id, hom_19_DOWN$entrezgene_id) 
length(comp39vs19_intersect_down)
```


### FEA p70 precystic GSE149739  (top 100 human orthologs)  
```{r}
fea_nf_GSE149739 <- enrich(as.character(hom_degs_UP$entrezgene_id), as.character(hom_degs_DOWN$entrezgene_id), 30)
```

Barplot of overall top enrichment terms
```{r}
bp_39 <- barplot(fea_nf_GSE149739$topcompared)
bp_39
```

Bubbleplot of cilia-related enriched terms
```{r eval=FALSE, include=FALSE}
cilia_39 <- cilia_terms(fea_nf_GSE149739)
bubbleplot(cilia_39) + labs(title = "Cilia-related Enriched Terms")
```

Bubbleplot of mitochondria-related enriched terms
```{r eval=FALSE, include=FALSE}
mito_39 <- mito_terms(fea_nf_GSE149739)
bubbleplot(mito_39) + labs(title = "Mitochondria-related Enriched Terms")
```

SOCE-related pathways for '19 and '56 overlap
```{r eval=FALSE, include=FALSE}
soce_39 <- soce(fea_nf_GSE149739$upregulated, fea_nf_GSE149739$downregulated)
bubbleplot(soce_39) + labs(title = "SOCE-related Enriched Terms")
```

### FEA p21 cystic GSE69556 (top 100 human orthologs)  
```{r}
fea_GSE69556 <- enrich(as.character(hom_56_UP$entrezgene_id), as.character(hom_56_DOWN$entrezgene_id), 30)
```

Barplot of overall top enrichment terms
```{r}
bp_56 <- barplot(fea_GSE69556$topcompared)
bp_56
```

Bubbleplot of cilia-related enriched terms
```{r eval=FALSE, include=FALSE}
cilia_56 <- cilia_terms(fea_GSE69556)
bubbleplot(cilia_56) + labs(title = "Cilia-related Enriched Terms")
```

Bubbleplot of mitochondria-related enriched terms
```{r eval=FALSE, include=FALSE}
mito_56 <- mito_terms(fea_GSE69556)
bubbleplot(mito_56) + labs(title = "Mitochondria-related Enriched Terms")
```

SOCE-related pathways for '19 and '56 overlap
```{r eval=FALSE, include=FALSE}
soce_56 <- soce(fea_GSE69556$upregulated, fea_GSE69556$downregulated)
bubbleplot(soce_56) + labs(title = "SOCE-related Enriched Terms")
```

### FEA p28 cystic GSE134719 (top 100 human orthologs)
```{r}
fea_GSE134719 <- enrich(as.character(hom_19_UP$entrezgene_id), as.character(hom_19_DOWN$entrezgene_id), 30)
```

Barplot of overall top enrichment terms
```{r}
bar19 <- barplot(fea_GSE134719$topcompared)
bar19
```

Bubbleplot of cilia-related enriched terms
```{r eval=FALSE, include=FALSE}
cilia_19 <- cilia_terms(fea_GSE134719)
bubbleplot(cilia_19) + labs(title = "Cilia-related Enriched Terms")
```

Bubbleplot of mito-related enriched terms
```{r eval=FALSE, include=FALSE}
mito_19 <- mito_terms(fea_GSE134719)
bubbleplot(mito_19) + labs(title = "Mitochondria-related Enriched Terms")
```

SOCE-related pathways for '19 and '56 overlap
```{r eval=FALSE, include=FALSE}
soce_19 <- soce(fea_GSE134719$upregulated, fea_GSE134719$downregulated)
bubbleplot(soce_19) + labs(title = "SOCE-related Enriched Terms")
```

Enrichment comparisons
```{r eval=FALSE, include=FALSE}
#enrichment for overlapping genes
nf_orig_overlap_fea <- enrich(pipecompare_intersect_up, pipecompare_intersect_down, 15) 

#enrichment for original alignment-specific genes
orig_specifc_fea <- enrich(pipecompare_oldspecific_up, pipecompare_oldspecific_down, 15)

#enrichment for nf alignment-specific genes
nf_specifc_fea <- enrich(pipecompare_nfspecific_up, pipecompare_nfspecific_down, 15)
```

Save
```{r eval=FALSE, include=FALSE}
##fea_nf_GSE149739
fea_nf_GSE149739_upterms <- apply(fea_nf_GSE149739$upregulated, 2, as.character)
write.csv(fea_nf_GSE149739_upterms, file = here("res", "fea", "fea_nf_GSE149739_upterms.csv"))

fea_nf_GSE149739_downterms <- apply(fea_nf_GSE149739$downregulated, 2, as.character)
write.csv(fea_nf_GSE149739_downterms, file = here("res", "fea", "fea_nf_GSE149739_downterms.csv"))

##fea_GSE69556
fea_GSE69556_upterms <- apply(fea_GSE69556$upregulated, 2, as.character)
write.csv(fea_GSE69556_upterms, file = here("res", "fea", "fea_GSE69556_upterms.csv"))

fea_GSE69556_downterms <- apply(fea_GSE69556$downregulated, 2, as.character)
write.csv(fea_GSE69556_downterms, file = here("res", "fea", "fea_GSE69556_downterms.csv"))

##fea_GSE134719
fea_GSE134719_upterms <- apply(fea_GSE134719$upregulated, 2, as.character)
write.csv(fea_GSE134719_upterms, file = here("res", "fea", "fea_GSE134719_upterms.csv"))

fea_GSE134719_downterms <- apply(fea_GSE134719$downregulated, 2, as.character)
write.csv(fea_GSE134719_downterms, file = here("res", "fea", "fea_GSE134719_downterms.csv"))

```

## Mouse genes FEA  
Using ALL mouse DEGs (0.05 alpha and -2 and 2 LFC cutoff) before mapping to human genes or selecting top 100 available in LINCS  

### Load in (mouse DEG) data

'39 nf-core aligned
```{r}
musdegs_39_UP <- read.csv(here("res", "deseq2_outputs", "GSE149739_mus_degs_UP_220421.csv"))

musdegs_39_DOWN <- read.csv(here("res", "deseq2_outputs", "GSE149739_mus_degs_DOWN_220421.csv"))
```

'19
```{r}
musdegs_19_UP <- read.csv(here("res", "deseq2_outputs", "GSE134719_mus_degs_UP_220421.csv"))
musdegs_19_DOWN <- read.csv(here("res", "deseq2_outputs", "GSE134719_mus_degs_DOWN_220421.csv"))
```

'56
```{r}
musdegs_56_UP <- read.csv(here("res", "deseq2_outputs", "GSE69556_mus_degs_UP_220421.csv"))

musdegs_56_DOWN <- read.csv(here("res", "deseq2_outputs", "GSE69556_mus_degs_DOWN_220421.csv"))
```

#### Gene overlaps
```{r}
degs_19_56_overlap_up <- merge(musdegs_19_UP, musdegs_56_UP, by = "Mouse.gene.stable.ID")

degs_19_56_overlap_down <- merge(musdegs_19_DOWN, musdegs_56_DOWN, by = "Mouse.gene.stable.ID")

degs_19_56_39_overlap_up <- merge(degs_19_56_overlap_up, musdegs_39_UP, by = "Mouse.gene.stable.ID")

degs_19_56_39_overlap_down <- merge(degs_19_56_overlap_down, musdegs_39_DOWN, by = "Mouse.gene.stable.ID")

degs_39_56_overlap_up <- merge(musdegs_39_UP, musdegs_56_UP, by = "Mouse.gene.stable.ID")

degs_39_56_overlap_down <- merge(musdegs_39_DOWN, musdegs_56_DOWN, by = "Mouse.gene.stable.ID")


degs_39_19_overlap_up <- merge(musdegs_39_UP, musdegs_19_UP, by = "Mouse.gene.stable.ID")

degs_39_19_overlap_down <- merge(musdegs_39_DOWN, musdegs_19_DOWN, by = "Mouse.gene.stable.ID")
```

### FEA p70 precystic '39 Mouse
```{r}
fea_mouse_39 <- mouseenrich(musdegs_39_UP, musdegs_39_DOWN, 30)

bp_mouse_39_overall <- barplot(dplyr::filter(fea_mouse_39$topcompared, intersection_size > 10)) + ggtitle("Pre-Cystic Top Upregulated ") 
bp_mouse_39_overall

fea_mouse_39$topcompared$term_name <- stringr::str_wrap(fea_mouse_39$topcompared$term_name, width = 60)

p70_fea_plot <- ggplot(dplyr::filter(fea_mouse_39$topcompared, intersection_size > 10), aes(x=intersection_size, y=term_name,  fill = recall)) + 
    geom_bar( stat = "identity") + 
    scale_fill_continuous(type = "viridis") + 
    labs(x = "Number of Genes Matched to Term", y = "Functional Terms", title = "Pre-Cystic Top Upregulated Enrichment") + 
    theme(text = element_text(size = 14), plot.title = element_text(size = 24))# +
    #theme_minimal(base_size = 8)
p70_fea_plot
```

bubbleplot
```{r}
bubbleplot_p70 <- bubbleplot(dplyr::filter(fea_mouse_39$topcompared, intersection_size > 6)) + labs(title = "Top Enriched Terms for Pre-Cystic P70 Dataset") + theme(text = element_text(size = 14), plot.title = element_text(size = 24))

bubbleplot(fea_mouse_39$topcompared) + labs(title = "Top Enriched Terms for Pre-Cystic P70 Dataset") + theme(text = element_text(size = 14), plot.title = element_text(size = 24))
```

```{r}
ggsave(p70_fea_plot, filename = here("res", "fea", "p70_fea_plot.pdf"))

ggsave(bubbleplot_p70, filename = here("res", "fea", "p70_fea_bubbleplot.pdf"), height = 7, width = 14)
```

bubbleplot top
```{r}
bubbleplot(dplyr::filter(fea_mouse_39$topcompared, intersection_size > 10)) #+ labs(title = "")
```


Cilia-related pathways
```{r eval=FALSE, include=FALSE}
cilia_mouse_39 <- cilia_terms(fea_mouse_39)
cilia_mouse_39_plot  <- bubbleplot(cilia_mouse_39) + labs(title = "Cilia-related Enriched Terms")
cilia_mouse_39_plot
```

Mitochondria-related pathways 
```{r eval=FALSE, include=FALSE}
mito_mouse_39 <- mito_terms(fea_mouse_39)
mito_mouse_39_plot <- bubbleplot(mito_mouse_39) + labs(title = "Mitochondria-related Enriched Terms")
mito_mouse_39_plot
```

Store-operated CA 2+ Entry pathways
```{r eval=FALSE, include=FALSE}
soce_39 <- soce(fea_mouse_39$upregulated, fea_mouse_39$downregulated)
mouse_39_soce <- bubbleplot(soce_39) + labs(title = "Store-operated CA 2+ Entry-related Enriched Terms")
mouse_39_soce
```

Fibrosis-related terms
```{r eval=FALSE, include=FALSE}
fibrosis_mouse_39 <- fibrosis_terms(fea_mouse_39)
fibrosis_mouse_39_plot <- bubbleplot(fibrosis_mouse_39) + labs(title = "Fibrosis-related Enriched Terms")
fibrosis_mouse_39_plot

```

Immune-related terms
```{r eval=FALSE, include=FALSE}
immune_mouse_39 <- immune_terms(fea_mouse_39)
immune_mouse_39_plot <- bubbleplot(immune_mouse_39) + labs(title = "Immune-related Enriched Terms")
immune_mouse_39_plot

```


Save plots
```{r}
ggsave(filename = here("res", "fea", "mouse_39_overall.pdf"), bp_mouse_39_overall)


```

### FEA p21 cystic '56 Mouse
```{r}
fea_mouse_56 <- mouseenrich(musdegs_56_UP, musdegs_56_DOWN, 30)

bp_mouse_56_overall <- barplot(dplyr::filter(fea_mouse_56$topcompared, intersection_size > 10))
bp_mouse_56_overall
```

bubbleplot
```{r}
fea_mouse_56$topcompared$term_name <- stringr::str_wrap(fea_mouse_56$topcompared$term_name, width = 70)


bubbleplot_p21 <- bubbleplot(dplyr::filter(fea_mouse_56$topcompared, intersection_size > 10)) + labs(title = "Top Enriched Terms for Cystic P21 Dataset") + theme(text = element_text(size = 14), plot.title = element_text(size = 24))
```

```{r}
ggsave(bubbleplot_p21, filename = here("res", "fea", "p21_fea_bubbleplot.pdf"), height = 7, width = 14)

```

Cilia-related pathways
```{r eval=FALSE, include=FALSE}
cilia_mouse_56 <- cilia_terms(fea_mouse_56)
cilia_mouse_56_plot <- bubbleplot(cilia_mouse_56) + labs(title = "Cilia-related Enriched Terms")
cilia_mouse_56_plot
```

Mitochondria-related pathways 
```{r eval=FALSE, include=FALSE}
mito_mouse_56 <- mito_terms(fea_mouse_56)
mito_mouse_56_plot <- bubbleplot(mito_mouse_56) + labs(title = "Mitochondria-related Enriched Terms")
mito_mouse_56_plot
```

Fibrosis-related terms
```{r eval=FALSE, include=FALSE}
fibrosis_mouse_56 <- fibrosis_terms(fea_mouse_56)
fibrosis_mouse_56_plot <- bubbleplot(fibrosis_mouse_56) + labs(title = "Fibrosis-related Enriched Terms")
fibrosis_mouse_56_plot

```

Immune-related terms
```{r eval=FALSE, include=FALSE}
immune_mouse_56 <- immune_terms(fea_mouse_56)
immune_mouse_56_plot <- bubbleplot(immune_mouse_56) + labs(title = "Immune-related Enriched Terms")
immune_mouse_56_plot

```

Store-operated CA 2+ Entry pathways
```{r eval=FALSE, include=FALSE}
soce_56 <- soce(fea_mouse_56$upregulated, fea_mouse_56$downregulated)
mouse_56_soce <- bubbleplot(soce_56) + labs(title = "Store-operated CA 2+ Entry-related Enriched Terms")
mouse_56_soce
```
Save plots
```{r}
ggsave(filename = here("res", "fea", "mouse_56_overall.pdf"), bp_mouse_56_overall)

```

### FEA p28 cystic '19 Mouse
```{r}
fea_mouse_19 <- mouseenrich(musdegs_19_UP, musdegs_19_DOWN, 40)

bp_mouse_19 <- barplot(fea_mouse_19$topcompared)
bp_mouse_19
```

bubbleplot
```{r}
fea_mouse_19$topcompared$term_name <- stringr::str_wrap(fea_mouse_19$topcompared$term_name, width = 70)


bubbleplot_p28 <- bubbleplot(dplyr::filter(fea_mouse_19$topcompared, intersection_size > 10)) + labs(title = "Top Enriched Terms for Cystic P28 Dataset") + theme(text = element_text(size = 14), plot.title = element_text(size = 24))
```

```{r}
ggsave(bubbleplot_p28, filename = here("res", "fea", "p28_fea_bubbleplot.pdf"), height = 7, width = 14)

```


Cilia-related pathways
```{r eval=FALSE, include=FALSE}
cilia_mouse_19 <- cilia_terms(fea_mouse_19)
cilia_mouse_19_plot <- bubbleplot(cilia_mouse_19) + labs(title = "Cilia-related Enriched Terms")
cilia_mouse_19_plot
```

Mitochondria-related pathways 
```{r eval=FALSE, include=FALSE}
mito_mouse_19 <- mito_terms(fea_mouse_19)
mito_mouse_19_plot <- bubbleplot(mito_mouse_19) + labs(title = "Mitochondria-related Enriched Terms")
mito_mouse_19_plot
```

Fibrosis-related terms
```{r eval=FALSE, include=FALSE}
fibrosis_mouse_19 <- fibrosis_terms(fea_mouse_19)
fibrosis_mouse_19_plot <- bubbleplot(fibrosis_mouse_19) + labs(title = "Fibrosis-related Enriched Terms")
fibrosis_mouse_19_plot

```

Immune-related terms
```{r eval=FALSE, include=FALSE}
immune_mouse_19 <- immune_terms(fea_mouse_19)
immune_mouse_19_plot <- bubbleplot(immune_mouse_19) + labs(title = "Immune-related Enriched Terms")
immune_mouse_19_plot

```

Store-operated CA 2+ Entry pathways
```{r eval=FALSE, include=FALSE}
soce_19 <- soce(fea_mouse_19$upregulated, fea_mouse_19$downregulated)
mouse_19_soce <- bubbleplot(soce_19) + labs(title = "Store-operated CA 2+ Entry-related Enriched Terms")
mouse_19_soce
```

Save gprofiler results
```{r eval=FALSE, include=FALSE}
##fea_mouse_39
fea_mouse_39_upterms <- apply(fea_mouse_39$upregulated, 2, as.character)
write.csv(fea_mouse_39_upterms, file = here("res", "fea", "fea_mouse_39_upterms.csv"))

fea_mouse_39_downterms <- apply(fea_mouse_39$downregulated, 2, as.character)
write.csv(fea_mouse_39_downterms, file = here("res", "fea", "fea_mouse_39_downterms.csv"))

##fea_mouse_56
fea_mouse_56_upterms <- apply(fea_mouse_56$upregulated, 2, as.character)
write.csv(fea_mouse_56_upterms, file = here("res", "fea", "fea_mouse_56_upterms.csv"))

fea_mouse_56_downterms <- apply(fea_mouse_56$downregulated, 2, as.character)
write.csv(fea_mouse_56_downterms, file = here("res", "fea", "fea_mouse_56_downterms.csv"))

##fea_mouse_19
fea_mouse_19_upterms <- apply(fea_mouse_19$upregulated, 2, as.character)
write.csv(fea_mouse_19_upterms, file = here("res", "fea", "fea_mouse_19_upterms.csv"))

fea_mouse_19_downterms <- apply(fea_mouse_19$downregulated, 2, as.character)
write.csv(fea_mouse_19_downterms, file = here("res", "fea", "fea_mouse_19_downterms.csv"))
```

Save plots
```{r}
ggsave(filename = here("res", "fea", "mouse_19_overall.pdf"), bp_mouse_19)

```

#### Cystic Overlap ('19 and '56)  
read in
```{r}
fea_56_up <- read.csv(here("res", "fea", "fea_mouse_56_upterms.csv"))

fea_56_down <- read.csv(here("res", "fea", "fea_mouse_56_downterms.csv"))

fea_19_up <- read.csv(here("res", "fea", "fea_mouse_19_upterms.csv"))

fea_19_down <- read.csv(here("res", "fea", "fea_mouse_19_downterms.csv"))

fea_39_up <- read.csv(here("res", "fea", "fea_mouse_39_upterms.csv"))

fea_39_down <- read.csv(here("res", "fea", "fea_mouse_39_downterms.csv"))
```

merge for overlap
```{r}
fea_cystic_overlap_up <- merge(fea_56_up, fea_19_up, by = "term_name")

fea_cystic_overlap_down <- merge(fea_56_down, fea_19_down, by = "term_name")
```

merge for overlap across all datasets
```{r}
fea_all_overlap_up <- merge(fea_cystic_overlap_up, fea_39_up, by = "term_name")

fea_all_overlap_down <- merge(fea_cystic_overlap_down, fea_39_down, by = "term_name")

```

cystic up overlap
```{r}
ggplot(slice_max(fea_cystic_overlap_up, order_by = recall.y, n = 30), aes(x=intersection_size.y, y=term_name,  fill = p_value.y)) + 
    geom_bar( stat = "identity") + 
    scale_fill_continuous(type = "viridis") + 
    labs(x = "Number of Genes Matched to Term", y = "Functional Terms", title = "Top Enriched Terms") + 
    theme_minimal(base_size = 8)


```
cystic down overlap
```{r}
ggplot(slice_max(fea_cystic_overlap_down, order_by = recall.y, n = 30), aes(x=intersection_size.y, y=term_name,  fill = p_value.y)) + 
    geom_bar( stat = "identity") + 
    scale_fill_continuous(type = "viridis") + 
    labs(x = "Number of Genes Matched to Term", y = "Functional Terms", title = "Top Enriched Terms") + 
    theme_minimal(base_size = 8)
```

```{r}
fea_cystic_overlap_combined <- rbind(fea_cystic_overlap_up, fea_cystic_overlap_down)

fea_cystic_overlap_combined <- fea_cystic_overlap_combined %>% rowwise() %>% mutate(total_genes = sum(intersection_size.x, intersection_size.y))
```

```{r}
ggplot(slice_max(fea_cystic_overlap_combined, order_by = recall.x, n = 10), aes(x=direction.x, y=term_name, size = total_genes, fill = p_value.y)) +
  geom_point(alpha=0.7, shape = 21) +
  scale_size(range = c(2, 10), name = "# Genes Matched to Term") + 
  scale_fill_distiller(palette = "Purples") + 
  labs(x = "Differential Expression Direction", y = "Functional Enrichment Terms") + 
  theme_minimal(base_size = 8)

```

```{r}
write.csv(fea_cystic_overlap_up, file = here("res", "fea", "fea_cystic_overlap_up.csv"))

write.csv(fea_cystic_overlap_down, file = here("res", "fea", "fea_cystic_overlap_down.csv"))

write.csv(fea_cystic_overlap_combined, file = here("res", "fea", "fea_cystic_overlap_combined.csv"))
```

#### Compare Mouse FEA  

Term overlap
```{r}
fea_overlap_19_56_up <- merge(fea_mouse_19$upregulated, fea_mouse_56$upregulated, by = "term_name")
fea_overlap_19_56_down <- merge(fea_mouse_19$downregulated, fea_mouse_56$downregulated, by = "term_name")

#fea_overlap_19_56 <- list(upregulated = fea_overlap_19_56_up, downregulated = fea_overlap_19_56_down)

fea_overap_all_up <- merge(fea_mouse_39$upregulated, fea_overlap_19_56_up, by = "term_name")
fea_overap_all_down <- merge(fea_mouse_39$downregulated, fea_overlap_19_56_down, by = "term_name")
```

Save
```{r}
fea_overlap_19_56_upterms <- apply(fea_overlap_19_56_up, 2, as.character)
write.csv(fea_overlap_19_56_upterms, file = here("res", "fea", "fea_overlap_19_56_upterms.csv"))

fea_overlap_19_56_downterms <- apply(fea_overlap_19_56_down, 2, as.character)
write.csv(fea_overlap_19_56_downterms, file = here("res", "fea", "fea_overlap_19_56_downterms.csv"))

fea_overap_all_upterms <- apply(fea_overap_all_up, 2, as.character)
write.csv(fea_overap_all_upterms, file = here("res", "fea", "fea_overap_all_upterms.csv"))

fea_overap_all_downterms <- apply(fea_overap_all_down, 2, as.character)
write.csv(fea_overap_all_downterms, file = here("res", "fea", "fea_overap_all_downterms.csv"))

```

Mitochondria-related terms overlap
```{r eval=FALSE, include=FALSE}
mitoch_overlap_19_56 <- mito_terms(fea_overlap_19_56)
#mitoch_overlap_19_56_plot <- bubbleplot(mitoch_overlap_19_56) + labs(title = "Mitochondria-related Enriched Terms")
#mitoch_overlap_19_56_plot
dim(mitoch_overlap_19_56)
```



#### Versions  
```{r}
R.Version()
```


```{r}
installed.packages()[names(sessionInfo()$otherPkgs), "Version"]
```
  
