---
title: "R Notebook"
output:
    html_document:
      toc: true
      toc_float: true

---


```{r message=FALSE, warning = FALSE}
library(gprofiler2)
library(viridis)
library(stringr)
library(signatureSearch)
library(ggplot2)
library(dplyr)
library(here)
library(viridis)
library(rrvgo)
```

## Custom functions for FEA

Enrichment function for up and down genes
```{r}
enrich <- function(upgenes, downgenes, topreturned) {#, filepath) {
  up_fea <- gost(upgenes, multi_query = FALSE, correction_method = "bonferroni", evcodes = TRUE, sources = c("GO:BP", "GO:MF", "GO:CC", "REAC", "WP"))
  #remove abitrary pathways -- any pathways with > 1000 genes
    up_fea <- up_fea$result %>% dplyr::filter(., term_size < 1000 & term_size > 5) %>% dplyr::mutate(., .keep = "all", direction = "upregulated")
    up_mat <- up_fea  %>% dplyr::arrange(., desc(p_value)) %>% dplyr::slice_head(., n = topreturned)
  
  down_fea <- gost(downgenes, multi_query = FALSE, correction_method = "bonferroni", evcodes = TRUE, sources = c("GO:BP", "GO:MF", "GO:CC", "REAC", "WP"))
  #remove abitrary pathways -- any pathways with > 1000 genes
  down_fea <- down_fea$result %>% dplyr::filter(., term_size < 1000 & term_size > 5) %>% dplyr::mutate(., .keep = "all", direction = "downregulated") 
  down_mat <- down_fea  %>% dplyr::arrange(., desc(p_value)) %>% dplyr::slice_head(., n = topreturned)
  combined_fea <- rbind(up_mat, down_mat)
  
  fea_res <- list(downregulated = down_fea, upregulated = up_fea, topcompared = combined_fea)

return(fea_res)
}
```

MOUSE GENES Enrichment function for up and down genes  

```{r}
#Runs an ordered query
fea <- function(genelist, topreturned){
  fea <- gost(genelist, organism = "mmusculus", ordered_query = FALSE, multi_query = FALSE, evcodes = TRUE, correction_method = "bonferroni", sources = c("GO:BP", "GO:MF", "GO:CC", "KEGG", "REAC", "MIRNA", "CORUM", "HP", "HPA", "WP"))
  #remove abitrary pathways -- any pathways with > 1000 genes
  fea <- fea$result %>% dplyr::filter(., term_size < 1000 & term_size > 10)
return(fea)
}

mouseenrich <- function(upgenes, downgenes, topreturned) {
  upgenes$log2FoldChange <- sort(as.numeric(upgenes$log2FoldChange), decreasing = TRUE)
  up_fea <- fea(upgenes$Mouse.gene.stable.ID, topreturned$Mouse.gene.stable.ID)
  up_fea <- dplyr::mutate(up_fea, .keep = "all", direction = "upregulated")
  
  downgenes$log2FoldChange <- sort(as.numeric(downgenes$log2FoldChange))
  down_fea <- fea(downgenes$Mouse.gene.stable.ID, topreturned$Mouse.gene.stable.ID)
  down_fea <- dplyr::mutate(down_fea, .keep = "all", direction = "downregulated")
  
  up_mat <- up_fea  %>% dplyr::arrange(., desc(recall)) %>% dplyr::slice_head(., n = topreturned)
  down_mat <- down_fea  %>% dplyr::arrange(., desc(recall)) %>% dplyr::slice_head(., n = topreturned)
  combined_fea <- rbind(up_mat, down_mat)
  fea_res <- list(downregulated = down_fea, upregulated = up_fea, topcompared = combined_fea)

return(fea_res)
}

```

Custom background for FEA (for enrichment using mouse to human to entrez to lincs conversion -- background gene list is all possible measured genes that have gone through those conversions)
```{r}
#Needs to only be possible genes -- all genes measured in salmon gene counts that map to human entrez
#mesgenes <- data@NAMES
#remove version numbers
#mesgenes <- gsub("\\..*", "", mesgenes)

#genebg <- convertMouseGeneList(mesgenes)
#write.csv2(genebg, file = "./data/background_genelist_genestoLINCSgenes.csv")

genebg <- read.csv(file = here("data", "background_genelist_genestoLINCSgenes.csv"))
```


bubbleplot, up and down for x axis and terms on y axis
```{r}
bubbleplot <- function(combined_fea){
  bubbleplot <- ggplot(combined_fea, aes(x=direction, y=term_name, size = intersection_size, fill = p_value)) +
  geom_point(alpha=0.7, shape = 21) +
  scale_size(range = c(2, 10), name = "# Genes Matched to Term") + 
  scale_fill_distiller(palette = "Purples") + 
  labs(x = "Differential Expression Direction", y = "Functional Enrichment Terms") + 
  theme_minimal(base_size = 8) #+ labs(title = "Top Enriched Terms")
#  ggsave(filepath, bubbleplot, width = 10, height = 7)
return(bubbleplot)
}
```

barplot, up and down merged, colored by p-value, terms on y axis and # genes on x axis
```{r}
barplot <- function(combined_fea){
  barplot <- ggplot(combined_fea, aes(x=intersection_size, y=term_name,  fill = recall)) + 
    geom_bar( stat = "identity") + 
    scale_fill_continuous(type = "viridis") + 
    labs(x = "Number of Genes Matched to Term", y = "Functional Terms") + 
    theme_minimal(base_size = 8)
return(barplot)
}

```


 ### Load in (top 100 human) data  
Load in top 100 up and down genes used as signatures for signatureSearch

'39 precystic p70
```{r}
hom_39_UP <- read.csv(here("res", "deseq2_outputs", "p70_signature_UP.csv"))
hom_39_DOWN <- read.csv(here("res", "deseq2_outputs", "p70_signature_DOWN.csv"))
```

'19 cystic p28
```{r}
hom_19_UP <- read.csv(here("res", "deseq2_outputs", "p28_signature_UP.csv"))
hom_19_DOWN <- read.csv(here("res", "deseq2_outputs", "p28_signature_DOWN.csv"))
```

'56 cystic p21
```{r}
hom_56_UP <- read.csv(here("res", "deseq2_outputs", "p21_signature_UP.csv"))
hom_56_DOWN <- read.csv(here("res", "deseq2_outputs", "p21_signature_DOWN.csv"))
```


Comparing Human-Entrez results across datasets 

"Hom" refers to homosapiens -- genes that have been mapped from mouse to human
### Compare number of DEGs - datasets
```{r}
#in common, 19 and 56
overlap_19_56_up <- merge(hom_19_UP, hom_56_UP, by = "entrezgene_id")
dim(overlap_19_56_up)

overlap_19_56_down <- merge(hom_19_DOWN, hom_56_DOWN, by = "entrezgene_id")
dim(overlap_19_56_down)

#in common, 39 and 56
overlap_39_56_up <- merge(hom_39_UP, hom_56_UP, by = "entrezgene_id")
dim(overlap_39_56_up)

overlap_39_56_down <- merge(hom_39_DOWN, hom_56_DOWN, by = "entrezgene_id")
dim(overlap_39_56_down)

#in common, 39 and 19
overlap_39_19_up <- merge(hom_39_UP, hom_19_UP, by = "entrezgene_id")
dim(overlap_39_19_up)

overlap_39_19_down <- merge(hom_39_DOWN, hom_19_DOWN, by = "entrezgene_id")
dim(overlap_39_19_down)

#in common, all
all_up <- merge(overlap_19_56_up, overlap_39_56_up, by = "entrezgene_id")
dim(all_up)

all_down <- merge(overlap_19_56_down, overlap_39_56_down, by = "entrezgene_id")
dim(all_down)
```

specific to pre-cystic
```{r}
p70_unique_up <- dplyr::anti_join(hom_39_UP, hom_19_UP, by = "hgnc_symbol")
p70_unique_up <- dplyr::anti_join(p70_unique_up, hom_56_UP, by = "hgnc_symbol")

p70_unique_down <- dplyr::anti_join(hom_39_DOWN, hom_19_DOWN, by = "hgnc_symbol")
p70_unique_down <- dplyr::anti_join(p70_unique_down, hom_56_DOWN, by = "hgnc_symbol")

```


#### Venn diagram of signatures  
```{r}
library(VennDiagram)
library(RColorBrewer)
myCol <- brewer.pal(3, "Pastel2")
set39 <- rbind(hom_39_UP, hom_39_DOWN)
set56 <- rbind(hom_56_UP, hom_56_DOWN)
set19 <- rbind(hom_19_UP, hom_19_DOWN)

venn.diagram(x = list(set39$ensembl_gene_id, set56$ensembl_gene_id, set19$ensembl_gene_id),
        category.names = c("Precystic P70" , "Cystic P21" , "Cystic P28"),
        filename = here("res/deseq2_outputs", "signatures_venndiagram.png"), # imagetype = "png",
        output=TRUE,
        
        # Output features
        imagetype="png" ,
        height = 480 , 
        width = 570 , 
        resolution = 300,
        compression = "lzw",
        
        # Circles
        lwd = 2,
        lty = 'blank',
        fill = myCol,
        
        # Numbers
        cex = .6,
        fontface = "bold",
        fontfamily = "sans",
        
        # Set names
        cat.cex = 0.55,
        cat.fontface = "bold",
        cat.default.pos = "outer",
        cat.pos = c(-27, 27, 135),
        cat.dist = c(0.055, 0.055, 0.085),
        cat.fontfamily = "sans",
        rotation = 1
)

```


### FEA p70 precystic GSE149739  (top 100 human orthologs)  
```{r}
fea_39 <- enrich(as.character(hom_39_UP$entrezgene_id), as.character(hom_39_DOWN$entrezgene_id), 20)
```

Barplot of overall top enrichment terms
```{r}
bp_39 <- barplot(fea_39$topcompared)
bp_39
```
Bubbleplot
```{r}

bubbleplot(fea_39$topcompared)
```

```{r}
fea_39$topcompared$term_name <- stringr::str_wrap(fea_39$topcompared$term_name, width = 100)

bp_39 <- bubbleplot(fea_39$topcompared) + labs(title = "Top Enriched Terms for Precystic P70 Signature") + theme(text = element_text(size = 12), plot.title = element_text(size = 24))

bp_39
```

```{r}
ggsave(bp_39, filename = here("res", "fea", "bubbleplot_signature_p70.pdf"), height = 7, width = 17)

```


### FEA p21 cystic GSE69556 (top 100 human orthologs)  
```{r}
fea_56 <- enrich(as.character(hom_56_UP$entrezgene_id), as.character(hom_56_DOWN$entrezgene_id), 20)
```

Barplot of overall top enrichment terms
```{r}
bp_56 <- barplot(fea_56$topcompared)
bp_56
```

```{r}

bp_56 <- bubbleplot(fea_56$topcompared) + labs(title = "Top Enriched Terms for Cystic P21 Signature") + theme(text = element_text(size = 14), plot.title = element_text(size = 24))
```

```{r}
ggsave(bp_56, filename = here("res", "fea", "bubbleplot_signature_p21.pdf"), height = 7, width = 14)

```

### FEA p28 cystic GSE134719 (top 100 human orthologs)
```{r}
fea_19 <- enrich(as.character(hom_19_UP$entrezgene_id), as.character(hom_19_DOWN$entrezgene_id), 20)
```

Barplot of overall top enrichment terms
```{r}
bar19 <- barplot(fea_19$topcompared)
bar19
```

```{r}

bubbleplot(fea_19$topcompared)
```


```{r}
fea_19$topcompared$term_name <- stringr::str_wrap(fea_19$topcompared$term_name, width = 100)

bp_19 <- bubbleplot(fea_19$topcompared) + labs(title = "Top Enriched Terms for Cystic P28 Signature") + theme(text = element_text(size = 14), plot.title = element_text(size = 24))

bp_19
```


```{r}
ggsave(bp_19, filename = here("res", "fea", "bubbleplot_signature_p28.pdf"), height = 7, width = 17)

```


Save
``` {r}
##fea_39
fea_39_upterms <- apply(fea_39$upregulated, 2, as.character)
write.csv(fea_39_upterms, file = here("res", "fea", "fea_39_upterms.csv"))

fea_39_downterms <- apply(fea_39$downregulated, 2, as.character)
write.csv(fea_39_downterms, file = here("res", "fea", "fea_39_downterms.csv"))

##fea_56
fea_56_upterms <- apply(fea_56$upregulated, 2, as.character)
write.csv(fea_56_upterms, file = here("res", "fea", "fea_56_upterms.csv"))

fea_56_downterms <- apply(fea_56$downregulated, 2, as.character)
write.csv(fea_56_downterms, file = here("res", "fea", "fea_56_downterms.csv"))

##fea_19
fea_19_upterms <- apply(fea_19$upregulated, 2, as.character)
write.csv(fea_19_upterms, file = here("res", "fea", "fea_19_upterms.csv"))

fea_19_downterms <- apply(fea_19$downregulated, 2, as.character)
write.csv(fea_19_downterms, file = here("res", "fea", "fea_19_downterms.csv"))

```

#### Term Overlaps

Term overlap
```{r}
cystic_term_overlap_up <- merge(fea_19$upregulated, fea_56$upregulated, by = "term_name")
cystic_term_overlap_down <- merge(fea_19$downregulated, fea_56$downregulated, by = "term_name")

p70_p21_term_overlap_up <- merge(fea_39$upregulated, fea_56$upregulated, by = "term_name")
p70_p21_term_overlap_down <- merge(fea_39$downregulated, fea_56$downregulated, by = "term_name")

p70_p28_term_overlap_up <- merge(fea_39$upregulated, fea_19$upregulated, by = "term_name")
p70_p28_term_overlap_down <- merge(fea_19$downregulated, fea_19$downregulated, by = "term_name")

fea_overap_all_up <- merge(fea_39$upregulated, cystic_term_overlap_up, by = "term_name")
fea_overap_all_down <- merge(fea_39$downregulated, cystic_term_overlap_down, by = "term_name")
```

##### rrvgo
Trying out -- give GO terms as input
```{r}
go_39_up <- dplyr::filter(fea_39$upregulated, source == "GO:BP")
go_39_down <- dplyr::filter(fea_39$downregulated, source == "GO:BP")
go_39 <- rbind(go_39_up, go_39_down)

simMatrix <- calculateSimMatrix(go_39$term_id, orgdb = "org.Hs.eg.db", ont = "BP", method = "Wang")

scores <- setNames(-log10(go_39$p_value), go_39$term_id)
reducedTerms <- reduceSimMatrix(simMatrix, scores, threshold=0.70, orgdb="org.Hs.eg.db")

sp <- scatterPlot(simMatrix, reducedTerms, labelSize = 2.5) + labs(title = "Precystic P70 Signature Enrichment Overview") #+ theme(text = element_text(size = 14), plot.title = element_text(size = 24))
sp
tm <- treemapPlot(reducedTerms, title = "Enrichment for Precystic P70 Signature") 
hm <- heatmapPlot(simMatrix, reducedTerms, annotateParent=TRUE, annotationLabel="parentTerm", fontsize=6)
```

```{r}
ggsave()
```

```{r}
go_56_up <- dplyr::filter(fea_56$upregulated, source == "GO:BP")
go_56_down <- dplyr::filter(fea_56$downregulated, source == "GO:BP")
go_56 <- rbind(go_56_up, go_56_down)

simMatrix_56 <- calculateSimMatrix(go_56$term_id, orgdb = "org.Hs.eg.db", ont = "BP", method = "Rel")

scores_56 <- setNames(-log10(go_56$p_value), go_56$term_id)
reducedTerms_56 <- reduceSimMatrix(simMatrix_56, scores_56, threshold=0.70, orgdb="org.Hs.eg.db")

#options(ggrepel.max.overlaps = 50)
sp_56 <- scatterPlot(simMatrix, reducedTerms) + labs(title = "Cystic P21 Signature Enrichment Overview")
sp_56
tm_56 <- treemapPlot(reducedTerms_56, title = "Enrichment for Cystic P21 Signature")
hm_56 <- heatmapPlot(simMatrix_56, reducedTerms_56, annotateParent=TRUE, annotationLabel="parentTerm", fontsize=6)
```


```{r}
go_19_up <- dplyr::filter(fea_19$upregulated, source == "GO:BP")
go_19_down <- dplyr::filter(fea_19$downregulated, source == "GO:BP")
go_19 <- rbind(go_19_up, go_19_down)

simMatrix_19 <- calculateSimMatrix(go_19$term_id, orgdb = "org.Hs.eg.db", ont = "BP", method = "Rel")

scores_19 <- setNames(-log10(go_19$p_value), go_19$term_id)
reducedTerms_19 <- reduceSimMatrix(simMatrix_19, scores_19, threshold=0.7, orgdb="org.Hs.eg.db")

sp_19 <- scatterPlot(simMatrix_19, reducedTerms_19, labelSize = 2.5) + labs(title = "Cystic P28 Signature Enrichment Overview")
sp_19
tm_19 <- treemapPlot(reducedTerms_19, title = "Enrichment for Cystic P28 Signature")
hm_19 <- heatmapPlot(simMatrix_19, reducedTerms_19, annotateParent=TRUE, annotationLabel="parentTerm", fontsize=6)
```



```{r}
ggsave(hm, filename = here("res", "test.pdf"))

```

#### clusterprofiler


```{r}
library(DOSE)
library(enrichplot)
library(clusterProfiler)
library(org.Hs.eg.db)

ego_39 <- enrichGO(gene          = hom_39_UP$entrezgene_id,
                #universe      = names(geneList),
                OrgDb         = org.Hs.eg.db,
                ont           = "BP",
                pAdjustMethod = "BH",
                pvalueCutoff  = 0.01,
                qvalueCutoff  = 0.05,
        readable      = TRUE)

#edo_39 <- gseDO(hom_39_UP$entrezgene_id)
dotplot(ego_39, showCategory=30) + ggtitle("dotplot for ORA")


```


## Mouse genes FEA  
Using ALL mouse DEGs (0.05 alpha and -2 and 2 LFC cutoff) before mapping to human genes or selecting top 100 available in LINCS  

### Load in (mouse DEG) data

'39 nf-core aligned
```{r}
musdegs_39_UP <- read.csv(here("res", "deseq2_outputs", "deseq2_upgenes_res_39.csv"))

musdegs_39_DOWN <- read.csv(here("res", "deseq2_outputs", "deseq2_downgenes_res_39.csv"))
```

'19
```{r}
musdegs_19_UP <- read.csv(here("res", "deseq2_outputs", "deseq2_upgenes_res_19.csv"))

musdegs_19_DOWN <- read.csv(here("res", "deseq2_outputs", "deseq2_downgenes_res_19.csv"))
```

'56
```{r}
musdegs_56_UP <- read.csv(here("res", "deseq2_outputs", "deseq2_upgenes_res_56.csv"))

musdegs_56_DOWN <- read.csv(here("res", "deseq2_outputs", "deseq2_upgenes_res_56.csv"))
```

#### Gene overlaps
Overlap comparing DEGs before human ortholog mapping, using Ensembl ID's
```{r}
degs_19_56_overlap_up <- merge(musdegs_19_UP, musdegs_56_UP, by = "Mouse.gene.stable.ID")

degs_19_56_overlap_down <- merge(musdegs_19_DOWN, musdegs_56_DOWN, by = "Mouse.gene.stable.ID")

degs_19_56_39_overlap_up <- merge(degs_19_56_overlap_up, musdegs_39_UP, by = "Mouse.gene.stable.ID")

degs_19_56_39_overlap_down <- merge(degs_19_56_overlap_down, musdegs_39_DOWN, by = "Mouse.gene.stable.ID")

degs_39_56_overlap_up <- merge(musdegs_39_UP, musdegs_56_UP, by = "Mouse.gene.stable.ID")

degs_39_56_overlap_down <- merge(musdegs_39_DOWN, musdegs_56_DOWN, by = "Mouse.gene.stable.ID")


degs_39_19_overlap_up <- merge(musdegs_39_UP, musdegs_19_UP, by = "Mouse.gene.stable.ID")

degs_39_19_overlap_down <- merge(musdegs_39_DOWN, musdegs_19_DOWN, by = "Mouse.gene.stable.ID")
```

### FEA p70 precystic '39 Mouse
```{r}
fea_mouse_39 <- mouseenrich(musdegs_39_UP, musdegs_39_DOWN, 30)

bp_mouse_39_overall <- barplot(dplyr::filter(fea_mouse_39$topcompared, intersection_size > 10)) + ggtitle("Pre-Cystic Top Upregulated ") 
bp_mouse_39_overall

fea_mouse_39$topcompared$term_name <- stringr::str_wrap(fea_mouse_39$topcompared$term_name, width = 60)

p70_fea_plot <- ggplot(dplyr::filter(fea_mouse_39$topcompared, intersection_size > 10), aes(x=intersection_size, y=term_name,  fill = recall)) + 
    geom_bar( stat = "identity") + 
    scale_fill_continuous(type = "viridis") + 
    labs(x = "Number of Genes Matched to Term", y = "Functional Terms", title = "Pre-Cystic Top Upregulated Enrichment") + 
    theme(text = element_text(size = 14), plot.title = element_text(size = 24))# +
    #theme_minimal(base_size = 8)
p70_fea_plot
```

bubbleplot
```{r}
bubbleplot_p70 <- bubbleplot(dplyr::filter(fea_mouse_39$topcompared, intersection_size > 6)) + labs(title = "Top Enriched Terms for Pre-Cystic P70 Dataset") + theme(text = element_text(size = 14), plot.title = element_text(size = 24))

bubbleplot(fea_mouse_39$topcompared) + labs(title = "Top Enriched Terms for Pre-Cystic P70 Dataset") + theme(text = element_text(size = 14), plot.title = element_text(size = 24))
```

``` {r}
ggsave(p70_fea_plot, filename = here("res", "fea", "p70_fea_plot.pdf"))

ggsave(bubbleplot_p70, filename = here("res", "fea", "p70_fea_bubbleplot.pdf"), height = 7, width = 14)
```

bubbleplot top
```{r}
bubbleplot(dplyr::filter(fea_mouse_39$topcompared, intersection_size > 10)) #+ labs(title = "")
```

Save plots
``` {r}
ggsave(filename = here("res", "fea", "mouse_39_overall.pdf"), bp_mouse_39_overall)

```


test--do FEA res change when first filtering out pseudogenes and ribosomal genes?
```{r}
#filter LFC
filtered_39_up <- dplyr::filter(genes, log2FoldChange > 2)
filtered_39_down <- dplyr::filter(genes, log2FoldChange < -2)

filtered_39_up <- dplyr::filter(filtered_39_up, pvalue < 0.05)
filtered_39_down <- dplyr::filter(filtered_39_down,  pvalue < 0.05)


fea_mouse_39_filtered_up <- fea(filtered_39_up$ensembl_gene_id, 10)
                             
fea_mouse_39_filtered_down <- fea(filtered_39_down$ensembl_gene_id, 10)

test_up <- dplyr::mutate(fea_mouse_39_filtered_up, .keep = "all", direction = "upregulated")
test_down <- dplyr::mutate(fea_mouse_39_filtered_down, .keep = "all", direction = "downregulated")

up_mat <- test_up  %>% dplyr::arrange(., desc(recall)) %>% dplyr::slice_head(., n = 30)
  down_mat <- test_down  %>% dplyr::arrange(., desc(recall)) %>% dplyr::slice_head(., n = 30)
combined_fea <- rbind(up_mat, down_mat)
  


ggplot(combined_fea, aes(x=intersection_size, y=term_name,  fill = recall)) + 
    geom_bar( stat = "identity") + 
    scale_fill_continuous(type = "viridis") + 
    labs(x = "Number of Genes Matched to Term", y = "Functional Terms") + 
    theme_minimal(base_size = 8)


test <- ggplot(dplyr::filter(combined_fea, intersection_size > 7), aes(x=direction, y=term_name, size = intersection_size, fill = p_value)) +
  geom_point(alpha=0.7, shape = 21) +
  scale_size(range = c(2, 10), name = "# Genes Matched to Term") + 
  scale_fill_distiller(palette = "Purples") + 
  labs(x = "Differential Expression Direction", y = "Functional Enrichment Terms") + 
  theme_minimal(base_size = 8)
```

Save plots
``` {r}
ggsave(filename = here("res", "fea", "test.pdf"), test)

```

Number of DEGs with pseudo/ribo genes filtered out, and unique symbols
```{r}
filtered_19_up <- dplyr::filter(genes_19, log2FoldChange > 2)
filtered_19_down <- dplyr::filter(genes_19, log2FoldChange < -2)

filtered_56_up <- dplyr::filter(genes_56, log2FoldChange > 2)
filtered_56_down <- dplyr::filter(genes_56, log2FoldChange < -2)


filtered_19_up <- dplyr::filter(filtered_19_up, pvalue < 0.05)
filtered_19_down <- dplyr::filter(filtered_19_down,  pvalue < 0.05)

filtered_56_up <- dplyr::filter(filtered_56_up, pvalue < 0.05)
filtered_56_down <- dplyr::filter(filtered_56_down,  pvalue < 0.05)


```

### FEA p21 cystic '56 Mouse
```{r}
fea_mouse_56 <- mouseenrich(musdegs_56_UP, musdegs_56_DOWN, 30)

bp_mouse_56_overall <- barplot(dplyr::filter(fea_mouse_56$topcompared, intersection_size > 10))
bp_mouse_56_overall
```

bubbleplot
```{r}
fea_mouse_56$topcompared$term_name <- stringr::str_wrap(fea_mouse_56$topcompared$term_name, width = 70)


bubbleplot_p21 <- bubbleplot(dplyr::filter(fea_mouse_56$topcompared, intersection_size > 10)) + labs(title = "Top Enriched Terms for Cystic P21 Dataset") + theme(text = element_text(size = 14), plot.title = element_text(size = 24))
```

``` {r}
ggsave(bubbleplot_p21, filename = here("res", "fea", "p21_fea_bubbleplot.pdf"), height = 7, width = 14)

```

Save plots
``` {r}
ggsave(filename = here("res", "fea", "mouse_56_overall.pdf"), bp_mouse_56_overall)

```

### FEA p28 cystic '19 Mouse
```{r}
fea_mouse_19 <- mouseenrich(musdegs_19_UP, musdegs_19_DOWN, 40)

bp_mouse_19 <- barplot(fea_mouse_19$topcompared)
bp_mouse_19
```

bubbleplot
```{r}
fea_mouse_19$topcompared$term_name <- stringr::str_wrap(fea_mouse_19$topcompared$term_name, width = 70)


bubbleplot_p28 <- bubbleplot(dplyr::filter(fea_mouse_19$topcompared, intersection_size > 10)) + labs(title = "Top Enriched Terms for Cystic P28 Dataset") + theme(text = element_text(size = 14), plot.title = element_text(size = 24))
```

``` {r}
ggsave(bubbleplot_p28, filename = here("res", "fea", "p28_fea_bubbleplot.pdf"), height = 7, width = 14)

```


Save gprofiler results
``` {r}
##fea_mouse_39
fea_mouse_39_upterms <- apply(fea_mouse_39$upregulated, 2, as.character)
write.csv(fea_mouse_39_upterms, file = here("res", "fea", "fea_mouse_39_upterms.csv"))

fea_mouse_39_downterms <- apply(fea_mouse_39$downregulated, 2, as.character)
write.csv(fea_mouse_39_downterms, file = here("res", "fea", "fea_mouse_39_downterms.csv"))

##fea_mouse_56
fea_mouse_56_upterms <- apply(fea_mouse_56$upregulated, 2, as.character)
write.csv(fea_mouse_56_upterms, file = here("res", "fea", "fea_mouse_56_upterms.csv"))

fea_mouse_56_downterms <- apply(fea_mouse_56$downregulated, 2, as.character)
write.csv(fea_mouse_56_downterms, file = here("res", "fea", "fea_mouse_56_downterms.csv"))

##fea_mouse_19
fea_mouse_19_upterms <- apply(fea_mouse_19$upregulated, 2, as.character)
write.csv(fea_mouse_19_upterms, file = here("res", "fea", "fea_mouse_19_upterms.csv"))

fea_mouse_19_downterms <- apply(fea_mouse_19$downregulated, 2, as.character)
write.csv(fea_mouse_19_downterms, file = here("res", "fea", "fea_mouse_19_downterms.csv"))
```

Save plots
``` {r}
ggsave(filename = here("res", "fea", "mouse_19_overall.pdf"), bp_mouse_19)

```


#### Cystic Overlap ('19 and '56)  
read in
```{r}
fea_56_up <- read.csv(here("res", "fea", "fea_mouse_56_upterms.csv"))

fea_56_down <- read.csv(here("res", "fea", "fea_mouse_56_downterms.csv"))

fea_19_up <- read.csv(here("res", "fea", "fea_mouse_19_upterms.csv"))

fea_19_down <- read.csv(here("res", "fea", "fea_mouse_19_downterms.csv"))

fea_39_up <- read.csv(here("res", "fea", "fea_mouse_39_upterms.csv"))

fea_39_down <- read.csv(here("res", "fea", "fea_mouse_39_downterms.csv"))
```

merge for overlap
```{r}
fea_cystic_overlap_up <- merge(fea_56_up, fea_19_up, by = "term_name")

fea_cystic_overlap_down <- merge(fea_56_down, fea_19_down, by = "term_name")
```

merge for overlap across all datasets
```{r}
fea_all_overlap_up <- merge(fea_cystic_overlap_up, fea_39_up, by = "term_name")

fea_all_overlap_down <- merge(fea_cystic_overlap_down, fea_39_down, by = "term_name")

```

cystic up overlap
```{r}
ggplot(slice_max(fea_cystic_overlap_up, order_by = recall.y, n = 30), aes(x=intersection_size.y, y=term_name,  fill = p_value.y)) + 
    geom_bar( stat = "identity") + 
    scale_fill_continuous(type = "viridis") + 
    labs(x = "Number of Genes Matched to Term", y = "Functional Terms", title = "Top Enriched Terms") + 
    theme_minimal(base_size = 8)


```

cystic down overlap
```{r}
ggplot(slice_max(fea_cystic_overlap_down, order_by = recall.y, n = 30), aes(x=intersection_size.y, y=term_name,  fill = p_value.y)) + 
    geom_bar( stat = "identity") + 
    scale_fill_continuous(type = "viridis") + 
    labs(x = "Number of Genes Matched to Term", y = "Functional Terms", title = "Top Enriched Terms") + 
    theme_minimal(base_size = 8)
```

```{r}
fea_cystic_overlap_combined <- rbind(fea_cystic_overlap_up, fea_cystic_overlap_down)

fea_cystic_overlap_combined <- fea_cystic_overlap_combined %>% rowwise() %>% mutate(total_genes = sum(intersection_size.x, intersection_size.y))
```

```{r}
ggplot(slice_max(fea_cystic_overlap_combined, order_by = recall.x, n = 10), aes(x=direction.x, y=term_name, size = total_genes, fill = p_value.y)) +
  geom_point(alpha=0.7, shape = 21) +
  scale_size(range = c(2, 10), name = "# Genes Matched to Term") + 
  scale_fill_distiller(palette = "Purples") + 
  labs(x = "Differential Expression Direction", y = "Functional Enrichment Terms") + 
  theme_minimal(base_size = 8)

```

``` {r}
write.csv(fea_cystic_overlap_up, file = here("res", "fea", "fea_cystic_overlap_up.csv"))

write.csv(fea_cystic_overlap_down, file = here("res", "fea", "fea_cystic_overlap_down.csv"))

write.csv(fea_cystic_overlap_combined, file = here("res", "fea", "fea_cystic_overlap_combined.csv"))
```

##### rrvgo
Trying out -- give GO terms as input
```{r}

go_39 <- dplyr::filter(fea_39_up, source == "GO:BP")

simMatrix <- calculateSimMatrix(go_39$term_id, orgdb = "org.Mm.eg.db", ont = "BP", method = "Wang")

scores <- setNames(-log10(go_39$p_value), go_39$term_id)
reducedTerms <- reduceSimMatrix(simMatrix, scores, threshold=0.7, orgdb="org.Mm.eg.db")

sp <- scatterPlot(simMatrix, reducedTerms)
tm <- treemapPlot(reducedTerms)
hm <- heatmapPlot(simMatrix, reducedTerms, annotateParent=TRUE, annotationLabel="parentTerm", fontsize=6)
```


```{r}
go_19 <- dplyr::filter(fea_19_up, source == "GO:BP")

simMatrix <- calculateSimMatrix(go_19$term_id, orgdb = "org.Mm.eg.db", ont = "BP", method = "Rel")

scores <- setNames(-log10(go_19$p_value), go_19$term_id)
reducedTerms <- reduceSimMatrix(simMatrix, scores, threshold=0.4, orgdb="org.Mm.eg.db")

sp <- scatterPlot(simMatrix, reducedTerms)
sp
tm <- treemapPlot(reducedTerms)
hm <- heatmapPlot(simMatrix, reducedTerms, annotateParent=TRUE, annotationLabel="parentTerm", fontsize=6)
```


```{r}
ggsave(hm, filename = here("res", "test.pdf"))

```

#### Compare Mouse FEA  

Term overlap
```{r}
fea_overlap_19_56_up <- merge(fea_mouse_19$upregulated, fea_mouse_56$upregulated, by = "term_name")
fea_overlap_19_56_down <- merge(fea_mouse_19$downregulated, fea_mouse_56$downregulated, by = "term_name")

#fea_overlap_19_56 <- list(upregulated = fea_overlap_19_56_up, downregulated = fea_overlap_19_56_down)

fea_overap_all_up <- merge(fea_mouse_39$upregulated, fea_overlap_19_56_up, by = "term_name")
fea_overap_all_down <- merge(fea_mouse_39$downregulated, fea_overlap_19_56_down, by = "term_name")
```

Save
``` {r}
fea_overlap_19_56_upterms <- apply(fea_overlap_19_56_up, 2, as.character)
write.csv(fea_overlap_19_56_upterms, file = here("res", "fea", "fea_overlap_19_56_upterms.csv"))

fea_overlap_19_56_downterms <- apply(fea_overlap_19_56_down, 2, as.character)
write.csv(fea_overlap_19_56_downterms, file = here("res", "fea", "fea_overlap_19_56_downterms.csv"))

fea_overap_all_upterms <- apply(fea_overap_all_up, 2, as.character)
write.csv(fea_overap_all_upterms, file = here("res", "fea", "fea_overap_all_upterms.csv"))

fea_overap_all_downterms <- apply(fea_overap_all_down, 2, as.character)
write.csv(fea_overap_all_downterms, file = here("res", "fea", "fea_overap_all_downterms.csv"))

```



#### Versions  
```{r}
R.Version()
```


```{r}
installed.packages()[names(sessionInfo()$otherPkgs), "Version"]
```
  
