---
title: "R Notebook"
output:
    html_document:
      toc: true
      toc_float: true

---


### Purpose  
This markdown follows differential expression analyses from __deseq2_analyses.Rmd__. DEGs and signatures from pre-cystic P70 ('39') data and cystic P21 ('56') and P28 ('19') data are compared to each other and used for functional enirichment analysis   
__Data Sets__
- '39 / Pre-Cystic P70: [GSE149739](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE149739)  
- '19 / Cystic P21: [GSE134719](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE134719)  
- '56 / Cystic P28: [GSE69556](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE69556)  


```{r message=FALSE, warning = FALSE}
library(gprofiler2)
library(viridis)
library(stringr)
library(signatureSearch)
library(ggplot2)
library(dplyr)
library(here)
library(viridis)
library(rrvgo)
library(GOSemSim)
library(VennDiagram)
library(RColorBrewer)
library(ComplexHeatmap)
library(ExperimentHub)
source(here("code", "functions.R"))
```

##Signature comparisons

#### Load in (top 100 human) data  
Load in top 100 up and down genes used as signatures for signatureSearch

'39 precystic p70
```{r}
hom_39_UP <- read.csv(here("res", "deseq2_outputs", "p70_signature_UP.csv"), row.names = 1)
hom_39_DOWN <- read.csv(here("res", "deseq2_outputs", "p70_signature_DOWN.csv"), row.names = 1)
```

'19 cystic p28
```{r}
hom_19_UP <- read.csv(here("res", "deseq2_outputs", "p28_signature_UP.csv"), row.names = 1)
hom_19_DOWN <- read.csv(here("res", "deseq2_outputs", "p28_signature_DOWN.csv"), row.names = 1)
```

'56 cystic p21
```{r}
hom_56_UP <- read.csv(here("res", "deseq2_outputs", "p21_signature_UP.csv"), row.names = 1)
hom_56_DOWN <- read.csv(here("res", "deseq2_outputs", "p21_signature_DOWN.csv"), row.names = 1)
```

specific to pre-cystic
```{r}
p70_unique_up <- dplyr::anti_join(hom_39_UP, hom_19_UP, by = "hgnc_symbol")
p70_unique_up <- dplyr::anti_join(p70_unique_up, hom_56_UP, by = "hgnc_symbol")

p70_unique_down <- dplyr::anti_join(hom_39_DOWN, hom_19_DOWN, by = "hgnc_symbol")
p70_unique_down <- dplyr::anti_join(p70_unique_down, hom_56_DOWN, by = "hgnc_symbol")

```


### Venn diagram of signatures  
```{r}
set39 <- c(hom_39_UP$entrezgene_id, hom_39_DOWN$entrezgene_id)
set56 <- c(hom_56_UP$entrezgene_id, hom_56_DOWN$entrezgene_id)
set19 <- c(hom_19_UP$entrezgene_id, hom_19_DOWN$entrezgene_id)
  
venn3(set39, set56, set19, categories = c("Precystic P70" , "Cystic P21" , "Cystic P28"), filename = here("res/deseq2_outputs", "signatures_venndiagram.png"))
  

```


### FEA p70 precystic GSE149739  (top 100 human orthologs)  
```{r}
fea_39 <- enrich(as.character(hom_39_UP$entrezgene_id), as.character(hom_39_DOWN$entrezgene_id), 11) 
```

Bubbleplot
```{r}
bubbleplot(fea_39$topcompared)
```

```{r}
fea_39$topcompared$term_name <- stringr::str_wrap(fea_39$topcompared$term_name, width = 70)

bp_39 <- bubbleplot(fea_39$topcompared) + labs(title = "Top Enriched Terms for \nPrecystic P70 Signature") + scale_x_discrete(labels=c("downregulated" = "down", "upregulated" = "up")) + theme(text = element_text(size = 20, face = "bold"), plot.title = element_text(size = 20))

bp_39
```

```{r}
ggsave(bp_39, filename = here("res", "fea", "bubbleplot_signature_p70.png"), height = 8
, width = 8.75)

```


### FEA p21 cystic GSE69556 (top 100 human orthologs)  
```{r}
fea_56 <- enrich(as.character(hom_56_UP$entrezgene_id), as.character(hom_56_DOWN$entrezgene_id), 9)
```


```{r}
fea_56$topcompared$term_name <- stringr::str_trunc(fea_56$topcompared$term_name, width = 55)

bp_56 <- bubbleplot(fea_56$topcompared) + labs(title = "Top Enriched Terms for \nCystic P21 Signature") + scale_x_discrete(labels=c("downregulated" = "down", "upregulated" = "up")) + theme(text = element_text(size = 20, face = "bold"), plot.title = element_text(size = 20))

bp_56
```

```{r}
ggsave(bp_56, filename = here("res", "fea", "bubbleplot_signature_p21.png"), height = 9, width = 9.76)

```

### FEA p28 cystic GSE134719 (top 100 human orthologs)
```{r}
fea_19 <- enrich(as.character(hom_19_UP$entrezgene_id), as.character(hom_19_DOWN$entrezgene_id), 11)
```


```{r}

bubbleplot(fea_19$topcompared)
```


```{r}
fea_19$topcompared$term_name <- stringr::str_trunc(fea_19$topcompared$term_name, width = 42)

bp_19 <- bubbleplot(fea_19$topcompared) + labs(title = "Top Enriched Terms for \nCystic P28 Signature") + scale_x_discrete(labels=c("downregulated" = "down", "upregulated" = "up")) + theme(text = element_text(size = 20, face = "bold"), plot.title = element_text(size = 20))

bp_19
```

```{r}
ggsave(bp_19, filename = here("res", "fea", "bubbleplot_signature_p28.png"), height = 8.7, width = 8.75)

```

Save
``` {r}
##fea_39
fea_39_upterms <- apply(fea_39$upregulated, 2, as.character)
fea_39_downterms <- apply(fea_39$downregulated, 2, as.character)
write.csv(rbind(fea_39_upterms, fea_39_downterms), file = here("res", "fea", "fea_p70sig.csv"))

##fea_56
fea_56_upterms <- apply(fea_56$upregulated, 2, as.character)
fea_56_downterms <- apply(fea_56$downregulated, 2, as.character)
write.csv(rbind(fea_56_upterms, fea_56_downterms), file = here("res", "fea", "fea_p21sig.csv"))

##fea_19
fea_19_upterms <- apply(fea_19$upregulated, 2, as.character)
fea_19_downterms <- apply(fea_19$downregulated, 2, as.character)
write.csv(rbind(fea_19_upterms, fea_19_downterms), file = here("res", "fea", "fea_p28sig.csv"))

```

#### Term Overlaps
Term overlap
```{r}
cystic_term_overlap_up <- merge(fea_19$upregulated, fea_56$upregulated, by = "term_name")
cystic_term_overlap_down <- merge(fea_19$downregulated, fea_56$downregulated, by = "term_name")

p70_p21_term_overlap_up <- merge(fea_39$upregulated, fea_56$upregulated, by = "term_name")
p70_p21_term_overlap_down <- merge(fea_39$downregulated, fea_56$downregulated, by = "term_name")

p70_p28_term_overlap_up <- merge(fea_39$upregulated, fea_19$upregulated, by = "term_name")
p70_p28_term_overlap_down <- merge(fea_19$downregulated, fea_19$downregulated, by = "term_name")

fea_overap_all_up <- merge(fea_39$upregulated, cystic_term_overlap_up, by = "term_name")
fea_overap_all_down <- merge(fea_39$downregulated, cystic_term_overlap_down, by = "term_name")
```

#### Pathway Similarity
rrvgo, uses GO terms as input
##### P70
```{r}
sim_39 <-   go_bp_sim(fea_39, "org.Hs.eg.db", threshold = 0.80)

sp <- scatterPlot2(sim_39$simMatrix, sim_39$reducedTerms, labelSize = 2.5) + labs(title = "Precystic P70 Signature Enrichment Overview") #+ geom_label_repel(max.overlaps = 20) #+ theme(text = element_text(size = 14), plot.title = element_text(size = 24))
sp
tm <- treemapPlot(sim_39$reducedTerms, title = "Enrichment for Precystic P70 Signature") 
library(treemap)
tm2 <- treemapPlot2(sim_39$reducedTerms, title = "Enrichment for Precystic P70 Signature") 

hm <- heatmapPlot(sim_39$simMatrix, sim_39$reducedTerms, annotateParent=TRUE, annotationLabel="parentTerm", fontsize=6)
```

```{r}
#ggsave(here("res/fea", "precysticp70_sig_enrichment_scatter.pdf"), sp, height = 5, width = 7)

#ggsave(here("res/fea", "precysticp70_sig_enrichment_treemap.pdf"), tm, height = 4, width = 7)

```

##### P21
```{r}
sim_56 <-   go_bp_sim(fea_56, "org.Hs.eg.db", threshold = 0.80)

#options(ggrepel.max.overlaps = 50)
sp_56 <- scatterPlot2(sim_56$simMatrix, sim_56$reducedTerms, labelSize = 2.5) + labs(title = "Cystic P21 Signature Enrichment Overview")
sp_56
tm_56 <- treemapPlot2(sim_56$reducedTerms, title = "Enrichment for Cystic P21 Signature")
hm_56 <- heatmapPlot(sim_56$simMatrix, sim_56$reducedTerms, annotateParent=TRUE, annotationLabel="parentTerm", fontsize=6)
```


```{r}
ggsave(here("res/fea", "cysticp21_sig_enrichment_scatter.pdf"), sp_56, height =5, width = 7)

#ggsave(here("res/fea", "precysticp70_sig_enrichment_treemap.pdf"), tm, height = 4, width = 7)

```

##### P28
```{r}
sim_19 <-   go_bp_sim(fea_19, "org.Hs.eg.db", threshold = 0.80)

#options(ggrepel.max.overlaps = 50)
sp_19 <- scatterPlot2(sim_19$simMatrix, sim_19$reducedTerms, labelSize = 2.5) + labs(title = "Cystic P21 Signature Enrichment Overview")
sp_19
tm_19 <- treemapPlot2(sim_19$reducedTerms, title = "Enrichment for Cystic P28 Signature")
hm_19 <- heatmapPlot(sim_19$simMatrix, sim_19$reducedTerms, annotateParent=TRUE, annotationLabel="parentTerm", fontsize=6)
```

```{r}
ggsave(here("res/fea", "cysticp28_sig_enrichment_scatter.pdf"), sp_19, height =5, width = 7)
```

## DEG Comparisons  
Using ALL mouse DEGs (0.05 alpha and -2 and 2 LFC cutoff) before mapping to human genes or selecting top 100 available in LINCS  

#### AKI Genes  
AKI-associated DEGs were pulled in to compare to the data sets to see if the cystic data sets look more similar to AKI. 157 upregulated and 88 downregulated
```{r}
aki_upgenes <- c("Adam8", "Adamts1", "Adm", "Akap12", "Akr1b8", "Aldh1a2", "Aloxe3", "Anxa1", "Anxa3", "Apobec3", "Arf2", "Asns", "Atf3", "Bcl3", "Birc3", "Cd14", "Cd44", "Cd68", "Cd9", "Cd93", "Cebpd", "Chrnb1", "Clcf1", "Cldn4", "Cldn7", "Ctgf", "Ctsc", "Cyr61", "Ddx21", "Dusp10", "Dusp5", "Edn1", "Efhd2", "Elf4", "Emp1", "Entpd1", "Epha2", "F2r", "F3", "Fam57a", "Flnc", "Fndc4", "Fosl1", "Fosl2", "Fut2", "Fxyd5", "Gas7", "Gdf15", "Gprc5a", "Havcr1", "Hbegf", "Hilpda", "Hmox1", "Hpcal4", "Igf2bp2", "Il1f6", "Il34", "Irak3", "Itga5", "Klf4", "Klf6", "Klf7", "Krt19", "Krt20", "Lamc2", "Lif", "Lrp8", "Lrrc32", "Lrrc8c", "Maff", "Map3k6", "Map4k4", "Mapk6", "Mmp3", "Mthfd1l", "Myc", "Myh9", "Nek6", "Ngf", "Nras", "Nt5c1a", "Nucb2", "Oasl1", "Pdgfb", "Pdk4", "Pdlim7", "Plat", "Plaur", "Plin2", "Plk3", "Plp2", "Ppm1j", "Ppp1r14b", "Pprc1", "Procr", "Prrg4", "Ptger4", "Ptgs2", "Ptpn12", "Pvr", "Pxdc1", "Qsox1", "Rab31", "Rad18", "Rap2b", "Rbm3", "Rell1", "Rhbdf2", "Rin1", "Rnd1", "Rnd3", "Rras2", "Rtn4", "Runx1", "Samd4", "Sbno2", "Sema7a", "Serpinb1a", "Serpine1", "Sfn", "Slc16a1", "Slc25a24", "Slc38a2", "Slc7a5", "Slc7a6", "Smad1", "Smox", "Socs3", "Sox9", "Sphk1", "Sprr1a", "Sprr2f", "Sprr2g", "Spry2", "Spsb1", "St3gal1", "Stil", "Syt12", "Taf1d", "Taf4b", "Tcerg1", "Tgfbr1", "Thbd", "Timp1", "Tinagl1", "Tmcc3", "Tmed5", "Tmem173", "Tnc", "Tnfrsf12a", "Tnfrsf1a", "Tpm3", "Trmt61a", "Tubb6", "Txnrd1", "Uchl1", "Vopp1")

aki_downgenes <- c("1700040L02Rik", "Abhd14b", "Acad12", "Acmsd", "Acot11", "Acox2", "Acss2", "Ak4", "Akr1c14", "Akr1c18", "Aldh4a1", "Apeh", "Atp6v1b1", "BC067074", "Bcat1", "Bdh1", "Bhmt2", "Bphl", "Bsnd", "Casr", "Cbs", "Cdkl1", "Ceacam2", "Cmah", "Crot", "Cth", "Cubn", "Cyp2j11", "D3Ertd751e", "Dhdh", "Dnajc28", "Fam107a", "Fggy", "Fmo5", "Fras1", "G6pc", "Galm", "Gatb", "Gatm", "Glyctk", "Gm10804", "Hgd", "Hnmt", "Hsd3b2", "Hykk", "Ift122", "Inpp5j", "Kcnj1", "Keg1", "Klhl3", "Kmo", "Lrrc31", "Lyplal1", "Map2k6", "Mccc1", "Mccc2", "Mep1b", "Mgam", "Nccrp1", "Oxgr1", "Pbld1", "Pde4c", "Pfkm", "Pnkd", "Pter", "Ranbp3l", "Rhcg", "Serpina1d", "Serpina1f", "Sfrp1", "Shmt2", "Slc15a2", "Slc16a9", "Slc22a13", "Slc22a19", "Slc22a26", "Slc23a1", "Slc25a21", "Slc8a1", "Slco4c1", "Smarca2", "Snx29", "Suox", "Tln2", "Tmem207", "Tmem25", "Tpmt", "Ugt8a")
```

combine all data sets
```{r}
all_39 <- read.csv(here("res/deseq2_outputs", "deseq2_p70_fullannotatedres.csv"))
all_56 <- read.csv(here("res/deseq2_outputs", "deseq2_p21_fullannotatedres.csv"))
all_19 <- read.csv(here("res/deseq2_outputs", "deseq2_p28_fullannotatedres.csv"))
#at least some lfc and padj > 0.1 
all_39 <- dplyr::filter(all_39, log2FoldChange > 0.15 | log2FoldChange < -0.15 & padj < 0.1)
all_56 <- dplyr::filter(all_56, log2FoldChange > 0.15 | log2FoldChange < -0.15 & padj < 0.1)
all_19 <- dplyr::filter(all_19, log2FoldChange > 0.15 | log2FoldChange < -0.15 & padj < 0.1)

combined_fc <- merge(all_56, all_19, by = "X")
combined_fc <- merge(all_39, combined_fc, by = "X")

lfc_aki_up <- combined_fc %>% dplyr::filter( X %in% aki_upgenes) %>% dplyr::mutate(direction = "aki_up") %>% dplyr::select(Genes = X, PrecysticP70 = log2FoldChange, CysticP21 = log2FoldChange.x, CysticP28 = log2FoldChange.y, Direction = direction)#152

lfc_aki_down <- combined_fc %>% dplyr::filter( X %in% aki_downgenes) %>% dplyr::mutate(direction = "aki_down") %>% dplyr::select(Genes = X, PrecysticP70 = log2FoldChange, CysticP21 = log2FoldChange.x, CysticP28 = log2FoldChange.y, Direction = direction)

aki_res <- rbind(lfc_aki_up, lfc_aki_down)
```

##### AKI Heatmap
```{r}
##prepare heatmap annotations
aki_direction <- dplyr::select(aki_res, Direction)
aki_direction$Direction <- factor(aki_direction$Direction, levels = c("aki_up", "aki_down"))

aki_res2 <- tibble::column_to_rownames(aki_res, var = "Genes") %>% dplyr::select(PrecysticP70, CysticP21, CysticP28)

samples <- str_sub(colnames(aki_res2), start = 1, end = -4) %>% as.data.frame(row.names = colnames(aki_res2))
colnames(samples) <- "Cystic_Status"
samples$Cystic_Status <- factor(samples$Cystic_Status, levels = c("Precystic", "Cystic"))
#show_col(viridis(8, option = "D")) #selecting colors

my_color = list(Direction = c(aki_down = "#575C6DFF", aki_up = "#FFEA46FF"), Cystic_Status = c(Precystic = "#00204DFF", Cystic = "#A69D75FF"))
```

plot
```{r}
pdf(here("res/deseq2_outputs", "heatmap_aki_genes.png"), width = 7, height = 13)
#aki_heatmap <- 
  pheatmap(aki_res2, annotation_colors = my_color, annotation_row = aki_direction, annotation_col = samples, cluster_rows = TRUE, cluster_cols = TRUE, show_rownames = TRUE, color = viridis(n= 10000, alpha = 1, begin = 0, end = 1, option = "viridis"), border_color = "NA", main = "Expression of AKI Genes Across Precystic/Cystic Data", angle_col = "0")
dev.off()
```

```{r }
write.csv(aki_res, file = here("res/deseq2_outputs", "aki_gene_lfc.csv"))

```


### Load in (mouse DEG) data

'39 
```{r}
musdegs_39_UP <- read.csv(here("res/deseq2_outputs", "deseq2_upgenes_res_39.csv"))
musdegs_39_DOWN <- read.csv(here("res/deseq2_outputs", "deseq2_downgenes_res_39.csv") )
```

'19
```{r}
musdegs_19_UP <- read.csv(here("res/deseq2_outputs", "deseq2_upgenes_res_19.csv") )
musdegs_19_DOWN <- read.csv(here("res/deseq2_outputs", "deseq2_downgenes_res_19.csv") )
```

'56
```{r}
musdegs_56_UP <- read.csv(here("res/deseq2_outputs", "deseq2_upgenes_res_56.csv") )
musdegs_56_DOWN <- read.csv(here("res/deseq2_outputs", "deseq2_downgenes_res_56.csv") )
```


### Venn diagram of all DEGs  
```{r}
degset39 <- c(musdegs_39_UP$Mouse.gene.stable.ID, musdegs_39_DOWN$Mouse.gene.stable.ID)
degset56 <- c(musdegs_56_UP$Mouse.gene.stable.ID, musdegs_56_DOWN$Mouse.gene.stable.ID)
degset19 <- c(musdegs_19_UP$Mouse.gene.stable.ID, musdegs_19_DOWN$Mouse.gene.stable.ID)
  
venn3(degset39, degset56, degset19, categories = c("Precystic P70" , "Cystic P21" , "Cystic P28"), filename = here("res/deseq2_outputs", "degs_venndiagram.png"))
```

### FEA p70 precystic '39 Mouse
```{r}
fea_mouse_39 <- mouseenrich(musdegs_39_UP$Mouse.gene.stable.ID, musdegs_39_DOWN$Mouse.gene.stable.ID, 30)

bp_mouse_39_overall <- barplot(fea_mouse_39$topcompared) + ggtitle("Pre-Cystic Top Upregulated ") 
bp_mouse_39_overall

mouse_39_bubble <- bubbleplot(fea_mouse_39$topcompared)
mouse_39_bubble

```

bubbleplot
```{r}
bubbleplot_p70 <- bubbleplot(dplyr::filter(fea_mouse_39$topcompared, intersection_size > 6)) + labs(title = "Top Enriched Terms for Pre-Cystic P70 Dataset") + theme(text = element_text(size = 14), plot.title = element_text(size = 24))

bubbleplot(fea_mouse_39$topcompared) + labs(title = "Top Enriched Terms for Pre-Cystic P70 Dataset") + theme(text = element_text(size = 14), plot.title = element_text(size = 24))
```

``` {r}
ggsave(bubbleplot_p70, filename = here("res", "fea", "p70_fea_bubbleplot.pdf"), height = 7, width = 14)
```


### FEA p21 cystic '56 Mouse
```{r}
fea_mouse_56 <- mouseenrich(musdegs_56_UP$Mouse.gene.stable.ID, musdegs_56_DOWN$Mouse.gene.stable.ID, 30)

```

bubbleplot
```{r}
fea_mouse_56$topcompared$term_name <- stringr::str_wrap(fea_mouse_56$topcompared$term_name, width = 70)


bubbleplot_p21 <- bubbleplot(fea_mouse_56$topcompared) + labs(title = "Top Enriched Terms for Cystic P21 Dataset") + theme(text = element_text(size = 14), plot.title = element_text(size = 24))
bubbleplot_p21
```

``` {r}
ggsave(bubbleplot_p21, filename = here("res", "fea", "p21_fea_bubbleplot.pdf"), height = 7, width = 14)

```


### FEA p28 cystic '19 Mouse
```{r}
fea_mouse_19 <- mouseenrich(musdegs_19_UP$Mouse.gene.stable.ID, musdegs_19_DOWN$Mouse.gene.stable.ID, 30)
```

bubbleplot
```{r}
fea_mouse_19$topcompared$term_name <- stringr::str_wrap(fea_mouse_19$topcompared$term_name, width = 70)


bubbleplot_p28 <- bubbleplot(fea_mouse_19$topcompared) + labs(title = "Top Enriched Terms for Cystic P28 Dataset") + theme(text = element_text(size = 14), plot.title = element_text(size = 24))
bubbleplot_p28
```

``` {r}
ggsave(bubbleplot_p28, filename = here("res", "fea", "p28_fea_bubbleplot.pdf"), height = 7, width = 14)

```


Save gprofiler results
``` {r}
#fea_mouse_39
fea_mouse_39_upterms <- apply(fea_mouse_39$upregulated, 2, as.character)
fea_mouse_39_downterms <- apply(fea_mouse_39$downregulated, 2, as.character)
write.csv(rbind(fea_mouse_39_upterms, fea_mouse_39_downterms), file = here("res", "fea", "fea_p70degs.csv"))

#feamouse_56
fea_mouse_56_upterms <- apply(fea_mouse_56$upregulated, 2, as.character)
fea_mouse_56_downterms <- apply(fea_mouse_56$downregulated, 2, as.character)
write.csv(rbind(fea_mouse_56_upterms, fea_mouse_56_downterms), file = here("res", "fea", "fea_p21degs.csv"))

#fea_mouse_19
fea_mouse_19_upterms <- apply(fea_mouse_19$upregulated, 2, as.character)
fea_mouse_19_downterms <- apply(fea_mouse_19$downregulated, 2, as.character)
write.csv(rbind(fea_mouse_19_upterms, fea_mouse_19_downterms), file = here("res", "fea", "fea_p28degs.csv"))
```

### Gene overlaps
Overlap comparing DEGs before human ortholog mapping, using Ensembl ID's
```{r}
degs_19_56_overlap_up <- merge(musdegs_19_UP, musdegs_56_UP, by = "Mouse.gene.stable.ID")
degs_19_56_overlap_down <- merge(musdegs_19_DOWN, musdegs_56_DOWN, by = "Mouse.gene.stable.ID")
degs_39_56_overlap_up <- merge(musdegs_39_UP, musdegs_56_UP, by = "Mouse.gene.stable.ID")
degs_39_56_overlap_down <- merge(musdegs_39_DOWN, musdegs_56_DOWN, by = "Mouse.gene.stable.ID")
degs_39_19_overlap_up <- merge(musdegs_39_UP, musdegs_19_UP, by = "Mouse.gene.stable.ID")
degs_39_19_overlap_down <- merge(musdegs_39_DOWN, musdegs_19_DOWN, by = "Mouse.gene.stable.ID")
degs_all_overlap_up <- merge(degs_19_56_overlap_up, musdegs_39_UP, by = "Mouse.gene.stable.ID")
degs_all_overlap_down <- merge(degs_19_56_overlap_down, musdegs_39_DOWN, by = "Mouse.gene.stable.ID")
alldegs_overlap <- rbind(degs_all_overlap_up, degs_all_overlap_down)
alldegs_lfc <- cbind(CysticP28 = alldegs_overlap$log2FoldChange.x, CysticP21 = alldegs_overlap$log2FoldChange.y, PrecysticP70 = alldegs_overlap$log2FoldChange)
row.names(alldegs_lfc) <- alldegs_overlap$X.x

```

###### Precyst Unique and Cystic Overlap
Cystic overlap enrichment
```{r}
cystic_geneoverlap_enrichment <- mouseenrich(degs_19_56_overlap_up$Mouse.gene.stable.ID, degs_19_56_overlap_down$Mouse.gene.stable.ID, 30)

cystic_geneoverlap_enrichment$topcompared$term_name <- stringr::str_wrap(cystic_geneoverlap_enrichment$topcompared$term_name, width = 70)


cystic_overlap_bubbleplot <- bubbleplot(cystic_geneoverlap_enrichment$topcompared)+ labs(title = "Top Enriched Terms for Overlapped Cystic DEGs") + theme(text = element_text(size = 10), plot.title = element_text(size = 11))

cystic_overlap_bubbleplot
```

Precystic unique enrichment
```{r}
#up genes
precystic_uniquegene_enrichment_up <- dplyr::anti_join(musdegs_39_UP, musdegs_19_UP, by = "Mouse.gene.stable.ID")
precystic_uniquegene_enrichment_up <- dplyr::anti_join(precystic_uniquegene_enrichment_up, musdegs_56_UP, by = "Mouse.gene.stable.ID")

#down genes
precystic_uniquegene_enrichment_down <- dplyr::anti_join(musdegs_39_DOWN, musdegs_19_DOWN, by = "Mouse.gene.stable.ID")
precystic_uniquegene_enrichment_down <- dplyr::anti_join(precystic_uniquegene_enrichment_down, musdegs_56_DOWN, by = "Mouse.gene.stable.ID")

precystic_unique_enrichment <- mouseenrich(precystic_uniquegene_enrichment_up$Mouse.gene.stable.ID, precystic_uniquegene_enrichment_down$Mouse.gene.stable.ID, 30)

precystic_unique_enrichment$topcompared$term_name <- stringr::str_wrap(precystic_unique_enrichment$topcompared$term_name, width = 60)

precyst_unique_bubbleplot <- bubbleplot(precystic_unique_enrichment$topcompared)+ labs(title = "Top Enriched Terms for Unique Precystic DEGs") + theme(text = element_text(size = 10), plot.title = element_text(size = 11))
precyst_unique_bubbleplot
```

save
```{r}
ggsave(cystic_overlap_bubbleplot, filename = here("res", "fea", "cystic_geneoverlap_fea.pdf"), height = 6, width = 8)

ggsave(precyst_unique_bubbleplot, filename = here("res", "fea", "precystic_geneunique_fea.pdf"), height = 6.5, width =7)

```

#### Cystic Overlap ('19 and '56)  

merge for overlap
```{r}
fea_cystic_overlap_up <- merge(fea_mouse_56$upregulated, fea_mouse_19$upregulated, by = "term_name")

fea_cystic_overlap_down <- merge(fea_mouse_56$downregulated, fea_mouse_19$downregulated, by = "term_name")
```

merge for overlap across all datasets
```{r}
fea_all_overlap_up <- merge(fea_cystic_overlap_up, fea_mouse_39_upterms, by = "term_name")

#no res for 39 down
#fea_all_overlap_down <- merge(fea_cystic_overlap_down, fea_mouse_39_downterms, by = "term_name")

```

FEA unique to precystic
```{r}
#termsin19 <- fea_mouse_39$upregulated$term_size[fea_mouse_39$upregulated]
fea_39_unique <- dplyr::anti_join(fea_mouse_39$upregulated, fea_mouse_56$upregulated, by = "term_name")
fea_39_unique <- dplyr::anti_join(fea_39_unique, fea_mouse_19$upregulated, by = "term_name")
#goterms <- fea$term_id[ fea$source == "GO:BP" & fea$direction == direction]


ggplot(slice_min(fea_39_unique, order_by = recall, n = 20), aes(x=intersection_size, y=term_name,  fill = p_value)) + 
    geom_bar( stat = "identity") + 
    scale_fill_continuous(type = "viridis") + 
    labs(x = "Number of Genes Matched to Term", y = "Functional Terms", title = "Terms Unique to Precystic DEG Enrichment") + 
    theme_minimal(base_size = 8)

```


##### Pathway Similarity
rrvgo package, uses GO terms for input
###### P70
```{r}
sim_deg_39 <-   go_bp_sim(fea_mouse_39, "org.Mm.eg.db", threshold = 0.80)

#options(ggrepel.max.overlaps = 50)
sp_deg_39 <- scatterPlot2(sim_deg_39$simMatrix, sim_deg_39$reducedTerms, labelSize = 2.5) + labs(title = "Pre-cystic P70 Signature Enrichment Overview")
sp_deg_39
tm_deg_39 <- treemapPlot2(sim_deg_39$reducedTerms, title = "Enrichment for Pre-cystic P70 DEGs")
hm_deg_39 <- heatmapPlot(sim_deg_39$simMatrix, sim_deg_39$reducedTerms, annotateParent=TRUE, annotationLabel="parentTerm", fontsize=6)

```

###### P21
```{r}
sim_deg_56 <-   go_bp_sim(fea_mouse_56, "org.Mm.eg.db", threshold = 0.80)

sp_deg_56 <- scatterPlot(sim_deg_56$simMatrix, sim_deg_56$reducedTerms, labelSize = 2.5) + labs(title = "Cystic P21 Signature Enrichment Overview")
sp_deg_56
tm_deg_56 <- treemapPlot2(sim_deg_56$reducedTerms, title = "Enrichment for Cystic P21 DEGs")
hm_deg_56 <- heatmapPlot(sim_deg_56$simMatrix, sim_deg_56$reducedTerms, annotateParent=TRUE, annotationLabel="parentTerm", fontsize=6)

```

###### P28
```{r}
sim_deg_19 <-   go_bp_sim(fea_mouse_19, "org.Mm.eg.db", threshold = 0.80)

sp_deg_19 <- scatterPlot2(sim_deg_19$simMatrix, sim_deg_19$reducedTerms, labelSize = 2.5) + labs(title = "Cystic P21 DEGs Enrichment Overview")
sp_deg_19
tm_deg_19 <- treemapPlot2(sim_deg_19$reducedTerms, title = "Enrichment for Cystic P28 DEGs")
hm_deg_19 <- heatmapPlot(sim_deg_19$simMatrix, sim_deg_19$reducedTerms, annotateParent=TRUE, annotationLabel="parentTerm", fontsize=6)
```

```{r}
#ggsave(hm, filename = here("res", "test.pdf"))

```
### Term overlaps

Cystic overlap
```{r}
fea_overlap_19_56_up <- merge(fea_mouse_19$upregulated, fea_mouse_56$upregulated, by = "term_name")

fea_overlap_19_56_down <- merge(fea_mouse_19$downregulated, fea_mouse_56$downregulated, by = "term_name")

fea_overlap_19_56_up$term_name
fea_overlap_19_56_down$term_name
```

Precystic overlap
```{r}
fea_overap_all_up <- merge(fea_mouse_39$upregulated, fea_overlap_19_56_up, by = "term_name")
fea_overap_all_down <- merge(fea_mouse_39$downregulated, fea_overlap_19_56_down, by = "term_name")

fea_overap_all_up$term_name
fea_overap_all_down$term_name
```

Save
``` {r}
fea_overlap_19_56_upterms <- apply(fea_overlap_19_56_up, 2, as.character)
write.csv(fea_overlap_19_56_upterms, file = here("res", "fea", "fea_overlap_19_56_upterms.csv"))

fea_overlap_19_56_downterms <- apply(fea_overlap_19_56_down, 2, as.character)
write.csv(fea_overlap_19_56_downterms, file = here("res", "fea", "fea_overlap_19_56_downterms.csv"))

fea_overap_all_upterms <- apply(fea_overap_all_up, 2, as.character)
write.csv(fea_overap_all_upterms, file = here("res", "fea", "fea_overap_all_upterms.csv"))

fea_overap_all_downterms <- apply(fea_overap_all_down, 2, as.character)
write.csv(fea_overap_all_downterms, file = here("res", "fea", "fea_overap_all_downterms.csv"))

```

#### Versions  
```{r}
R.Version()
```


```{r}
installed.packages()[names(sessionInfo()$otherPkgs), "Version"]
```
  
