---
title: "DESeq2 GSE149739: nf-core aligned"
author: "Lizzy Ramsey"
date: "3/30/2022"
output:
    html_document:
      toc: true
      toc_float: true
      toc_depth: 5

---

```{r setup, include=FALSE}
#filepath for code and data
knitr::opts_knit$set(root.dir = "/Users/eramsey/Downloads/kidney_drugrepurposing/pkd_drugrepurposing")
```
### Purpose  
This markdown uses nf-core for RNAseq alignment and DESeq2 DDS output  
Dataset: [GSE149739](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE149739)  
Processed data filepath: /data/project/lasseigne_lab/Lizzy/PKD_drugrepurposing_220324/  

#### Load in libraries
```{r message=FALSE, warning = FALSE}
library(DESeq2)
library(stringr)
library(apeglm)
library(pheatmap)
library(dplyr)
library(ggrepel)
library(signatureSearch)
library(ExperimentHub); library(rhdf5)
library(biomaRt)
```

#### Custom functions  

__salmon2dds__: Salmon merged gene counts from nf-core RNAseq pipeline (results/star_salmon/salmon.merged.gene_counts.rds) and SRA metadata as inputs, function data wrangles and runs DESeq2, resulting in a dds object with design by genotype    

```{r}
#Salmon merged gene counts to dds results
salmon2dds <- function(counts, metadata) {
  #count matrix in the sumarized experiment object
  cts <- counts@assays@data@listData[["counts"]]
  #round, salmon outputs have decimals
  cts <- round(cts)
  #colnames(cts) <- str_sub(colnames(cts), 1, -2) ##ALTERATION just for GSE69556
  #remove version numbers from transcript id's
  rownames(cts) <- gsub("\\..*", "", rownames(cts))
  
  #remove all non-alphanumeric characters from Genotype
  metadata$Genotype <- str_replace_all(metadata$Genotype, "[^[:alnum:]]", "")
  #contrasts need to be factor levels
  metadata$Genotype <- as.factor(metadata$Genotype)
#print(rownames(metadata))
#print(colnames(cts))
  dds <- DESeqDataSetFromMatrix(cts, metadata, design = ~ Genotype)
return(dds)
}
```

__uplfc__ & __downlfc__: Formats deseq2 res() output to select for upregulated and downregulated genes by log2 fold change threshold > 2 and < -2.  

```{r}
uplfc <- function(res){
  #NAs removed first
  res <- res[!is.na(res$log2FoldChange),]
  #p-adjusted value cutoff 0.05 and log2fc greater than 2.0 and less than -2.0
  upfc <- res[res$log2FoldChange > 2.0,]

  #pull just logfold change values from DESeqResults object and name the values by associated gene
  up <- upfc$log2FoldChange
  #names(up) <- rownames(upfc)
  up <- cbind(Mouse.gene.stable.ID = rownames(upfc), UpLFC = up)
  
return(as.data.frame(up))
}

downlfc <- function(res){
  #NAs removed first
  res <- res[!is.na(res$log2FoldChange),]
  #p-adjusted value cutoff 0.05 and log2fc greater than 2.0 and less than -2.0
  downfc <- res[res$log2FoldChange < -2.0,]

  #pull just logfold change values from DESeqResults object and name the values by associated gene
  down <- downfc$log2FoldChange
  #names(down) <- rownames(downfc)
  down <- cbind(Mouse.gene.stable.ID = rownames(downfc), DownLFC = down)
  
return(as.data.frame(down))
}
```

__convertMouseGeneList__: Uses Ensembl human GRCh38.p13 and Ensembl mouse C57BL_6NJ_v1 genes and to find human orthologous genes (more details [here](https://docs.google.com/document/d/1jU3EOVaZXJMzxwH9SsssHoArz8pMdcfW8MSr8hbeE4Q/edit) ), and biomart to convert to human entrez  

```{r}
convertMouseGeneList <- function(lfc){
  #read in mose to human ens annotation
  ens_annot <- read.csv("/Users/eramsey/Downloads/kidney_drugrepurposing/pkd_drugrepurposing/data/annot_ens_humanmouse.csv")
  #conversion table
  ens_mouse2human <- merge(lfc, ens_annot, by = "Mouse.gene.stable.ID")
  #biomart to conver human ens to human entrez
  human <- useEnsembl("ensembl", dataset = "hsapiens_gene_ensembl", mirror = "useast")
  genes <- getBM(filters = "ensembl_gene_id", attributes = c("ensembl_gene_id","entrezgene_id"), values = ens_mouse2human$Gene.stable.ID, mart = human)
  
  names(ens_mouse2human) <- c("mouse_ensembl_gene_id", "LFC", "ensembl_gene_id")
  conv_tbl <- merge(genes, ens_mouse2human, by = "ensembl_gene_id")

return(conv_tbl)
}
```

__toplincs__: Uses uplfc and downlfc outputs to then find top 100 genes that are available in LINCS (best input for signature reversion)  
```{r message=FALSE}
eh <- ExperimentHub()
lincs <- eh[["EH3226"]]; lincs_expr <- eh[["EH3227"]]

lincs_genes <- rhdf5:::h5read(lincs, "rownames", drop=TRUE)

toplincs <- function(conv_tbl, direction){
  lincs_genes <- cbind(entrezgene_id = lincs_genes)
  conv_tbl <- merge(conv_tbl, lincs_genes, by = "entrezgene_id")
  conv_tbl$LFC <- as.numeric(conv_tbl$LFC)
  if(direction == "up"){
  top100 <- slice_max(conv_tbl, n = 100, order_by = LFC)
  }
  else{
    top100 <- slice_min(conv_tbl, n = 100, order_by = LFC)
  }
return(top100)
}
```

__lincsGenes__: Wraps convertMouseGeneList to map orthologs, and toplincs to filter for LINCS-available genes 
```{r}
lincsGenes <- function(lfc, direction){
  conv_tbl <- convertMouseGeneList(lfc)
  top100 <- toplincs(conv_tbl, direction = direction)
  return(top100)
}
```

#### Read in data  
Filepath to counts data: /data/project/lasseigne_lab/Lizzy/PKD_drugrepurposing_220324/  
Metadata: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE149739 

```{r}
cts_39 <- readRDS("~/Downloads/kidney_drugrepurposing/pkd_drugrepurposing/data/GSE149739_salmonmerged_gene_counts.rds")

meta_39 <- read.csv("~/Downloads/kidney_drugrepurposing/pkd_drugrepurposing/data/GSE149739_metadata.txt", row.names = 1)
```

Remove IFT88
```{r}
cts_39_2 <- cts_39@assays@data@listData[["counts"]]
  #round, salmon outputs have decimals
cts_39_2 <- round(cts_39_2)
  #colnames(cts) <- str_sub(colnames(cts), 1, -2) ##ALTERATION just for GSE69556
  #remove version numbers from transcript id's
  rownames(cts_39_2) <- gsub("\\..*", "", rownames(cts_39_2))
  
  
meta_39_2 <- meta_39
  #remove all non-alphanumeric characters from Genotype
meta_39_2$Genotype <- str_replace_all(meta_39_2$Genotype, "[^[:alnum:]]", "")

#remove IFT88
meta_39_2 <- dplyr::filter(meta_39_2, Genotype != "Ift88flflPkd2flflPax8rtTATetOcre")
  #contrasts need to be factor levels
  meta_39_2$Genotype <- as.factor(meta_39_2$Genotype)
#print(rownames(metadata))
#print(colnames(cts))
  
cts_39_2 <- dplyr::select(cts_39_2, c(rownames(meta_39_2)))

#check ordering matches
colnames((cts_39_2)) == rownames(meta_39_2) #TRUE, good to proceed to DESeq2!
  
dds <- DESeqDataSetFromMatrix(cts_39_2, meta_39_2, design = ~ Genotype)

dds$Genotype <- relevel(dds$Genotype, ref = "Wildtype")

#run DESeq: estimates size factors, estimates dispersion, fits negative binomial GLM's and performs Wald stats
dds <- DESeq(dds) 

res <- results(dds, alpha = 0.05, contrast = c("Genotype", "Pkd2flflPax8rtTATetOcre", "Wildtype"))

resultsNames(res)
```

### DDS
```{r}
dds_39 <- salmon2dds(cts_39, meta_39)
```


#### Run DESeq2
```{r}
#relevel to make wildtype the reference instead of first alphabetical factor
dds_39$Genotype <- relevel(dds_39$Genotype, ref = "Wildtype")

#run DESeq: estimates size factors, estimates dispersion, fits negative binomial GLM's and performs Wald stats
dds_39 <- DESeq(dds_39) #can also look at setting replaceOutliers lower for dealing with outliers flagged by extreme Cook's distances
#save(ds, file = "./res/deseq2_outputs/GSE149739_dds_kasi2021alignment.rds")
```

#### Results  

```{r}
#extract results table with log2 fold changes, p values and adjusted p values for desired coefficients
#alpha 0.05 cutoff
res_nf <- results(dds_39, alpha = 0.05, contrast = c("Genotype", "Pkd2flflPax8rtTATetOcre", "Wildtype"))

```

DESeq2 results
```{r}
dsq2_39 <- as.data.frame(res_nf)

dsq2_39_up <- dplyr::filter(dsq2_39, log2FoldChange > 2.0)

dsq2_39_down <- dplyr::filter(dsq2_39, log2FoldChange < -2.0)
```

Save full deseq results
```{r}
write.csv(dsq2_39, file = "./res/deseq2_outputs/deseq2_fullres_39.csv")

write.csv(dsq2_39_up, file = "./res/deseq2_outputs/deseq2_upgenes_res_39.csv")

write.csv(dsq2_39_down, file = "./res/deseq2_outputs/deseq2_downgenes_res_39.csv")
```

#### LFC cutoffs
Using < -2 and > 2 LFC threshold for up and downregulated genes 
```{r}
musdegs_39_UP <- uplfc(res_nf)
head(musdegs_39_UP)
musdegs_39_DOWN <- downlfc(res_nf)
head(musdegs_39_DOWN)
```


```{r}
write.csv(musdegs_39_UP, file = "/Users/eramsey/Downloads/kidney_drugrepurposing/pkd_drugrepurposing/res/deseq2_outputs/GSE149739_mus_degs_UP_220421.csv")
write.csv(musdegs_39_DOWN, file = "/Users/eramsey/Downloads/kidney_drugrepurposing/pkd_drugrepurposing/res/deseq2_outputs/GSE149739_mus_degs_DOWN_220421.csv")

```


Variance stabilized transformation  
vst() uses a parametric fit for the dispersion. In this case, the closed-form expression for the variance stabilizing transformation is used by the vst function. If a local fit is used (option fitType="locfit" to estimateDispersions) a numerical integration is used instead. The transformed data should be approximated variance stabilized and also includes correction for size factors or normalization factors. The transformed data is on the log2 scale for large counts.  
  
rlog() is better to use for this dataset -- better for high count variance by sample (in this case, one sample has over 2x the reads compared to the others) and for smaller datasets "if you have many samples (e.g. 100s), the rlog function might take too long, and so the vst function will be a faster choice. The rlog and VST have similar properties, but the rlog requires fitting a shrinkage term for each sample and each gene which takes time. See the DESeq2 paper for more discussion on the differences (Love, Huber, and Anders 2014)."  
```{r}
#vsd_nf <- vst(dds_39, blind = FALSE) #if blind, could minimize large count differences due to anticipated experimental variation

rld_nf <- rlog(dds_39, blind = FALSE) 
```

PCA
```{r}
plotPCA(rld_nf, intgroup=c("Genotype")) + geom_label_repel(aes(label = rld_nf@colData@rownames), box.padding   = 0.15, label.padding = 0.15, point.padding = 0.3, segment.color = 'grey50') + scale_colour_viridis_d() + theme(axis.title.x = element_text(size = 20)) + theme_minimal()
```

### Mouse to Human Ortholog Mapping
Map to human entrez and filter by LINCS genes
```{r}
hom_degs_UP <- lincsGenes(musdegs_39_UP, "up")

hom_degs_DOWN <-  lincsGenes(musdegs_39_DOWN, "down")
```


```{r}
write.csv(hom_degs_UP, file = "/Users/eramsey/Downloads/kidney_drugrepurposing/pkd_drugrepurposing/res/deseq2_outputs/GSE149739_degs_UP_220421.csv")
write.csv(hom_degs_DOWN, file = "/Users/eramsey/Downloads/kidney_drugrepurposing/pkd_drugrepurposing/res/deseq2_outputs/GSE149739_degs_DOWN_220421.csv")
```



__Start of Ryan's Code__

### Heatmap plotting
```{r}
#Call in pheatmap package
library(pheatmap)
#formatting for pheatmap
#Gene_name and Gene ID dataframe for merging
conversion <- as.data.frame(cts_39@elementMetadata)
#Remove .# after ENMUSG names
conversion$gene_id <- gsub("\\..*", "", conversion$gene_id)
#retrieve counts data and ENMUSG #
cts_df <- as.data.frame(assays((rld_nf)))
#gene_id into a column
cts_df <- tibble:::rownames_to_column(cts_df, var = "gene_id")
#Merge gene symbols into counts and gene ID
conv_cts <- merge(cts_df, conversion, by = "gene_id")
#Remove unnecessary data
conv_cts <- conv_cts[,!(colnames(conv_cts) %in% c("gene_id", "group", "group_name"))]
#Filter out repeating data
conv_cts <- dplyr::distinct(conv_cts, gene_name, .keep_all = TRUE)
#Convert back to rownames but with gene symbols
conv_cts <- tibble:::column_to_rownames(conv_cts, "gene_name")
#Rename the count data to correspond with experimental group or control
#colnames(conv_cts) <- c("WT", "WT", "WT", "PKD2", "PKD2", "PKD2", "Ift88", "Ift88", "Ift88")

#remove ift88
conv_cts <- subset(conv_cts, select = c("SRR11680775", "SRR11680776", "SRR11680777", "SRR11680778", "SRR11680779", "SRR11680780"))
colnames(conv_cts) <- c("WT", "WT", "WT", "PKD2", "PKD2", "PKD2")
```


greatest LFC, wrangling
```{r}
genes <-  tibble:::rownames_to_column(dsq2_39, var = "gene_id")
#Merge gene symbols into counts and gene ID
genes <- merge(genes, conversion, by = "gene_id")
#Remove unnecessary data
genes <- genes[,!(colnames(genes) %in% c("gene_id", "group", "group_name"))]
#Filter out repeating data
genes <- dplyr::distinct(genes, gene_name, .keep_all = TRUE)
#Convert back to rownames but with gene symbols
genes <-  tibble:::column_to_rownames(genes, "gene_name")

genes <- arrange(genes, desc(log2FoldChange))#[1:20]
top20lfc <- slice_max(genes, (log2FoldChange^2), n = 20)
```

```{r}
pheatmap(conv_cts[rownames(top20lfc), ], cluster_rows = TRUE, cluster_cols = TRUE, show_rownames = TRUE, color = viridis(n= 10000, alpha = 1, begin = 0, end = 1, option = "viridis"), border_color = "NA", main = "Top 20 LFC Genes Pre-Cyst P70", angle_col = 45)

```

```{r}
#Create a heatmap of most varied genes
#sort by variance
Pre_Cyst_Variance <- apply(conv_cts, 1, var)
#top 20 most variable genes
top20_pre_cyst_variable <- names(sort(Pre_Cyst_Variance, decreasing = TRUE))[1:20]

```

```{r}
#### Versions  
#create heatmap
pheatmap(conv_cts[top20_pre_cyst_variable,], cluster_rows=TRUE, cluster_cols=TRUE, show_rownames=TRUE, color = viridis(n= 10000, alpha = 1, begin = 0, end = 1, option = "viridis"), border_color = "NA", main = "20 Most Varied Genes Pre-Cyst", angle_col = 45) #, annotation_col=meta_39$Genotype)
           
   #heatmap w/o gene symbols 20 greatest genes
genestop20 <- order(resLFC_nf$log2FoldChange, decreasing = TRUE)[1:20]
pheatmap(assay(cts_39)[genestop20, ], cluster_rows = TRUE, show_rownames = TRUE, cluster_cols = TRUE, color = viridis(n= 10000, alpha = 1, begin = 0, end = 1, option = "viridis"), border_color = "NA", main = "Top 20 Positive Genes Pre-Cyst", angle_col = 45)
#Heatmap w/o gene symbols 20 lowest genes
genesbottom20 <- order((resLFC_nf$log2FoldChange), decreasing = FALSE)[1:20]
pheatmap(assay(cts_39)[genesbottom20, ], cluster_rows = TRUE, show_rownames = TRUE, cluster_cols = TRUE, color = viridis(n= 10000, alpha = 1, begin = 0, end = 1, option = "viridis"), border_color = "NA", main = "Bottom 20 Positive Genes Pre-Cyst", angle_col = 45)
```

NEEDS REVIEW
Plotting alluvial plot for drug matching
```{r}
#Load in drug target list
gene_targets <- read.csv("C:\\Users\\mattr\\Documents\\Lasseigne_Lab_KURE\\TFEA_Lasseigne_Lab_Brownlee\\gene_drug_targets_39.csv")
#Extract drug and target columns
gene_target_alluvial <- gene_targets[,c(2,19)]
#Keep rows of drugs and targets that matched dataset results
gene_target_alluvial <- gene_target_alluvial[c(1,5,15,16,25,49,51,59,62,63,71,76,84,90,93,99,114,116,123,132,144,151,157),]
#Rename column "pert" to "Drug"
names(gene_target_alluvial)[1] <- "Drug"
#Recall packages for alluvial plotting
library(ggplot2)

library(ggalluvial)
#Create alluvial plot w/ viridis colors
target_alluvial_plot <- ggplot(data = gene_target_alluvial, aes(axis1 = Drug, axis2 = Target)) +
  geom_alluvium(color = "black", aes(fill = Target )) + # , colour = trend)) + 
  geom_stratum(width = 3/12, color = "black") + #the border of the stratum
  scale_x_discrete(expand = c(.07, .07)) + 
  scale_y_discrete(expand = c(.007, .007)) + 
  scale_fill_viridis_d(aesthetics = c("colour", "fill")) +
  geom_text(stat = "stratum", size = 3.5, fontface = "bold", aes(label = after_stat(stratum))) +
  labs(title = "Drug Target Matches") +
  theme_void() 
#View alluvial plot
target_alluvial_plot
```

Volcano plots
```{r}
  #Recall enhancedvoclano
  library(EnhancedVolcano)
  #Plot a basic volcano plot
  EnhancedVolcano(res_nf, lab = rownames(res_nf), x = 'log2FoldChange', y = 'pvalue', pCutoff = 10e-32, FCcutoff = 0.5, pointSize = 3.0, labSize = 6.0)
  #Set up data for Pre-Cyst Volcano Plot
TopT <- as.data.frame(res_nf)
#ENMUSG names to columns
TopT_test <- tibble:::rownames_to_column(TopT)
#Rename column of ENMSUG names
names(TopT_test)[1] <- "gene_id"
#Merge gene symbol into dataframe
TopT_finish <- left_join(TopT_test, conversion, by = "gene_id")
#ommit NA's
TopT_finish <- na.omit(TopT_finish)
#remove unneeded columns
TopT_finish <- TopT_finish[,!(colnames(TopT_finish) %in% c("gene_id", "baseMean", "lfcSE", "stat"))]
#Remove all repetitive rows
TopT_finish <- dplyr::distinct(TopT_finish, gene_name, .keep_all = TRUE)
#gene symbols to rownames
TopT_finish <- tibble:::column_to_rownames(TopT_finish, "gene_name")
View(TopT_finish)
#volcano plot w/ strict pvalue cutoff
EnhancedVolcano(TopT_finish, lab = rownames(TopT_finish), x = 'log2FoldChange', y = 'pvalue', pCutoff = 10e-32, FCcutoff = 0.5, pointSize = 3.0, labSize = 6.0, col = c("#440154", "#21918c", "#5ec962", "#fde725"))
#volcano plot w/ normal cutoffs and viridis colors
EnhancedVolcano(TopT_finish, lab = rownames(TopT_finish), x = 'log2FoldChange', y = 'pvalue', pointSize = 3.0, labSize = 4.0, col = c("#440154", "#21918c", "#5ec962", "#fde725"))
```

```{r}
# Read in Drug Targets
gene_targets <- read.csv("C:\\Users\\mattr\\Documents\\Lasseigne_Lab_KURE\\TFEA_Lasseigne_Lab_Brownlee\\gene_drug_targets_39.csv")
#New dataframe with drug and Target
gene_targets_unique <- data.frame(gene_targets$pert, gene_targets$Target)
#Rename columns
gene_targets_unique <- setNames(gene_targets_unique, c("Drug", "gene_name"))
#Check for DEG drug targets
#Dataframe w/ gene symbol and ENMUSG ID
conversion <- as.data.frame(cts_39@elementMetadata)
#Remove .#'s following ENMUSG
conversion$gene_id <- gsub("\\..*", "", conversion$gene_id)
#dataframe of LFC >2 from results
deg_39_up <- dsq2_39_up
#ENMUSG to column
deg_39_up <- tibble:::rownames_to_column(deg_39_up)
#rename column
names(deg_39_up)[1] <- "gene_id"
#merge gene symbols into dataframe
deg_39_up <- left_join(deg_39_up, conversion, by = "gene_id")
#remove unneeded columns
deg_39_up <- deg_39_up[,!(colnames(deg_39_up) %in% c("gene_id", "baseMean", "lfcSE", "stat"))]
#remove repeated symbols
deg_39_up <- dplyr::distinct(deg_39_up, gene_name, .keep_all = TRUE)
#deg_39_up <- tibble:::column_to_rownames(deg_39_up, "gene_name")
# deg_39_up is in lowercase and gene_targets_unique_df is uppercase. must fix!
deg_39_up$gene_name <- toupper(deg_39_up$gene_name)
#now fixed, merge for similarities in targets
target_gene_overlap_DESEQ2_39 <- merge(deg_39_up, gene_targets_unique, by = "gene_name")
View(target_gene_overlap_DESEQ2_39)
#save to file
write.csv(target_gene_overlap_DESEQ2_39, file = "C:\\Users\\mattr\\Documents\\Lasseigne_Lab_KURE\\TFEA_Lasseigne_Lab_Brownlee\\target_gene_overlap_DESEQ2_39.csv")
```

```{r}
rhub_launched_Drug_Targets <- read.csv("C:\\Users\\mattr\\Documents\\Lasseigne_Lab_KURE\\TFEA_Lasseigne_Lab_Brownlee\\rhub_launcheddrugs_39.csv")
#separate gene targets into individual rows
rhub_launched_Drug_Targets <- tidyr::separate_rows(rhub_launched_Drug_Targets, Target, sep = ", ")
#remove drugs with empty 'target' field
rhub_launched_Drug_Targets <- dplyr::filter(rhub_launched_Drug_Targets, Target != "")
#Make new dataframe of just drug and gene name
Rhub_Launched_Drug_Targets_Unique <- data.frame(rhub_launched_Drug_Targets$pert, rhub_launched_Drug_Targets$Target)
#rename for merging function
Rhub_Launched_Drug_Targets_Unique <- setNames(Rhub_Launched_Drug_Targets_Unique, c("Drug", "gene_name"))
#Same process as mentioned above. refer to previous annotations.
conversion <- as.data.frame(cts_39@elementMetadata)
conversion$gene_id <- gsub("\\..*", "", conversion$gene_id)
deg_39_up <- dsq2_39_up
deg_39_up <- tibble:::rownames_to_column(deg_39_up)
names(deg_39_up)[1] <- "gene_id"
deg_39_up <- left_join(deg_39_up, conversion, by = "gene_id")
deg_39_up <- deg_39_up[,!(colnames(deg_39_up) %in% c("gene_id", "baseMean", "lfcSE", "stat"))]
deg_39_up <- dplyr::distinct(deg_39_up, gene_name, .keep_all = TRUE)
deg_39_up$gene_name <- toupper(deg_39_up$gene_name)
rhub_target_gene_overlap_DESEQ2_39 <- merge(deg_39_up, Rhub_Launched_Drug_Targets_Unique, by = "gene_name")
View(rhub_target_gene_overlap_DESEQ2_39)
#write.csv(rhub_target_gene_overlap_DESEQ2_39, file = "C:\\Users\\mattr\\Documents\\Lasseigne_Lab_KURE\\TFEA_Lasseigne_Lab_Brownlee\\rhub_target_gene_overlap_DESEQ2_39.csv")
```


```{r}
R.Version()
```


```{r}
installed.packages()[names(sessionInfo()$otherPkgs), "Version"]
```
  
