---
title: "DESeq2 GSE149739: nf-core aligned"
author: "Lizzy Ramsey"
date: "3/30/2022"
output:
    html_document:
      toc: true
      toc_float: true
      toc_depth: 5

---


### Purpose  
This markdown uses nf-core for RNAseq alignment, STAR salmon counts. All data from publicly available mouse kidney RNAseq data comparing PKD2-KO and WT.  
__p70__: AKA 39/GSE149739/precystic/p70, induced-KO model, samples collected at p70 when tubules were dilated with no overt cyst formation. Total n = 6 samples, 3 Pkd2fl/fl;Pax8rtTA;TetO-Cre and 3 WT  
Dataset: [GSE149739](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE149739)  
Processed data filepath: /data/project/lasseigne_lab/Lizzy/PKD_drugrepurposing_220324/  

__p21__: AKA 56/GSE69556/cystic/p21, kidney-specific KO model, samples collected at p21 from full-blown cystic kidneys. Total n = 7, 3 Pkhd1-Cre; Pkd2fl/fl and 4 WT   
Dataset: [GSE69556](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE69556)  
Processed data filepath: /data/project/lasseigne_lab/Lizzy/PKD_drugrepurposing_220324/  

__p28__: AKA 19/GSE134719/cystic/p28, kidney-specific KO model, samples collected at p28 from full-blown cystic kidneys. Total n = 23, 11 Pkhd1-Cre; Pkd2fl/fl and 12 WT     
Dataset: [GSE134719](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE134719)  
Processed data filepath: /data/project/lasseigne_lab/Lizzy/PKD_drugrepurposing_220324/  


#### Load in libraries
```{r message=FALSE, warning = FALSE}
library(DESeq2)
library(stringr)
library(apeglm)
library(pheatmap)
library(dplyr)
library(ggrepel)
library(signatureSearch)
library(ExperimentHub); library(rhdf5)
library(biomaRt)
library(viridis)
library(here)
library(EnhancedVolcano)
```

#### Custom functions  

__salmon2dds__: Salmon merged gene counts from nf-core RNAseq pipeline (results/star_salmon/salmon.merged.gene_counts.rds) and SRA metadata as inputs, function data wrangles and runs DESeq2, resulting in a dds object with design by genotype. Just for dataset  GSE69556 (aka 56, p21), set argument data = not "regular", for extra characters in counts rownames just in that dataset  

```{r}
#Salmon merged gene counts to dds results
salmon2dds <- function(counts, metadata, dataset = "regular") {
  #count matrix in the sumarized experiment object
  cts <- counts@assays@data@listData[["counts"]]
  #round, salmon outputs have decimals
  cts <- round(cts)
  if (dataset != "regular"){
    colnames(cts) <- str_sub(colnames(cts), 1, -2) ##ALTERATION just for GSE69556  
  }
  
  #remove version numbers from transcript id's
  rownames(cts) <- gsub("\\..*", "", rownames(cts))
  
  #remove all non-alphanumeric characters from Genotype
  metadata$Genotype <- str_replace_all(metadata$Genotype, "[^[:alnum:]]", "")
  #contrasts need to be factor levels
  metadata$Genotype <- as.factor(metadata$Genotype)

  #check ordering matches
  print("Checking order of count column names to metadata rownames... If false, fix and rerun!")
  colnames((cts)) == rownames(metadata)
  dds <- DESeqDataSetFromMatrix(cts, metadata, design = ~ Genotype)

return(dds)
}


```


__uplfc__ & __downlfc__: Formats deseq2 res() output to select for upregulated and downregulated genes by log2 fold change threshold > 2 and < -2.  

```{r}
uplfc <- function(res){
  #NAs removed first
  res <- res[!is.na(res$log2FoldChange),]
  #p-adjusted value cutoff 0.05 and log2fc greater than 2.0 and less than -2.0
  upfc <- res[res$log2FoldChange > 2.0,]

  #pull just logfold change values from DESeqResults object and name the values by associated gene
  up <- upfc$log2FoldChange
  #names(up) <- rownames(upfc)
  up <- cbind(Mouse.gene.stable.ID = rownames(upfc), LFC = up)
  
return(as.data.frame(up))
}

downlfc <- function(res){
  #NAs removed first
  res <- res[!is.na(res$log2FoldChange),]
  #p-adjusted value cutoff 0.05 and log2fc greater than 2.0 and less than -2.0
  downfc <- res[res$log2FoldChange < -2.0,]

  #pull just logfold change values from DESeqResults object and name the values by associated gene
  down <- downfc$log2FoldChange
  #names(down) <- rownames(downfc)
  down <- cbind(Mouse.gene.stable.ID = rownames(downfc), LFC = down)
  
return(as.data.frame(down))
}
```

__convertMouseGeneList__: Uses Ensembl human GRCh38.p13 and Ensembl mouse C57BL_6NJ_v1 genes and to find human orthologous genes (more details [here](https://docs.google.com/document/d/1jU3EOVaZXJMzxwH9SsssHoArz8pMdcfW8MSr8hbeE4Q/edit) ), and biomart to convert to human entrez  

```{r}
convertMouseGeneList <- function(lfc){
  #read in mose to human ens annotation
  ens_annot <- read.csv(here("data", "annot_ens_humanmouse.csv"))
  #conversion table
  ens_mouse2human <- merge(lfc, ens_annot, by = "Mouse.gene.stable.ID")
  #biomart to conver human ens to human entrez
  human <- useEnsembl("ensembl", dataset = "hsapiens_gene_ensembl", mirror = "useast")
  genes <- getBM(filters = "ensembl_gene_id", attributes = c("ensembl_gene_id","entrezgene_id"), values = ens_mouse2human$Gene.stable.ID, mart = human)
  
  names(ens_mouse2human) <- c("mouse_ensembl_gene_id", "LFC", "ensembl_gene_id")
  conv_tbl <- merge(genes, ens_mouse2human, by = "ensembl_gene_id")
return(conv_tbl)
}
```

__toplincs__: Uses uplfc and downlfc outputs to then find top 100 genes that are available in LINCS (best input for signature reversion)  
```{r message=FALSE}
eh <- ExperimentHub()
lincs <- eh[["EH3226"]]; lincs_expr <- eh[["EH3227"]]

lincs_genes <- rhdf5:::h5read(lincs, "rownames", drop=TRUE)

toplincs <- function(conv_tbl, direction){
  lincs_genes <- cbind(entrezgene_id = lincs_genes)
  conv_tbl <- merge(conv_tbl, lincs_genes, by = "entrezgene_id")
  conv_tbl$LFC <- as.numeric(conv_tbl$LFC)
  if(direction == "up"){
  top100 <- slice_max(conv_tbl, n = 100, order_by = LFC)
  }
  else{
    top100 <- slice_min(conv_tbl, n = 100, order_by = LFC)
  }
return(top100)
}
```

__lincsGenes__: Wraps convertMouseGeneList to map orthologs, and toplincs to filter for LINCS-available genes  
```{r}
lincsGenes <- function(lfc, direction){
  conv_tbl <- convertMouseGeneList(lfc)
  top100 <- toplincs(conv_tbl, direction = direction)
  return(top100)
}
```

##DEA p70 precystic GSE149739

#### Read in data  
Filepath to counts data: /data/project/lasseigne_lab/Lizzy/PKD_drugrepurposing_220324/  
Metadata: https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE149739 

```{r}
cts_39 <- readRDS(here("data", "GSE149739_salmonmerged_gene_counts.rds"))

meta_39 <- read.csv(here("data", "GSE149739_metadata.txt"), row.names = 1)
```

Remove IFT88
```{r}
cts_39_2 <- cts_39@assays@data@listData[["counts"]]
  #round, salmon outputs have decimals
cts_39_2 <- round(cts_39_2)
  #colnames(cts) <- str_sub(colnames(cts), 1, -2) ##ALTERATION just for GSE69556
  #remove version numbers from transcript id's
  rownames(cts_39_2) <- gsub("\\..*", "", rownames(cts_39_2))
  
  
meta_39_2 <- meta_39
  #remove all non-alphanumeric characters from Genotype
meta_39_2$Genotype <- str_replace_all(meta_39_2$Genotype, "[^[:alnum:]]", "")

#remove IFT88
meta_39_2 <- dplyr::filter(meta_39_2, Genotype != "Ift88flflPkd2flflPax8rtTATetOcre")
  #contrasts need to be factor levels
  meta_39_2$Genotype <- as.factor(meta_39_2$Genotype)

  
cts_39_2 <- dplyr::select(cts_39_2, c(rownames(meta_39_2)))

#check ordering matches
colnames((cts_39_2)) == rownames(meta_39_2) #TRUE, good to proceed to DESeq2!
  
dds <- DESeqDataSetFromMatrix(cts_39_2, meta_39_2, design = ~ Genotype)

dds$Genotype <- relevel(dds$Genotype, ref = "Wildtype")

#run DESeq: estimates size factors, estimates dispersion, fits negative binomial GLM's and performs Wald stats
dds <- DESeq(dds) 

res <- results(dds, alpha = 0.05, contrast = c("Genotype", "Pkd2flflPax8rtTATetOcre", "Wildtype"))

resultsNames(res)
```

#### DDS
```{r}
dds_39 <- salmon2dds(cts_39, meta_39)
```


#### Run DESeq2
```{r}
#relevel to make wildtype the reference instead of first alphabetical factor
dds_39$Genotype <- relevel(dds_39$Genotype, ref = "Wildtype")

#run DESeq: estimates size factors, estimates dispersion, fits negative binomial GLM's and performs Wald stats
dds_39 <- DESeq(dds_39) #can also look at setting replaceOutliers lower for dealing with outliers flagged by extreme Cook's distances
#save(ds, file = here("res", "deseq2_outputs", "GSE149739_dds_kasi2021alignment.rds")
```

#### Results  

```{r}
#extract results table with log2 fold changes, p values and adjusted p values for desired coefficients
#alpha 0.05 cutoff
res_nf <- results(dds_39, alpha = 0.05, contrast = c("Genotype", "Pkd2flflPax8rtTATetOcre", "Wildtype"))

```

DESeq2 results
```{r}
dsq2_39 <- as.data.frame(res_nf)

dsq2_39_up <- dplyr::filter(dsq2_39, log2FoldChange > 2.0)

dsq2_39_down <- dplyr::filter(dsq2_39, log2FoldChange < -2.0)
```

Save full deseq results
```{r}
write.csv(dsq2_39, file = here("res", "deseq2_outputs", "deseq2_fullres_39.csv"))

write.csv(dsq2_39_up, file = here("res", "deseq2_outputs", "deseq2_upgenes_res_39.csv"))

write.csv(dsq2_39_down, file = here("res", "deseq2_outputs", "deseq2_downgenes_res_39.csv"))
```

#### LFC cutoffs
Using < -2 and > 2 LFC threshold for up and downregulated genes 
```{r}
musdegs_39_UP <- uplfc(res_nf)
head(musdegs_39_UP)
musdegs_39_DOWN <- downlfc(res_nf)
head(musdegs_39_DOWN)
```


```{r eval=FALSE, include=FALSE}
write.csv(musdegs_39_UP, file = here("res", "deseq2_outputs", "GSE149739_mus_degs_UP_220421.csv"))

write.csv(musdegs_39_DOWN, file = here("res", "deseq2_outputs", "GSE149739_mus_degs_DOWN_220421.csv"))

```


Variance stabilized transformation  
vst() uses a parametric fit for the dispersion. In this case, the closed-form expression for the variance stabilizing transformation is used by the vst function. If a local fit is used (option fitType="locfit" to estimateDispersions) a numerical integration is used instead. The transformed data should be approximated variance stabilized and also includes correction for size factors or normalization factors. The transformed data is on the log2 scale for large counts.  
  
rlog() is better to use for this dataset -- better for high count variance by sample (in this case, one sample has over 2x the reads compared to the others) and for smaller datasets "if you have many samples (e.g. 100s), the rlog function might take too long, and so the vst function will be a faster choice. The rlog and VST have similar properties, but the rlog requires fitting a shrinkage term for each sample and each gene which takes time. See the DESeq2 paper for more discussion on the differences (Love, Huber, and Anders 2014)."  
```{r}
#vsd_nf <- vst(dds_39, blind = FALSE) #if blind, could minimize large count differences due to anticipated experimental variation

rld_nf <- rlog(dds_39, blind = FALSE) 
```

PCA
```{r}
plotPCA(rld_nf, intgroup=c("Genotype")) + geom_label_repel(aes(label = rld_nf@colData@rownames), box.padding   = 0.15, label.padding = 0.15, point.padding = 0.3, segment.color = 'grey50') + scale_colour_viridis_d() + theme(axis.title.x = element_text(size = 20)) + theme_minimal()
```

#### Mouse to Human Ortholog Mapping
Map to human entrez and filter by LINCS genes
```{r}
hom_degs_UP <- lincsGenes(musdegs_39_UP, "up")
 
hom_degs_DOWN <-  lincsGenes(musdegs_39_DOWN, "down")
```


```{r eval=FALSE, include=FALSE}
write.csv(hom_degs_UP, file = here("res", "deseq2_outputs", "GSE149739_degs_UP_220421.csv"))
write.csv(hom_degs_DOWN, file = here("res", "deseq2_outputs", "GSE149739_degs_DOWN_220421.csv"))
```


#### Heatmap plotting
```{r}
#Call in pheatmap package
library(pheatmap)
#formatting for pheatmap
#Gene_name and Gene ID dataframe for merging
conversion <- as.data.frame(cts_39@elementMetadata)
#Remove .# after ENMUSG names
conversion$gene_id <- gsub("\\..*", "", conversion$gene_id)
#retrieve counts data and ENMUSG #
cts_df <- as.data.frame(assays((rld_nf)))
#gene_id into a column
cts_df <- tibble:::rownames_to_column(cts_df, var = "gene_id")
#Merge gene symbols into counts and gene ID
conv_cts <- merge(cts_df, conversion, by = "gene_id")
#Remove unnecessary data
conv_cts <- conv_cts[,!(colnames(conv_cts) %in% c("gene_id", "group", "group_name"))]
#Filter out repeating data
conv_cts <- dplyr::distinct(conv_cts, gene_name, .keep_all = TRUE)
#Convert back to rownames but with gene symbols
conv_cts <- tibble:::column_to_rownames(conv_cts, "gene_name")
#Rename the count data to correspond with experimental group or control
#colnames(conv_cts) <- c("WT", "WT", "WT", "PKD2", "PKD2", "PKD2", "Ift88", "Ift88", "Ift88")

#remove ift88
conv_cts <- subset(conv_cts, select = c("SRR11680775", "SRR11680776", "SRR11680777", "SRR11680778", "SRR11680779", "SRR11680780"))
colnames(conv_cts) <- c("WT", "WT", "WT", "PKD2", "PKD2", "PKD2")
```


greatest LFC, wrangling
```{r}
genes <-  tibble:::rownames_to_column(dsq2_39, var = "gene_id")
#Merge gene symbols into counts and gene ID
genes <- merge(genes, conversion, by = "gene_id")
#Remove unnecessary data
genes <- genes[,!(colnames(genes) %in% c("gene_id", "group", "group_name"))]
#Filter out repeating data
genes <- dplyr::distinct(genes, gene_name, .keep_all = TRUE)
#Convert back to rownames but with gene symbols
genes <-  tibble:::column_to_rownames(genes, "gene_name")

genes <- arrange(genes, desc(log2FoldChange))#[1:20]
top20lfc <- slice_max(genes, (log2FoldChange^2), n = 20)
```


```{r}
heatmap_p70 <- pheatmap(conv_cts[rownames(top20lfc), ], cluster_rows = TRUE, cluster_cols = TRUE, show_rownames = TRUE, color = viridis(n= 10000, alpha = 1, begin = 0, end = 1, option = "viridis"), border_color = "NA", main = "Top 20 LFC Genes Pre-Cyst P70", angle_col = 45)

```

```{r}
#ggsave(heatmap_p70, filename = here("res", "fea", "heatmap_p70.pdf", height = 7, width = 14)
```


#### Volcano plots

#####p70 DEGs to p70 targets
convert from mouse ensembl to human symbol
```{r}
ens_annot <- read.csv(here("data", "annot_ens_humanmouse.csv"))
ens_mouse2human <- tibble:::rownames_to_column(dsq2_39, var = "Mouse.gene.stable.ID")
  #conversion table
ens_mouse2human <- merge(ens_mouse2human, ens_annot, by = "Mouse.gene.stable.ID")
#biomart to conver human ens to human entrez
human <- useEnsembl("ensembl", dataset = "hsapiens_gene_ensembl", mirror = "useast")
convgenes <- getBM(filters = "ensembl_gene_id", attributes = c("ensembl_gene_id","entrezgene_id", "hgnc_symbol"), values = ens_mouse2human$Gene.stable.ID, mart = human)
  
conv_tbl <- merge(convgenes, ens_mouse2human, by.x = "ensembl_gene_id", by.y = "Gene.stable.ID")

```

Drug targets from FDA approved candidates from signature reversion of p70
```{r}
targets_39 <- read.csv(here("res", "sigsearch_outputs", "rhub_launcheddrugs_39.csv"), row.names = 1)
targets_39 <- tidyr::separate_rows(targets_39, "t_gn_sym", sep = "; ")
```

```{r eval=FALSE, include=FALSE}
write.csv(targets_39, file = here("res", "sigsearch_outputs", "drugtargets_p70.csv"))

```

```{r}
conv_tbl <- tidyr::drop_na(conv_tbl)

EnhancedVolcano(conv_tbl, lab = conv_tbl$hgnc_symbol, x = 'log2FoldChange', y = 'padj', selectLab = unique(targets_39$t_gn_sym), pCutoff = 0.05, FCcutoff = 0.5, pointSize = 3.0, labSize = 4.0, col = c("#440154", "#21918c", "#5ec962", "#fde725"), title = "Drug Targets Found in P70 Pre-Cystic DEGs")#, labFace = 'bold', boxedLabels = TRUE,drawConnectors = TRUE,widthConnectors = 0.5,colConnectors = 'black')

```


##DEA p21 cystic GSE69556


### Read in data  
__Filepath to fastq's and metadata__: /data/project/lasseigne_lab/DATASET_dir/pkdDrugRepurposing/pkd2/GSE69556/SraRunTable.txt  
__Filepath to counts data__: /data/project/lasseigne_lab/Lizzy/PKD_drugrepurposing_220324/GSE69556_nfcore/results/star_salmon/salmon.merged.gene_counts.rds  
```{r}
cts_56 <- readRDS(here("data", "GSE69556_salmonmerged_gene_counts.rds"))
#cts_56 <- GSE69556_salmonmerged_gene_counts
#colnames(cts_56) <- str_sub(colnames(cts_56), 1, -2)

meta_56 <- read.csv(here("data", "GSE69556_metadata.txt"), row.names = 1)
```

#### Run DDS
```{r}
dds_56 <- salmon2dds(cts_56, meta_56, dataset = "56")
str(dds_56)
```

#### Run DESeq2
```{r}
#relevel to make wildtype the reference instead of first alphabetical factor
dds_56$Genotype <- relevel(dds_56$Genotype, ref = "wildtype")

#run DESeq: estimates size factors, estimates dispersion, fits negative binomial GLM's and performs Wald stats
dds_56 <- DESeq(dds_56) #can also look at setting replaceOutliers lower for dealing with outliers 
```

#### Results  

```{r}
#extract results table with log2 fold changes, p values and adjusted p values for desired coefficients
res_56 <- results(dds_56, alpha = 0.05, contrast = c("Genotype", "Pkhd1CrePkd2FF", "wildtype"))

```

```{r eval=FALSE, include=FALSE}
write.csv(res_56, file = here("res", "deseq2_outputs", deseq2_fullres_56.csv"))

```

LFC cutoffs
```{r}
#NAs removed  p-adjusted value cutoff 0.05 and log2fc greater than 2.0 and less than -2.0
musdegs_56_UP <- uplfc(res_56)

musdegs_56_DOWN <- downlfc(res_56)
```

Map to human entrez and filter by LINCS genes
```{r}
hom_56_UP <- lincsGenes(musdegs_56_UP, "up")

hom_56_DOWN <-  lincsGenes(musdegs_56_DOWN, "down")
```


```{r eval=FALSE, include=FALSE}
write.csv(musdegs_56_UP, file = here("res", "deseq2_outputs", "GSE69556_mus_degs_UP_220421.csv"))
write.csv(musdegs_56_DOWN, file = here("res", "deseq2_outputs", "GSE69556_mus_degs_DOWN_220421.csv"))

write.csv(hom_56_UP, file = here("res", "deseq2_outputs", "GSE69556_degs_UP_220421.csv"))
write.csv(hom_56_DOWN, file = here("res", "deseq2_outputs", "GSE69556_degs_DOWN_220421.csv"))

```


Variance stabilized transformation  

rlog() may be better to use for this dataset -- better for high count variance by sample (in this case, one sample has over 2x the reads compared to the others) and for smaller datasets "if you have many samples (e.g. 100s), the rlog function might take too long, and so the vst function will be a faster choice. The rlog and VST have similar properties, but the rlog requires fitting a shrinkage term for each sample and each gene which takes time. See the DESeq2 paper for more discussion on the differences (Love, Huber, and Anders 2014)."  
```{r}
rld_56 <- rlog(dds_56, blind = FALSE) 
```

#### Visualizations  

PCA
```{r}
#regularized log transformation
plotPCA(rld_56, intgroup=c("Genotype"))
#label samples
plotPCA(rld_56, intgroup=c("Genotype")) + geom_label_repel(aes(label = rld_56@colData@rownames), box.padding   = 0.15, label.padding = 0.15, point.padding = 0.3, segment.color = 'grey50') 
```

#### Heatmap plotting
```{r}
#Call in pheatmap package
library(pheatmap)
#formatting for pheatmap
#Gene_name and Gene ID dataframe for merging
conversion <- as.data.frame(cts_56@elementMetadata)
#Remove .# after ENMUSG names
conversion$gene_id <- gsub("\\..*", "", conversion$gene_id)
#retrieve counts data and ENMUSG #
cts_df_56 <- as.data.frame(assays((rld_56)))
#gene_id into a column
cts_df_56 <- tibble:::rownames_to_column(cts_df_56, var = "gene_id")
#Merge gene symbols into counts and gene ID
conv_cts_56 <- merge(cts_df_56, conversion, by = "gene_id")
#Remove unnecessary data
conv_cts_56 <- conv_cts_56[,!(colnames(conv_cts_56) %in% c("gene_id", "group", "group_name"))]
#Filter out repeating data
conv_cts_56 <- dplyr::distinct(conv_cts_56, gene_name, .keep_all = TRUE)
#Convert back to rownames but with gene symbols
conv_cts_56 <- tibble:::column_to_rownames(conv_cts_56, "gene_name")
#Rename the count data to correspond with experimental group or control
colnames(conv_cts_56) <- c("wildtype", "wildtype", "wildtype", "wildtype", "PKD2", "PKD2", "PKD2")
```

greatest LFC, wrangling
```{r}
dsq2_56 <- as.data.frame(res_56)
genes_56 <-  tibble:::rownames_to_column(dsq2_56, var = "gene_id")
#Merge gene symbols into counts and gene ID
genes_56 <- merge(genes_56, conversion, by = "gene_id")
#Remove unnecessary data
genes_56 <- genes_56[,!(colnames(genes_56) %in% c("gene_id", "group", "group_name"))]
#Filter out repeating data
genes_56 <- dplyr::distinct(genes_56, gene_name, .keep_all = TRUE)
#Convert back to rownames but with gene symbols
genes_56 <-  tibble:::column_to_rownames(genes_56, "gene_name")

genes_56 <- arrange(genes_56, desc(log2FoldChange))#[1:20]
top20lfc_56 <- slice_max(genes_56, (log2FoldChange^2), n = 20)
```

```{r}
heatmap_21 <- pheatmap(conv_cts_56[rownames(top20lfc_56), ], cluster_rows = TRUE, cluster_cols = TRUE, show_rownames = TRUE, color = viridis(n= 10000, alpha = 1, begin = 0, end = 1, option = "viridis"), border_color = "NA", main = "Top 20 LFC Genes Cystic P21", angle_col = 45)

```

```{r eval=FALSE, include=FALSE}
ggsave(heatmap_21, filename = "./res/fea/heatmap_p21.pdf", height = 7, width = 14)

```


#### p21 DEGs to p70 Targets

convert from mouse ensembl to human symbol
```{r}
dsq2_56 <- read.csv(here("res", "deseq2_outputs", "deseq2_fullres_56.csv"), row.names = 1)
ens_mouse2human_56 <- tibble:::rownames_to_column(dsq2_56, var = "Mouse.gene.stable.ID")
  #conversion table
ens_mouse2human_56 <- merge(ens_mouse2human_56, ens_annot, by = "Mouse.gene.stable.ID")
#biomart to conver human ens to human entrez
human_56 <- useEnsembl("ensembl", dataset = "hsapiens_gene_ensembl", mirror = "useast")
convgenes_56 <- getBM(filters = "ensembl_gene_id", attributes = c("ensembl_gene_id","entrezgene_id", "hgnc_symbol"), values = ens_mouse2human_56$Gene.stable.ID, mart = human)
  
conv_tbl_56 <- merge(convgenes_56, ens_mouse2human_56, by.x = "ensembl_gene_id", by.y = "Gene.stable.ID")

conv_tbl_56 <- tidyr::drop_na(conv_tbl_56)
```


```{r}
EnhancedVolcano(conv_tbl_56, lab = conv_tbl_56$hgnc_symbol, x = 'log2FoldChange', y = 'padj', selectLab = unique(targets_39$t_gn_sym), pCutoff = 0.05, FCcutoff = 0.5, pointSize = 3.0, labSize = 4.0, col = c("#440154", "#21918c", "#5ec962", "#fde725"), title = "P70 Pre-Cystic Drug Targets Found in P21 Cystic DEGs")#, labFace = 'bold', boxedLabels = TRUE,drawConnectors = TRUE,widthConnectors = 0.5,colConnectors = 'black')

```

##DEA p28 cystic GSE134719


#### Read in data  
__Filepath to fastq's and metadata__: /data/project/lasseigne_lab/DATASET_dir/pkdDrugRepurposing/pkd2/GSE134719/SraRunTable.txt  
__Filepath to counts data__: /data/project/lasseigne_lab/Lizzy/PKD_drugrepurposing_220324/GSE134719_nfcore/results/star_salmon/salmon.merged.gene_counts.rds  
```{r}
cts_19 <- readRDS(here("data", "GSE134719_salmonmerged_gene_counts.rds"))

meta_19 <- read.csv(here("data", "GSE134719_metadata.txt"), row.names = 1)
```

#### DDS
```{r}
dds_19 <- salmon2dds(cts_19, meta_19)
str(dds_19)
```

#### Run DESeq2
```{r}
#relevel to make wildtype the reference instead of first alphabetical factor
dds_19$Genotype <- relevel(dds_19$Genotype, ref = "wildtype")

#run DESeq: estimates size factors, estimates dispersion, fits negative binomial GLM's and performs Wald stats
dds_19 <- DESeq(dds_19) #can also look at setting replaceOutliers lower for dealing with outliers 
```

#### Results  

```{r}
#extract results table with log2 fold changes, p values and adjusted p values for desired coefficients
res_19 <- results(dds_19, alpha = 0.05, contrast = c("Genotype", "Pdk2KO", "wildtype"))

```

Save full deseq2 results
```{r eval=FALSE, include=FALSE}
write.csv(res_19, file = here("res", "deseq2_outputs", "deseq2_fullres_19.csv"))

```

#### LFC cutoffs
```{r}
musdegs_19_UP <- uplfc(res_19)

musdegs_19_DOWN <- downlfc(res_19)
```

Variance stabilized transformation  

rlog() may be better to use for this dataset -- better for high count variance by sample (in this case, one sample has over 2x the reads compared to the others) and for smaller datasets "if you have many samples (e.g. 100s), the rlog function might take too long, and so the vst function will be a faster choice. The rlog and VST have similar properties, but the rlog requires fitting a shrinkage term for each sample and each gene which takes time. See the DESeq2 paper for more discussion on the differences (Love, Huber, and Anders 2014)."  
```{r}
rld_19 <- rlog(dds_19, blind = FALSE) 
```


#### Visualizations  

PCA
```{r}
#regularized log transformation
#plotPCA(rld_19, intgroup=c("Genotype"))
#label samples
plotPCA(rld_19, intgroup=c("Genotype")) + geom_label_repel(aes(label = rld_19@colData@rownames), box.padding   = 0.15, label.padding = 0.15, point.padding = 0.3, segment.color = 'grey50') 
```


#### Mouse to Human Ortholog Mapping
Map to human entrez and filter by LINCS genes
```{r}
hom_19_UP <- lincsGenes(musdegs_19_UP, "up")

hom_19_DOWN <-  lincsGenes(musdegs_19_DOWN, "down")
```


```{r eval=FALSE, include=FALSE}
write.csv(musdegs_19_UP, file = here("res", "deseq2_outputs", "GSE134719_mus_degs_UP_220421.csv"))
write.csv(musdegs_19_DOWN, file = here("res", "deseq2_outputs", "GSE134719_mus_degs_DOWN_220421.csv"))
 
write.csv(hom_19_UP, file = here("res", "deseq2_outputs", "GSE134719_degs_UP_220421.csv"))
write.csv(hom_19_DOWN, file = here("res", "deseq2_outputs", "GSE134719_degs_DOWN_220421.csv"))

```


__Start of Ryan's Code__

#### Heatmap plotting
```{r}
#Call in pheatmap package
library(pheatmap)
#formatting for pheatmap
#Gene_name and Gene ID dataframe for merging
conversion <- as.data.frame(cts_19@elementMetadata)
#Remove .# after ENMUSG names
conversion$gene_id <- gsub("\\..*", "", conversion$gene_id)
#retrieve counts data and ENMUSG #
cts_df_19 <- as.data.frame(assays((rld_19)))
#gene_id into a column
cts_df_19 <- tibble:::rownames_to_column(cts_df_19, var = "gene_id")
#Merge gene symbols into counts and gene ID
conv_cts_19 <- merge(cts_df_19, conversion, by = "gene_id")
#Remove unnecessary data
conv_cts_19 <- conv_cts_19[,!(colnames(conv_cts_19) %in% c("gene_id", "group", "group_name"))]
#Filter out repeating data
conv_cts_19 <- dplyr::distinct(conv_cts_19, gene_name, .keep_all = TRUE)
#Convert back to rownames but with gene symbols
conv_cts_19 <- tibble:::column_to_rownames(conv_cts_19, "gene_name")
#Rename the count data to correspond with experimental group or control
colnames(conv_cts_19) <- c("Pkd2KO", "Pkd2KO", "Pkd2KO", "wildtype", "wildtype", "wildtype", "wildtype", "Pkd2KO", "Pkd2KO", "Pkd2KO", "Pkd2KO", "wildtype", "wildtype", "wildtype", "wildtype", "Pkd2KO", "Pkkd2KO", "Pkd2KO", "Pkd2KO", "wildtype", "wildtype", "wildtype", "wildtype")

```


greatest LFC, wrangling
```{r}
dsq2_19 <- as.data.frame(res_19)
genes_19 <-  tibble:::rownames_to_column(dsq2_19, var = "gene_id")
#Merge gene symbols into counts and gene ID
genes_19 <- merge(genes_19, conversion, by = "gene_id")
#Remove unnecessary data
genes_19 <- genes_19[,!(colnames(genes_19) %in% c("gene_id", "group", "group_name"))]
#Filter out repeating data
genes_19 <- dplyr::distinct(genes_19, gene_name, .keep_all = TRUE)
#Convert back to rownames but with gene symbols
genes_19 <-  tibble:::column_to_rownames(genes_19, "gene_name")

genes_19 <- arrange(genes_19, desc(log2FoldChange))#[1:20]
top20lfc_19 <- slice_max(genes_19, (log2FoldChange^2), n = 20)
```

```{r}
heatmap_p28 <- pheatmap(conv_cts_19[rownames(top20lfc_19), ], cluster_rows = TRUE, cluster_cols = TRUE, show_rownames = TRUE, color = viridis(n= 10000, alpha = 1, begin = 0, end = 1, option = "viridis"), border_color = "NA", main = "Top 20 LFC Genes Cystic P28", angle_col = 45)

```

```{r eval=FALSE, include=FALSE}
ggsave(heatmap_p28, filename = "./res/fea/heatmap_p28.pdf", height = 7, width = 14)

```


#### p28 DEGs to p70 Targets

convert from mouse ensembl to human symbol
```{r}
dsq2_19 <- read.csv(here("res", "deseq2_outputs", "deseq2_fullres_19.csv"), row.names = 1)
ens_mouse2human_19 <- tibble:::rownames_to_column(dsq2_19, var = "Mouse.gene.stable.ID")
  #conversion table
ens_mouse2human_19 <- merge(ens_mouse2human_19, ens_annot, by = "Mouse.gene.stable.ID")
#biomart to conver human ens to human entrez
human_19 <- useEnsembl("ensembl", dataset = "hsapiens_gene_ensembl", mirror = "useast")
convgenes_19 <- getBM(filters = "ensembl_gene_id", attributes = c("ensembl_gene_id","entrezgene_id", "hgnc_symbol"), values = ens_mouse2human_19$Gene.stable.ID, mart = human)
  
conv_tbl_19 <- merge(convgenes_19, ens_mouse2human_19, by.x = "ensembl_gene_id", by.y = "Gene.stable.ID")

conv_tbl_19 <- tidyr::drop_na(conv_tbl_19)

```


```{r}

EnhancedVolcano(conv_tbl_19, lab = conv_tbl_19$hgnc_symbol, x = 'log2FoldChange', y = 'padj', selectLab = unique(targets_39$t_gn_sym), pCutoff = 0.05, FCcutoff = 0.5, pointSize = 3.0, labSize = 4.0, col = c("#440154", "#21918c", "#5ec962", "#fde725"), title = "P70 Pre-Cystic Drug Targets Found in P28 Cystic DEGs")#, labFace = 'bold', boxedLabels = TRUE,drawConnectors = TRUE,widthConnectors = 0.5,colConnectors = 'black') 

```


Number of targets differentially expressed in P21
```{r}
targets_p21 <- merge(conv_tbl_56, targets_39, by.x = "hgnc_symbol", by.y = "t_gn_sym")
targets_p21 <- dplyr::filter(targets_p21, log2FoldChange > 2.0 | log2FoldChange < -2.0)

#number of unique drugs
length(unique(targets_p21$pert))

#number of unique targets
length(unique(targets_p21$hgnc_symbol))
```

Number of targets differentially expressed in P28
```{r}
targets_p28 <- merge(conv_tbl_19, targets_39, by.x = "hgnc_symbol", by.y = "t_gn_sym")
targets_p28 <- dplyr::filter(targets_p28, log2FoldChange > 2.0 | log2FoldChange < -2.0)

#number of unique drugs
length(unique(targets_p28$pert))

#number of unique targets
length(unique(targets_p28$hgnc_symbol))
```

expressed in both
```{r}
targets_cystic <- merge(targets_p28, targets_p21, by = "hgnc_symbol")
#number of unique drugs
length(unique(targets_cystic$pert))

#number of unique targets
length(unique(targets_cystic$hgnc_symbol))
```

```{r}
R.Version()
```


```{r}
installed.packages()[names(sessionInfo()$otherPkgs), "Version"]
```
  
