---
title: "pig_transgenic_target_analysis"
author: "Lizzy Wilk"
date: "3/16/2023"
output:
    html_document:
      toc: true
      toc_float: true
      toc_depth: 5

---

### Purpose  
This markdown uses STAR-salmon counts from nf-core for RNAseq alignment. All data from publicly available humanized Pkd2 variant (pathogenic for human ADPKD) from pig kidney RNAseq data comparing Pkd2-mutants and WT.  
__pig__: AKA 39/GSE149739/precystic/p70, induced-KO model, samples collected at p70 when tubules were dilated with no overt cyst formation. Total n = 6 samples, 3 Pkd2fl/fl;Pax8rtTA;TetO-Cre and 3 WT  
Dataset: [GSE149739](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE149739)  
Processed data filepath: /data/project/lasseigne_lab/Lizzy/PKD_drugrepurposing_220324/  

#### Load in libraries
```{r message=FALSE, warning = FALSE}
library(DESeq2)
library(apeglm)
library(tidyverse)
library(ggrepel)
library(signatureSearch)
library(ExperimentHub); library(rhdf5)
library(biomaRt)
library(viridis)
library(here)
#library(EnhancedVolcano)
source(here("code", "functions.R"))
library(gprofiler2)
```

#### Read in data  
Filepath to counts data: /data/project/lasseigne_lab/Lizzy/nfcore_rnaseq/results/star_salmon/salmon.merged.gene_counts.rds  
Metadata and SRA: 

```{r}
pig_rse <- readRDS(here("data/salmon.merged.gene_counts_pkd2pig.rds"))
pig_meta <- read.csv(here("data/metadata_pkd2pig.txt"), row.names = 1)
pig_meta$Genotype <- pig_meta$Ecotype #new column to use as genotype for DDS function
```

# All samples

deseq2 with all samples
```{r}
rownames(pig_meta) <- c("WT_SRR21600987", "WT_SRR21600988", "WT_SRR21600989", "WT_SRR21600990", "PKD2_SRR21600991", "PKD2_SRR21600992", "PKD2_SRR21600993", "PKD2_SRR21600994", "PKD2_SRR21600995")

dds <- salmon2dds(pig_rse@assays@data$counts, pig_meta)
```

run deseq2
```{r}
#relevel to make wildtype the reference instead of first alphabetical factor
dds@colData@listData$Genotype <- relevel(dds@colData@listData$Genotype, ref = "wildtype")
#run DESeq: estimates size factors, estimates dispersion, fits negative binomial GLM's and performs Wald stats
dsq2 <- DESeq(dds) 
```

rlog
```{r}
rld <- rlog(dds)
```

#### PCA
```{r}
ntop = 500
#plotPCA_vst <- function (object,  ntop = 500, assay=length(assays(object))) {
    rv         <- rowVars(assay(rld))
    select     <- order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))]
    pca        <- prcomp(t(assay(rld)[select, ]), center=TRUE, scale=FALSE)
    percentVar <- pca$sdev^2/sum(pca$sdev^2)
    df         <- cbind( as.data.frame(colData(rld)), pca$x)
    #Order points so extreme samples are more likely to get label
    ord        <- order(abs(rank(df$PC1)-median(df$PC1)), abs(rank(df$PC2)-median(df$PC2)))
    df         <- df[ord,]
    attr(df, "percentVar") <- data.frame(PC=seq(along=percentVar), percentVar=100*percentVar)
#    return(df)
#}
pca.data <- df

top <- c(500, Inf)
for (n_top_var in ntop) {
    #pca.data      <- plotPCA_vst(dds, assay=vst_name, ntop=n_top_var)
    percentVar    <- round(attr(pca.data, "percentVar")$percentVar)
    plot_subtitle <- ifelse(n_top_var==Inf, "All genes", paste("Top", n_top_var, "genes"))
    pl <- ggplot(pca.data, aes(PC1, PC2, label=paste0(" ", Genotype, " ", Sample.Name, " "))) +
        geom_point() +
        geom_text(check_overlap=TRUE, vjust=0.5, hjust="inward") +
        xlab(paste0("PC1: ",percentVar[1],"% variance")) +
        ylab(paste0("PC2: ",percentVar[2],"% variance")) +
        labs(title = paste0("First PCs on transformed data"), subtitle = plot_subtitle) +
        theme(legend.position="top",
            #panel.grid.major = element_blank(),
            #panel.grid.minor = element_blank(),
            panel.background = element_blank())#,
#            panel.border = element_rect(colour = "black", fill=NA, size=1))
    print(pl)
}
```


#### Results  

```{r}
#extract results table with log2 fold changes, p values and adjusted p values for desired coefficients
#alpha 0.05 cutoff
resultsNames(dsq2) #list coefficients
res <- lfcShrink(dsq2, coef = c("Genotype_humanizedmutantPKD2transgenicpig_vs_wildtype"), type="apeglm")
res
```

DESeq2 results
```{r}
res_df <- as.data.frame(res)
```

Save full deseq results
```{r}
write.csv(res_df, file = here("res/deseq2_outputs/pig_outlier_included/deseq2_fullres_humanized_pig_pkd2.csv"))
saveRDS(res, file = here("res/deseq2_outputs/pig_outlier_included/deseq2_fullres_humanized_pig_pkd2.rds"))
```

#### LFC cutoffs
Using < -2 and > 2 LFC threshold for up and downregulated genes 
```{r}
res_df <- res_df[!is.na(res_df$padj),]

pig_degs_up <- res_df[res_df$log2FoldChange > 1.0 & res_df$padj < 0.05,]
  #uplfc(res_df, lfcutoff = 1.1) #downfc <- res[res$log2FoldChange < lfcutoff & res$padj < 0.05,]
#pig_degs_down <- downlfc(res_df, lfcutoff = -1.1)
pig_degs_down <- res_df[res_df$log2FoldChange < -1.0 & res_df$padj < 0.05,]
```

Total genes above threshold before mapping from mouse ensembl to human symbol/entrez
```{r}
#from paper:  449 upregulated genes and 389 downregulated genes
dim(pig_degs_up) # at 1.1, 255 at 1.0
dim(pig_degs_down) # at -1.1, 338 at -1.0
```

```{r }
write.csv(pig_degs_up, file = here("res/deseq2_outputs/pig_outlier_included/pig_degs_up.csv"))
write.csv(pig_degs_down, file = here("res/deseq2_outputs/pig_outlier_included/pig_degs_down.csv"))
```


#### Visualizations
PCA
```{r}
plotPCA(rld, intgroup=c("Genotype")) + geom_label_repel(aes(label = rld@colData@rownames), box.padding   = 0.15, label.padding = 0.15, point.padding = 0.3, segment.color = 'grey50') + scale_colour_viridis_d() + theme(axis.title.x = element_text(size = 20)) + theme_minimal()
```

#### Pig to Human Ortholog Mapping
Downloaded from ensembl.org/biomart using pig genes (Sscrofa11.1) and human genes (GRCh38.p13) (filters for orthologous genes only) (accessed 230317)
Map to human entrez and filter by LINCS genes
```{r}
pig_human_conv <- read.delim("../data/mart_export_pig2human.txt")

pig_deseqres_anno <- res_df %>% rownames_to_column(., var = "Gene.stable.ID") %>% inner_join(., pig_human_conv, by = "Gene.stable.ID")

degs_up <- pig_degs_up %>% rownames_to_column(., var = "Gene.stable.ID") %>% inner_join(., pig_human_conv, by = "Gene.stable.ID")

degs_down <- pig_degs_down %>% rownames_to_column(., var = "Gene.stable.ID") %>% inner_join(., pig_human_conv, by = "Gene.stable.ID")

#hom_degs_UP <- lincsGenes(musdegs_p70_UP, "up")
 
#hom_degs_DOWN <-  lincsGenes(musdegs_p70_DOWN, "down")
#should be 100 or less
#length(unique(hom_degs_UP$entrezgene_id)) #99
#length(unique(hom_degs_DOWN$entrezgene_id)) #31
```

```{r}
write.csv(degs_up, file = here("res/deseq2_outputs/pig_outlier_included/pig_humanortho_degs_up.csv"))
write.csv(degs_down, file = here("res/deseq2_outputs/pig_outlier_included/pig_humanortho_degs_down.csv"))
```

#### Compare DEGs to Targets  

```{r}
drug_targets <- read.csv(here("res/sigsearch_outputs/drug_targets_oe_cysticdegs.csv"), row.names = 1)

#drug targets available in pig - human orthologs
drug_targets_avail_pig <- inner_join(drug_targets, pig_deseqres_anno, by = c("Drug_Targets" = "Gene.name.1"))

drug_targets_in_pkd2_pig <- dplyr::filter(drug_targets, Drug_Targets %in% degs_up$Gene.name.1)
```

double check up genes are same as up genes from the paper (pulling from supp Table 1, sheets 3 and 4 for up and down enrichment)
```{r}
library(readxl)
reported_upgenes <- read_xlsx("~/Downloads/Table 1.XLSX", sheet = 3, skip = 1)
reported_upgenes <- separate_rows(reported_upgenes, Genes, sep = ", ")
reported_upgenes <- unique(reported_upgenes$Genes)

reported_downgenes <- read_xlsx("~/Downloads/Table 1.XLSX", sheet = 4, skip = 1)
reported_downgenes <- separate_rows(reported_downgenes, Genes, sep = ", ")
reported_downgenes <- unique(reported_downgenes$Genes)

intersect(degs_up$Gene.name.1, reported_upgenes) #only 41 of reported
#intersect(degs_down$Gene.name.1, reported_upgenes) #triple checking there's few to none that are opposite of what's reported

intersect(degs_down$Gene.name.1, reported_downgenes)#only 30 of reported
#intersect(degs_up$Gene.name.1, reported_downgenes) #triple checking there's few to none that are opposite of what's reported
```



# Remove outlier -- WT sample explaining the majority of variation, separating apart from all other samples in PC1 (full multiQC results )
#### Read in data  
Filepath to counts data: /data/project/lasseigne_lab/Lizzy/nfcore_rnaseq/results/star_salmon/salmon.merged.gene_counts.rds  
Metadata and SRA: 

```{r}
pig_rse <- readRDS(here("data/salmon.merged.gene_counts_pkd2pig.rds"))
pig_meta <- read.csv(here("data/metadata_pkd2pig.txt"), row.names = 1)
pig_meta$Genotype <- pig_meta$Ecotype #new column to use as genotype for DDS function
```


```{r}
pig_cts <- dplyr::select(pig_rse@assays@data$counts, !WT_SRR21600987)
pig_meta <- pig_meta[rownames(pig_meta) != "SRR21600987",]
#update rownames with "WT" or "PKD2" classification
rownames(pig_meta) <- c("WT_SRR21600988", "WT_SRR21600989", "WT_SRR21600990", "PKD2_SRR21600991", "PKD2_SRR21600992", "PKD2_SRR21600993", "PKD2_SRR21600994", "PKD2_SRR21600995")
```

#### DDS
```{r}
dds <- salmon2dds(pig_cts, pig_meta)
```


#### Run DESeq2
```{r}
#relevel to make wildtype the reference instead of first alphabetical factor
dds@colData@listData$Genotype <- relevel(dds@colData@listData$Genotype, ref = "wildtype")
#run DESeq: estimates size factors, estimates dispersion, fits negative binomial GLM's and performs Wald stats
dds <- DESeq(dds) 
```

#### Results  

```{r}
#extract results table with log2 fold changes, p values and adjusted p values for desired coefficients
#alpha 0.05 cutoff
resultsNames(dds) #list coefficients
res <- lfcShrink(dds, coef = c("Genotype_humanizedmutantPKD2transgenicpig_vs_wildtype"), type="apeglm")
res
```

DESeq2 results
```{r}
res_df <- as.data.frame(res)
```

Save full deseq results
```{r}
write.csv(res_df, file = here("res/deseq2_outputs/deseq2_fullres_humanized_pig_pkd2.csv"))
saveRDS(res, file = here("res/deseq2_outputs/deseq2_fullres_humanized_pig_pkd2.rds"))
```

#### LFC cutoffs
Using < -2 and > 2 LFC threshold for up and downregulated genes 
```{r}
res_df <- res_df[!is.na(res_df$padj),]

pig_degs_up <- res_df[res_df$log2FoldChange > 1.0 & res_df$padj < 0.05,]
  #uplfc(res_df, lfcutoff = 1.1) #downfc <- res[res$log2FoldChange < lfcutoff & res$padj < 0.05,]
#pig_degs_down <- downlfc(res_df, lfcutoff = -1.1)
pig_degs_down <- res_df[res_df$log2FoldChange < -1.0 & res_df$padj < 0.05,]
```

Total genes above threshold before mapping from mouse ensembl to human symbol/entrez
```{r}
dim(pig_degs_up) #548 at 1.1, 628 at 1.0 
dim(pig_degs_down) #566 at -1.1, 624 at 1.0
##from paper:  449 upregulated genes and 389 downregulated genes
```

```{r }
write.csv(pig_degs_up, file = here("res/deseq2_outputs/pig_degs_up.csv"))
write.csv(pig_degs_down, file = here("res/deseq2_outputs/pig_degs_down.csv"))
```


Regularized log transformation  
rlog() is better to use for these datasets -- better for high count variance by sample (in this case, one sample has over 2x the reads compared to the others in one data set) and for smaller datasets "if you have many samples (e.g. 100s), the rlog function might take too long, and so the vst function will be a faster choice. The rlog and VST have similar properties, but the rlog requires fitting a shrinkage term for each sample and each gene which takes time. See the DESeq2 paper for more discussion on the differences (Love, Huber, and Anders 2014)."  
```{r}
#vsd_nf <- vst(dds_p70, blind = FALSE) #if blind, could minimize large count differences due to anticipated experimental variation
rld <- rlog(dds, blind = FALSE) 
```

#### Visualizations
PCA
```{r}
plotPCA(rld, intgroup=c("Genotype")) + geom_label_repel(aes(label = rld@colData@rownames), box.padding   = 0.15, label.padding = 0.15, point.padding = 0.3, segment.color = 'grey50') + scale_colour_viridis_d() + theme(axis.title.x = element_text(size = 20)) + theme_minimal()
```

#### Pig to Human Ortholog Mapping
Downloaded from ensembl.org/biomart using pig genes (Sscrofa11.1) and human genes (GRCh38.p13) (filters for orthologous genes only) (accessed 230317)
Map to human ortholog genes
```{r}
pig_human_conv <- read.delim("../data/mart_export_pig2human.txt")

pig_deseqres_anno <- res_df %>% rownames_to_column(., var = "Gene.stable.ID") %>% inner_join(., pig_human_conv, by = "Gene.stable.ID")

human_conv <- gprofiler2::gconvert(pig_deseqres_anno$Gene.stable.ID.1, organism = "hsapiens", target = "ENTREZGENE_ACC")

pig_human_conv <- left_join(pig_human_conv, human_conv, by = c("Gene.stable.ID.1" = "input"))
```

up and down gene lists w/ gene annotations (ensembl, entrez, names, and pig - human orthologs)
```{r}
degs_up <- pig_degs_up %>% rownames_to_column(., var = "Gene.stable.ID") %>% inner_join(., pig_human_conv, by = "Gene.stable.ID", multiple = "all")

degs_down <- pig_degs_down %>% rownames_to_column(., var = "Gene.stable.ID") %>% inner_join(., pig_human_conv, by = "Gene.stable.ID", multiple = "all")

```

```{r}
write.csv(degs_up, file = here("res/deseq2_outputs/pig_humanortho_degs_up.csv"))
write.csv(degs_down, file = here("res/deseq2_outputs/pig_humanortho_degs_down.csv"))
```

#### Compare DEGs to Targets  

```{r}
drug_targets <- read.csv(here("res/sigsearch_outputs/drug_targets_oe_cysticdegs.csv"), row.names = 1)

#drug targets available in pig - human orthologs
drug_targets_avail_pig <- inner_join(drug_targets, pig_deseqres_anno, by = c("Drug_Targets" = "Gene.name.1"))

drug_targets_in_pkd2_pig <- dplyr::filter(drug_targets, Drug_Targets %in% degs_up$Gene.name.1)
```

double check up genes are same as up genes from the paper (pulling from supp Table 1, sheets 3 and 4 for up and down enrichment)
```{r}
library(readxl)
reported_upgenes <- read_xlsx("~/Downloads/Table 1.XLSX", sheet = 3, skip = 1)
reported_upgenes <- separate_rows(reported_upgenes, Genes, sep = ", ")
reported_upgenes <- unique(reported_upgenes$Genes)

reported_downgenes <- read_xlsx("~/Downloads/Table 1.XLSX", sheet = 4, skip = 1)
reported_downgenes <- separate_rows(reported_downgenes, Genes, sep = ", ")
reported_downgenes <- unique(reported_downgenes$Genes)

intersect(degs_up$Gene.name.1, reported_upgenes) #only 46 of reported
#intersect(degs_down$Gene.name.1, reported_upgenes) #triple checking there's few to none that are opposite of what's reported

intersect(degs_down$Gene.name.1, reported_downgenes)#only 31 of reported
#intersect(degs_up$Gene.name.1, reported_downgenes)
```

##### Top Drugs, All Targets Comparison  
Instead of just the prioritized targets, looking at all prioritzed targets of the top drugs
```{r}
prioritized_drugs <- read.csv(here("res/sigsearch_outputs/drugtarget_cysticoverlap_simpletbl.csv"), row.names = 1)
prioritized_drugs <- tidyr::separate_rows(prioritized_drugs, Drug_Targets, sep = "; ")

#prioritized_drugs_pig_oe <- filter(prioritized_drugs, Drug_Targets %in% pig_humanortho_degs_up$Gene.name.1)
prioritized_drugs_pig_oe <- inner_join(prioritized_drugs,pig_humanortho_degs_up, by = c( "Drug_Targets" = "Gene.name.1"))
```

```{r}
#unique drugs
unique(prioritized_drugs_pig_oe$Drug)
```

```{r}
#unique targets
#unique drugs
unique(prioritized_drugs_pig_oe$Drug_Targets)
```


#### Plots
alluvial
```{r}
target_alluvial <- ggplot(data = prioritized_drugs_pig_oe,
       aes(axis1 = Drug, axis2 = Drug_Targets)) +
  geom_alluvium(color = "black", aes(fill = Drug_Targets %in% drug_targets_in_pkd2_pig$Drug_Targets)) + # , colour = trend)) + 
  geom_stratum(width = 3/12, color = "black") + #the border of the stratum
  #scale_x_discrete(expand = c(.07, .07)) + 
  #scale_y_discrete(expand = c(.007, .007)) + 
  scale_fill_viridis_d("Target \nUpregulated in \nPkd2-KO Cystic Mice", aesthetics = c("colour", "fill")) +
  geom_text(stat = "stratum", size = 3.5, fontface = "bold", aes(label = after_stat(stratum))) +
  #labs(title = "Prioritized Drugs with Targets \nUpregulated in Pkd2-Pig") +
  #theme(legend.title = element_text("Target Upregulated in Pkd2-KO Cystic Mice")) 
  theme(axis.title.x = element_blank(), plot.title = element_text(size = 18, face = "bold", hjust = 0.5, vjust = -3.5), axis.text.x = element_blank(), axis.text.y = element_blank(), axis.ticks = element_blank()) +
        #, title = element_text("Target Upregulated in Pkd2-KO Cystic Mice")) + 
  ggtitle(label = "Prioritized Drugs with \nTargets Upregulated in Pkd2-Pig") + 
  theme(panel.grid.major = element_blank(), panel.grid.minor = element_blank(),
panel.background = element_blank() )

target_alluvial
ggsave("../res/paper_supplementary/pig_drugtarget_alluvial.png", target_alluvial)
```
target_deg_plot <- function(targetdegs, cutoff, title){

  #targetdegs <- dplyr::distinct(targetdegs, t_gn_sym, .keep_all = TRUE)

  volcanoplot <- EnhancedVolcano(targetdegs$degs, lab = targetdegs$degs$hgnc_symbol,
  x = 'log2FoldChange', y = 'padj', selectLab = unique(targetdegs$targetdegs$t_gn_sym),
  pCutoff = 0.05, FCcutoff = cutoff, pointSize = 3.0, labSize = 5.0,
  col = c("#440154", "#21918c", "#fde725", "#5ec962"), title = title,
  labFace = 'bold', boxedLabels = TRUE,drawConnectors = TRUE,widthConnectors = 0.5,colConnectors = 'black',
  max.overlaps = 25)
  return(volcanoplot)
}
volcano plot
```{r}
#vp_df <- res_df %>% rownames_to_column(., var = "Gene.Stable.ID") %>% full_join(., prioritized_drugs_pig_oe, by = "Gene.Stable.ID", multiple = "all")

vp <- EnhancedVolcano::EnhancedVolcano(pig_deseqres_anno, lab = pig_deseqres_anno$Gene.name.1,
  x = 'log2FoldChange', y = 'padj', selectLab = unique(prioritized_drugs_pig_oe$Drug_Targets,),
  pCutoff = 0.05, FCcutoff = 1, pointSize = 3.0, labSize = 5.0,
  col = c("#440154", "#21918c", "#fde725", "#5ec962"), title = "Prioritized Drugs with \nTargets Upregulated in Pkd2-Pig",
  labFace = 'bold', boxedLabels = TRUE,drawConnectors = TRUE,widthConnectors = 0.5,colConnectors = 'black',
  max.overlaps = 25)

vp
ggsave("../res/paper_supplementary/pig_drugtarget_volcanoplot.png", vp)
```

## Signature by DEGs to LINCS-Gene Mapping
Filter for LINCS-available genes and then take top 100 up and down DEGs for signatures
```{r}
eh <- ExperimentHub()
lincs <- eh[["EH3226"]]; lincs_expr <- eh[["EH3227"]]

lincs_genes <- rhdf5:::h5read(lincs, "rownames", drop=TRUE)

sig_up <- degs_up %>% filter(., target %in% lincs_genes) %>% distinct(target, .keep_all = TRUE) %>% slice_max(., n = 100, order_by = log2FoldChange)

sig_down <- degs_down %>% filter(., target %in% lincs_genes) %>% distinct(target, .keep_all = TRUE)%>% slice_min(., n = 100, order_by = log2FoldChange)
```

```{r}
write.csv(sig_up, file = here("res/deseq2_outputs/pig_signature_up.csv"))
write.csv(sig_down, file = here("res/deseq2_outputs/pig_signature_down.csv"))
```


```{r}
R.Version()
```


```{r}
installed.packages()[names(sessionInfo()$otherPkgs), "Version"]
```
