---
title: "sigsearch_GSE69556"
author: "Lizzy Ramsey"
date: "4/7/2022"
output:
    html_document:
      toc: true
      toc_depth: 4
      toc_float: true

---

```{r setup, include=FALSE}
#filepath for code and data
knitr::opts_knit$set(root.dir = "/Users/eramsey/Downloads/kidney_drugrepurposing/pkd_drugrepurposing")

```

### Purpose  
This markdown follows differential expression analyses from deseq2_GSE134719.Rmd, deseq2_GSE149739.Rmd, and deseq2_GSE69556.Rmd. DEGs from these markdowns are used as input for signatureSearch. Results are annotated with Drug Repurposing Hub and onSides data.  
__Data Sets__
- '39 / Pre-Cystic P70: [GSE149739](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE149739)  
- '19 / Cystic P21: [GSE134719](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE134719)  
- '56 / Cystic P28: [GSE69556](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE69556)  

Load libraries  
```{r message=FALSE, warning = FALSE}
library(signatureSearch)
library(ggplot2)
library(ggalluvial)
library(readr)
library(signatureSearch)
library(ExperimentHub); library(rhdf5)
```

Pull LINCS data from Experiment Hub
```{r message=FALSE, warning = FALSE}
eh <- ExperimentHub()
cmap <- eh[["EH3223"]]; cmap_expr <- eh[["EH3224"]]
lincs <- eh[["EH3226"]]; lincs_expr <- eh[["EH3227"]]
h5ls(lincs)
db_path <- system.file("extdata", "sample_db.h5", package = "signatureSearch")

```

#### Custom functions  
__ssRun__: Wrapper for qsiq(), signature query, using up and down genes, and gess_lincs(), the 'lincs' signatureSearch method for gene expression signature search, sorting the results by WTCS and calculating Tau values, and then filtering for inversely related signatures and for kidney cell lines (HA1E and NKDBA) only  
```{r}
ssRun <- function(upgenes, downgenes, cells){
  q <- qSig(query = list(upset=as.character(upgenes), downset=as.character(downgenes)), gess_method="LINCS", refdb= lincs)
  gess <- gess_lincs(q, sortby="WTCS", tau=TRUE, workers=1)
  gess_res <- result(gess)
  gess_res <- dplyr::filter(gess_res, WTCS < 0)
  if(cells == "kidney"){
    gess_kidney <- gess_res %>% dplyr::filter(cell == "HA1E" | cell == "NKDBA")
    return(gess_kidney)
  }
  else{
return(gess_res)
  }
}

```


### GSE149739 (precystic dataset)

Load in top up and down DEGs  
```{r}
hom_degs_UP <- read.csv("/Users/eramsey/Downloads/kidney_drugrepurposing/pkd_drugrepurposing/res/deseq2_outputs/GSE149739_degs_UP_220421.csv")

hom_degs_DOWN <- read.csv("/Users/eramsey/Downloads/kidney_drugrepurposing/pkd_drugrepurposing/res/deseq2_outputs/GSE149739_degs_DOWN_220421.csv")
```

Run signatureSearch, filter for kidney cell lines  
```{r}
gess_kidneycells_39 <- ssRun(upgenes = hom_degs_UP$entrezgene_id, downgenes = hom_degs_DOWN$entrezgene_id, cells = "kidney")
```

#### Visualize Top 20
```{r}
gess_res_vis(gess_kidneycells_39, drugs = unique(gess_kidneycells_39$pert)[1:20], col = "WTCS")
```

```{r}
knitr::kable(gess_kidneycells_39, col.names = colnames(gess_kidneycells_39))

```

```{r eval=FALSE, include=FALSE}
write.csv(gess_kidneycells_39, "./res/sigsearch_outputs/gess_kidneycells_39_220421.csv")
```

## GSE69556  

Load in top up and down DEGs  
```{r}
hom_56_UP <- read.csv("/Users/eramsey/Downloads/kidney_drugrepurposing/pkd_drugrepurposing/res/deseq2_outputs/GSE69556_degs_UP_220421.csv")

hom_56_DOWN <- read.csv("/Users/eramsey/Downloads/kidney_drugrepurposing/pkd_drugrepurposing/res/deseq2_outputs/GSE69556_degs_DOWN_220421.csv")
```

Run signatureSearch, filter for kidney cell lines  
```{r}
gess_kidneycells_56 <- ssRun(upgenes = as.character(hom_56_UP$entrezgene_id), downgenes = as.character(hom_56_DOWN$entrezgene_id), cells = "kidney")
```

#### Visualize Top 20
```{r}
gess_res_vis(gess_kidneycells_56, drugs = unique(gess_kidneycells_56$pert)[1:20], col = "WTCS")
```

```{r}
knitr::kable(gess_kidneycells_56, col.names = colnames(gess_kidneycells_56))

```

Save gess and filtered gess result
```{r eval=FALSE, include=FALSE}
write.csv(gess_kidneycells_56, "./res/sigsearch_outputs/gess_kidneycells_56_220421.csv")
```

## GSE134719  
Read in data  
```{r}
hom_19_UP <- read.csv("/Users/eramsey/Downloads/kidney_drugrepurposing/pkd_drugrepurposing/res/deseq2_outputs/GSE134719_degs_UP_220421.csv")
hom_19_DOWN <- read.csv("/Users/eramsey/Downloads/kidney_drugrepurposing/pkd_drugrepurposing/res/deseq2_outputs/GSE134719_degs_DOWN_220421.csv")
```

Run signatureSearch, filter for kidney cell lines  
```{r}
gess_kidneycells_19 <- ssRun(upgenes = as.character(hom_19_UP$entrezgene_id), downgenes = as.character(hom_19_DOWN$entrezgene_id), cells = "kidney")
```

#### Visualize Top 20
```{r}
gess_res_vis(gess_kidneycells_19, drugs = unique(gess_kidneycells_19$pert)[1:20], col = "WTCS")
```

```{r}
knitr::kable(gess_kidneycells_19, col.names = colnames(gess_kidneycells_19))
```

Save gess and filtered gess result
```{r eval=FALSE, include=FALSE}
write.csv(gess_kidneycells_19, "./res/sigsearch_outputs/gess_kidneycells_19_220421.csv")
```

## Compare datasets  
Dimensions  
```{r}
dim(gess_kidneycells_39)
dim(gess_kidneycells_56)
dim(gess_kidneycells_19)
```

Drug overlaps  
```{r}
overlap_39_56 <- dplyr::intersect(gess_kidneycells_39$pert, gess_kidneycells_56$pert) 

overlap_39_19 <- dplyr::intersect(gess_kidneycells_39$pert, gess_kidneycells_19$pert) 

overlap_56_19 <- dplyr::intersect(gess_kidneycells_56$pert, gess_kidneycells_19$pert) 

```

Number of drug overlaps  
```{r}
length(overlap_39_56)
length(overlap_39_19)
length(overlap_56_19)

```

### SignatureSearch Plots  
```{r}
bggenes <- read.csv(file = "./data/background_genelist_genestoLINCSgenes.csv", row.names = 1)
#KEGG enrichment based on hypergeometric test
kegg_39 <- enrichKEGG2(gene = as.character(c(hom_degs_UP$entrezgene_id, hom_degs_DOWN$entrezgene_id)), universe = bggenes$entrezgene_id)
moa_39 <- enrichMOA(gene = as.character(c(hom_degs_UP$entrezgene_id, hom_degs_DOWN$entrezgene_id)))
```

#### TSEA  
Target Set Enrichment Analysis. Uses target protein, cellular networks and pathways affected by the top ranking drugs  

'39 TSEA 
```{r warning = FALSE}
#target set enrichment analysis TSEA, looks at target proteins for the drugs and thedrug MOA's or protein family domains
tsea_GO_39 <- tsea_dup_hyperG(drugs = gess_kidneycells_39$pert[1:100], universe = "Default", type = "GO", ont="MF", pvalueCutoff=0.05, pAdjustMethod="BH")

tsea_KEGG_39 <- tsea_dup_hyperG(drugs = gess_kidneycells_39$pert[1:50], universe = "Default", type = "KEGG", pvalueCutoff=0.05, pAdjustMethod="BH")

```


```{r warning = FALSE}
dtnetplot(drugs = drugs(tsea_KEGG_39), set = "hsa04024", ont = "MF", desc="cAMP signaling pathway")
```

```{r warning = FALSE}
dtnetplot(drugs = drugs(tsea_KEGG_39), set = "hsa04750", ont = "MF", desc="Inflammatory mediator regulation of TRP channels")
```

Save TSEA tables
```{r eval=FALSE, include=FALSE}
write.csv(tsea_GO_39@result, file = "./res/sigsearch_outputs/tsea_go_39.csv")

write.csv(tsea_KEGG_39@result, file = "./res/sigsearch_outputs/tsea_kegg_39.csv")
```

'56 TSEA 
```{r}
tsea_GO_56 <- tsea_dup_hyperG(drugs = gess_kidneycells_56$pert[1:30], universe = "Default", type = "GO", ont="MF", pvalueCutoff=0.05, pAdjustMethod="BH")

tsea_KEGG_56 <- tsea_dup_hyperG(drugs = gess_kidneycells_56$pert[1:30], universe = "Default", type = "KEGG", pvalueCutoff=0.05, pAdjustMethod="BH")

dtnetplot(drugs = drugs(tsea_GO_56), set = "GO:0004708", ont = "MF", desc="MAP kinase kinase activity")

```

```{r warning = FALSE}
dtnetplot(drugs = drugs(tsea_GO_56), set = "GO:0016712", ont = "MF", desc="oxidoreductase activity, acting on paired donors,")

dtnetplot(drugs = drugs(tsea_GO_56), set = "GO:0005262", ont = "MF", desc="calcium channel activity")
```

```{r}
dtnetplot(drugs = drugs(tsea_KEGG_56), set = "hsa04657", desc="IL-17 signaling pathway")
```

Save TSEA tables
```{r eval=FALSE, include=FALSE}
write.csv(tsea_GO_56@result, file = "./res/sigsearch_outputs/tsea_go_56.csv")

write.csv(tsea_KEGG_56@result, file = "./res/sigsearch_outputs/tsea_kegg_56.csv")
```

'19 TSEA
```{r}
tsea_GO_19 <- tsea_dup_hyperG(drugs = gess_kidneycells_19$pert[1:30], universe = "Default", type = "GO", ont="MF", pvalueCutoff=0.05, pAdjustMethod="BH")

tsea_KEGG_19 <- tsea_dup_hyperG(drugs = gess_kidneycells_19$pert[1:30], universe = "Default", type = "KEGG", pvalueCutoff=0.05, pAdjustMethod="BH")

dtnetplot(drugs = drugs(tsea_GO_19), set = "GO:0051117", ont = "MF", desc="ATPase binding")
```

```{r}
dtnetplot(drugs = drugs(tsea_GO_19), set = "GO:0051117", ont = "MF", desc="ATPase binding")

```

Save TSEA tables
```{r eval=FALSE, include=FALSE}
write.csv(tsea_GO_19@result, file = "./res/sigsearch_outputs/tsea_go_19.csv")

write.csv(tsea_KEGG_19@result, file = "./res/sigsearch_outputs/tsea_kegg_19.csv")
```

#####Plots TSEA  

Read back in  
```{r}
tsea_GO_39 <- read.csv(file = "./res/sigsearch_outputs/tsea_go_39.csv", row.names = 1)

```


#### DSEA
```{r}
dsea_39 <- dsea_hyperG(
  drugs = gess_kidneycells_39$pert[1:100],
  type = "GO",
  ont = "MF",
  pvalueCutoff = 0.05,
  pAdjustMethod = "BH",
  qvalueCutoff = 0.2,
  minGSSize = 10,
  maxGSSize = 500
)
```


```{r eval=FALSE, include=FALSE}
write.csv(dsea_39@result, file = "./res/sigsearch_outputs/dsea_39.csv")
dsea_39 <- read.csv(file = "./res/sigsearch_outputs/dsea_39.csv")
```


### Drug Annotations  
LINCS [Drug Repurposing Hub](https://www.nature.com/articles/nm.4306), contains drug names, MOAs, original indications, and targets
```{r}
repurpHub <- read.delim("./data/Repurposing_Hub_export.txt") #6798 drugs

rHub_19 <- dplyr::inner_join(gess_kidneycells_19, repurpHub, by= c("pert" = "Name"))

rHub_56 <- dplyr::inner_join(gess_kidneycells_56, repurpHub, by= c("pert" = "Name"))

rHub_39 <- dplyr::inner_join(gess_kidneycells_39, repurpHub, by= c("pert" = "Name"))
```

P70/'39 Already-Launched Drugs
```{r}
rhub_launched_39 <- dplyr::filter(rHub_39, Phase == "Launched")

write.csv(rhub_launched_39, "./res/sigsearch_outputs/rhub_launcheddrugs_39.csv")
```

Alluvial plot of drugs-targets with just launched drugs
```{r}
rhub_launched_39_top30 <- dplyr::slice_min(rhub_launched_39, order_by = Tau, n = 20)

rhub_launched_39_top_septarget <- tidyr::separate_rows(rhub_launched_39_top30, Target, sep = ", ")
#remove drugs with empty 'target' field
rhub_launched_39_top_septarget <- dplyr::filter(rhub_launched_39_top_septarget, Target != "")
```

plot
```{r}
target_alluvial <- ggplot(data = rhub_launched_39_top_septarget,
       aes(axis1 = pert, axis2 = Target)) +
  geom_alluvium(color = "black", aes(fill = WTCS)) + # , colour = trend)) + 
  geom_stratum(width = 3/12, color = "black") + #the border of the stratum
  scale_x_discrete(expand = c(.07, .07)) + 
  scale_y_discrete(expand = c(.007, .007)) + 
  scale_fill_viridis_c(aesthetics = c("colour", "fill")) +
  geom_text(stat = "stratum", size = 3.5, fontface = "bold", aes(label = after_stat(stratum))) +
  labs(title = "Targets of Top 20 Launched Drugs") +
  theme_void() 


target_alluvial
```

Save
```{r eval=FALSE, include=FALSE}
#target-separated, rhub-annotated drug results for 39
gene_drug_targets <- rhub_launched_39_top_septarget
write.csv(gene_drug_targets, file = "./res/sigsearch_outputs/gene_drug_targets_39.csv")
```


MOA's separated, string wrap labels
```{r}
rhub_launched_39_top_moasep <- tidyr::separate_rows(rhub_launched_39_top30, MOA, sep = ",")

rhub_launched_39_top_moasep$MOA <- stringr::str_wrap(rhub_launched_39_top_moasep$MOA, width = 20)

rhub_launched_39_top_moasep$Disease.Area <- stringr::str_wrap(rhub_launched_39_top_moasep$Disease.Area, width = 30)
```

plot
```{r}
moa_alluvial <- ggplot(data = rhub_launched_39_top_moasep,
       aes(axis1 = pert, axis2 = MOA)) +
  geom_alluvium(color = "black", aes(fill = WTCS)) + # , colour = trend)) + 
  geom_stratum(width = 3/12, color = "black") + #the border of the stratum
  scale_x_discrete(expand = c(.07, .07)) + 
  scale_y_discrete(expand = c(.007, .007)) + 
  scale_fill_viridis_c(aesthetics = c("colour", "fill")) +
  geom_text(stat = "stratum", size = 3.5, fontface = "bold", aes(label = after_stat(stratum))) +
  labs(title = "MOA of Top 30 Launched Drugs") +
  theme_void() 


moa_alluvial
```

```{r eval=FALSE, include=FALSE}
ggsave("./figures/moa_alluvial_wtcs_2.pdf", moa_alluvial, width = 10, height = 15)

```


Combined repurposing hub annotated dataset drugs
```{r}
rHub_19_and_56 <-  merge(rHub_19, rHub_56, by = "pert")

rHub_all_inters <- merge(rHub_19_and_56, rHub_39,  by = "pert")
```


Save
```{r eval=FALSE, include=FALSE}
write.csv(rHub_19, file = "./res/sigsearch_outputs/ssresults_rHub_19.csv")

write.csv(rHub_56, file = "./res/sigsearch_outputs/ssresults_rHub_56.csv")

write.csv(rHub_39, file = "./res/sigsearch_outputs/ssresults_rHub_39.csv")

write.csv(rHub_19_and_56, file = "./res/sigsearch_outputs/ssresults_rHub_19_and_56.csv")

write.csv(rHub_all_inters, file = "./res/sigsearch_outputs/ssresults_rHub_all_inters.csv")

```


##### Side effects  
Data from [onSides](https://github.com/tatonetti-lab/onsides)  
__adverse_reactions_bylabel__ - All extracted adverse reactions from the ADVERSE REACTIONS section of all labels. Each drug will have multiple labels over its lifetime (revisions, generic alternatives, etc.). This table contains the results of extracting adverse reactions from every label available for download from DailyMed. 2,764,338 rows.  
__label_map__ - Map between the xml_id (xml filename), zip_id (zip filename), and set_id (drug identifier). 45,989 rows.  

```{r}
reactions <- read.csv("/Users/eramsey/Downloads/kidney_drugrepurposing/pkd_drugrepurposing/data/onsides_v01/adverse_reactions.csv")

reactions_unique <- unique(reactions$concept_name)

warnings <- read.csv("/Users/eramsey/Downloads/kidney_drugrepurposing/pkd_drugrepurposing/data/onsides_v01/boxed_warnings.csv")

labels <- read.csv("/Users/eramsey/Downloads/kidney_drugrepurposing/pkd_drugrepurposing/data/onsides_v01/latest_labels_bydrug.csv")

```

Looking at reactions for 'launched' drugs from P70/'39 results
```{r}
onsides_39 <- dplyr::left_join(rhub_launched_39, labels, by = c("pert" = "ingredients"))

onsides_39 <- dplyr::left_join(onsides_39, reactions, by = c("latest_xml_id" = "xml_id"))

```


```{r}
write.csv(onsides_39, file = "./res/sigsearch_outputs/adr_onsides_precystic_drugcandidates")
```

#### Versions  
```{r}
R.Version()
```


```{r}
installed.packages()[names(sessionInfo()$otherPkgs), "Version"]
```
  
